diff --git a/amdgpu/amdgpu_bo.c b/amdgpu/amdgpu_bo.c
--- a/amdgpu/amdgpu_bo.c
+++ b/amdgpu/amdgpu_bo.c
@@ -78,6 +78,10 @@ drm_public int amdgpu_bo_alloc(amdgpu_device_handle dev,
	args.in.bo_size = alloc_buffer->alloc_size;
	args.in.alignment = alloc_buffer->phys_alignment;
 
+	if (alloc_buffer->preferred_heap & AMDGPU_GEM_DOMAIN_VRAM) {
+		  alloc_buffer->flags |= AMDGPU_GEM_CREATE_VRAM_CLEARED;
+	  }
+
	/* Set the placement. */
	args.in.domains = alloc_buffer->preferred_heap;
	args.in.domain_flags = alloc_buffer->flags;

