diff --git a/services/core/java/com/android/server/am/ProcessList.java b/services/core/java/com/android/server/am/ProcessList.java

--- a/services/core/java/com/android/server/am/ProcessList.java
+++ b/services/core/java/com/android/server/am/ProcessList.java
@@ -2588,7 +2588,7 @@ public final class ProcessList {
                                         "Unable to create process group for "
                                         + app.processName + " (" + startResult.pid + ")");
                             } else {
-                                throw new AssertionError("Unable to create process group for "
+                                Slog.e(ActivityManagerService.TAG, "Unable to create process group for "
                                     + app.processName + " (" + startResult.pid + ")");
                             }
                         } else {

