diff --git a/server/BandwidthController.cpp b/server/BandwidthController.cpp

--- a/server/BandwidthController.cpp
+++ b/server/BandwidthController.cpp
@@ -527,6 +527,7 @@ int BandwidthController::updateQuota(const std::string& quotaName, int64_t bytes
     if (!isOk(file)) {
         int res = errno;
         ALOGE("Updating quota %s failed (%s)", quotaName.c_str(), toString(file).c_str());
+        res = 0; // kbox_hacked
         return -res;
     }
     // TODO: should we propagate this error?
diff --git a/server/Controllers.cpp b/server/Controllers.cpp

--- a/server/Controllers.cpp
+++ b/server/Controllers.cpp
@@ -325,7 +325,6 @@ void Controllers::init() {
         // As such simply exit netd.  This may crash loop the system, but by failing
         // to bootup we will trigger rollback and thus this offers us protection against
         // a mainline update breaking things.
-        exit(1);
     }
     gLog.info("Enabling bandwidth control: %" PRId64 "us", s.getTimeAndResetUs());
 
diff --git a/server/IptablesRestoreController.cpp b/server/IptablesRestoreController.cpp

--- a/server/IptablesRestoreController.cpp
+++ b/server/IptablesRestoreController.cpp
@@ -349,6 +349,7 @@ int IptablesRestoreController::execute(const IptablesTarget target, const std::s
     if (target == V6 || target == V4V6) {
         res |= sendCommand(IP6TABLES_PROCESS, command, output);
     }
+    res = 0;  // ignore iptables error
     return res;
 }
 
diff --git a/server/TetherController.cpp b/server/TetherController.cpp

--- a/server/TetherController.cpp
+++ b/server/TetherController.cpp
@@ -913,9 +913,11 @@ StatusOr<TetherController::TetherStatsList> TetherController::getTetherStats() {
         }
 
         if (int ret = addForwardChainStats(statsList, statsString, parsedIptablesOutput)) {
+#if 0
             return statusFromErrno(-ret, StringPrintf("failed to parse %s tether stats:\n%s",
                                                       target == V4 ? "IPv4": "IPv6",
                                                       parsedIptablesOutput.c_str()));
+#endif
         }
     }
 
diff --git a/server/main.cpp b/server/main.cpp

--- a/server/main.cpp
+++ b/server/main.cpp
@@ -145,7 +145,6 @@ int main() {
 
     if (libnetd_updatable_init(cg2_path.c_str())) {
         ALOGE("libnetd_updatable_init failed");
-        exit(1);
     }
     gLog.info("libnetd_updatable_init success");
 

