diff --git a/drivers/Kconfig b/drivers/Kconfig
index 840137d90..024c8d51c 100644
--- a/drivers/Kconfig
+++ b/drivers/Kconfig
@@ -246,4 +246,8 @@ source "drivers/cpuinspect/Kconfig"
 
 source "drivers/arm/Kconfig"
 
+if KBOX
+	source "drivers/tango32-module/Kconfig"
+endif
+
 endmenu
diff --git a/drivers/Makefile b/drivers/Makefile
index 3f0e6f169..10ffefce5 100644
--- a/drivers/Makefile
+++ b/drivers/Makefile
@@ -196,3 +196,4 @@ obj-$(CONFIG_COUNTER)		+= counter/
 obj-$(CONFIG_ROH)		+= roh/
 obj-$(CONFIG_UB)      += ub/
 obj-$(CONFIG_ARM_SPE_MEM_SAMPLING)		+= arm/spe/
+obj-$(CONFIG_TANGO32)		+= tango32-module/
-- 
2.33.0

