diff --git a/data/keyboards/Generic.kl b/data/keyboards/Generic.kl
index f8d3bffbe00b..b547c8bd72e8 100644
--- a/data/keyboards/Generic.kl
+++ b/data/keyboards/Generic.kl
@@ -22,8 +22,8 @@
 
 key 1     ESCAPE
 key 2     1
-key 3     2
-key 4     3
+key 3     HOME
+key 4     BACK
 key 5     4
 key 6     5
 key 7     6
@@ -206,7 +206,7 @@ key 180   NUMPAD_RIGHT_PAREN
 # key 184   F14
 # key 185   F15
 # key 186   F16
-# key 187   F17
+key 187   APP_SWITCH
 # key 188   F18
 # key 189   F19
 # key 190   F20
diff --git a/services/core/jni/gnss/Gnss.cpp b/services/core/jni/gnss/Gnss.cpp
index da8928b5f97f..bf62044c68e6 100644
--- a/services/core/jni/gnss/Gnss.cpp
+++ b/services/core/jni/gnss/Gnss.cpp
@@ -48,35 +48,6 @@ using android::hardware::gnss::IGnssPsds;
 
 namespace {
 
-GnssLocationAidl createGnssLocation(jint gnssLocationFlags, jdouble latitudeDegrees,
-                                    jdouble longitudeDegrees, jdouble altitudeMeters,
-                                    jfloat speedMetersPerSec, jfloat bearingDegrees,
-                                    jfloat horizontalAccuracyMeters, jfloat verticalAccuracyMeters,
-                                    jfloat speedAccuracyMetersPerSecond,
-                                    jfloat bearingAccuracyDegrees, jlong timestamp,
-                                    jint elapsedRealtimeFlags, jlong elapsedRealtimeNanos,
-                                    jdouble elapsedRealtimeUncertaintyNanos) {
-    GnssLocationAidl location;
-    location.gnssLocationFlags = static_cast<int>(gnssLocationFlags);
-    location.latitudeDegrees = static_cast<double>(latitudeDegrees);
-    location.longitudeDegrees = static_cast<double>(longitudeDegrees);
-    location.altitudeMeters = static_cast<double>(altitudeMeters);
-    location.speedMetersPerSec = static_cast<double>(speedMetersPerSec);
-    location.bearingDegrees = static_cast<double>(bearingDegrees);
-    location.horizontalAccuracyMeters = static_cast<double>(horizontalAccuracyMeters);
-    location.verticalAccuracyMeters = static_cast<double>(verticalAccuracyMeters);
-    location.speedAccuracyMetersPerSecond = static_cast<double>(speedAccuracyMetersPerSecond);
-    location.bearingAccuracyDegrees = static_cast<double>(bearingAccuracyDegrees);
-    location.timestampMillis = static_cast<uint64_t>(timestamp);
-
-    location.elapsedRealtime.flags = static_cast<int>(elapsedRealtimeFlags);
-    location.elapsedRealtime.timestampNs = static_cast<uint64_t>(elapsedRealtimeNanos);
-    location.elapsedRealtime.timeUncertaintyNs =
-            static_cast<double>(elapsedRealtimeUncertaintyNanos);
-
-    return location;
-}
-
 GnssLocation_V1_0 createGnssLocation_V1_0(jint gnssLocationFlags, jdouble latitudeDegrees,
                                           jdouble longitudeDegrees, jdouble altitudeMeters,
                                           jfloat speedMetersPerSec, jfloat bearingDegrees,
@@ -133,7 +104,7 @@ GnssHal::GnssHal() {
     if (gnssHalAidl != nullptr) {
         ALOGD("Successfully got GNSS AIDL handle. Version=%d.", gnssHalAidl->getInterfaceVersion());
         if (gnssHalAidl->getInterfaceVersion() >= 2) {
-            return;
+            ALOGD("Return to GNSS HIDL handle.");
         }
     }
 
@@ -235,11 +206,6 @@ void GnssHal::close() {
 }
 
 jboolean GnssHal::start() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        auto status = gnssHalAidl->start();
-        return checkAidlStatus(status, "IGnssAidl start() failed.");
-    }
-
     if (gnssHal == nullptr) {
         return JNI_FALSE;
     }
@@ -249,11 +215,6 @@ jboolean GnssHal::start() {
 }
 
 jboolean GnssHal::stop() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        auto status = gnssHalAidl->stop();
-        return checkAidlStatus(status, "IGnssAidl stop() failed.");
-    }
-
     if (gnssHal == nullptr) {
         return JNI_FALSE;
     }
@@ -264,10 +225,6 @@ jboolean GnssHal::stop() {
 
 jboolean GnssHal::startSvStatus() {
     isSvStatusRegistered = true;
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        auto status = gnssHalAidl->startSvStatus();
-        return checkAidlStatus(status, "IGnssAidl startSvStatus() failed.");
-    }
     if (gnssHal == nullptr) {
         return JNI_FALSE;
     }
@@ -276,10 +233,6 @@ jboolean GnssHal::startSvStatus() {
 
 jboolean GnssHal::stopSvStatus() {
     isSvStatusRegistered = false;
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        auto status = gnssHalAidl->stopSvStatus();
-        return checkAidlStatus(status, "IGnssAidl stopSvStatus() failed.");
-    }
     if (gnssHal == nullptr) {
         return JNI_FALSE;
     }
@@ -288,10 +241,6 @@ jboolean GnssHal::stopSvStatus() {
 
 jboolean GnssHal::startNmea() {
     isNmeaRegistered = true;
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        auto status = gnssHalAidl->startNmea();
-        return checkAidlStatus(status, "IGnssAidl startNmea() failed.");
-    }
     if (gnssHal == nullptr) {
         return JNI_FALSE;
     }
@@ -300,10 +249,6 @@ jboolean GnssHal::startNmea() {
 
 jboolean GnssHal::stopNmea() {
     isNmeaRegistered = false;
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        auto status = gnssHalAidl->stopNmea();
-        return checkAidlStatus(status, "IGnssAidl stopNmea() failed.");
-    }
     if (gnssHal == nullptr) {
         return JNI_FALSE;
     }
@@ -326,18 +271,6 @@ jint GnssHal::readNmea(jbyteArray& nmeaArray, jint& buffer_size) {
 jboolean GnssHal::setPositionMode(jint mode, jint recurrence, jint min_interval,
                                   jint preferred_accuracy, jint preferred_time,
                                   jboolean low_power_mode) {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        IGnssAidl::PositionModeOptions options;
-        options.mode = static_cast<IGnssAidl::GnssPositionMode>(mode);
-        options.recurrence = static_cast<IGnssAidl::GnssPositionRecurrence>(recurrence);
-        options.minIntervalMs = min_interval;
-        options.preferredAccuracyMeters = preferred_accuracy;
-        options.preferredTimeMs = preferred_time;
-        options.lowPowerMode = low_power_mode;
-        auto status = gnssHalAidl->setPositionMode(options);
-        return checkAidlStatus(status, "IGnssAidl setPositionMode() failed.");
-    }
-
     Return<bool> result = false;
     if (gnssHal_V1_1 != nullptr) {
         result = gnssHal_V1_1->setPositionMode_1_1(static_cast<IGnss_V1_0::GnssPositionMode>(mode),
@@ -355,12 +288,6 @@ jboolean GnssHal::setPositionMode(jint mode, jint recurrence, jint min_interval,
 }
 
 void GnssHal::deleteAidingData(jint flags) {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        auto status = gnssHalAidl->deleteAidingData(static_cast<IGnssAidl::GnssAidingData>(flags));
-        checkAidlStatus(status, "IGnssAidl deleteAidingData() failed.");
-        return;
-    }
-
     if (gnssHal == nullptr) {
         return;
     }
@@ -370,12 +297,6 @@ void GnssHal::deleteAidingData(jint flags) {
 }
 
 void GnssHal::injectTime(jlong time, jlong timeReference, jint uncertainty) {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        auto status = gnssHalAidl->injectTime(time, timeReference, uncertainty);
-        checkAidlStatus(status, "IGnssAidl injectTime() failed.");
-        return;
-    }
-
     if (gnssHal == nullptr) {
         return;
     }
@@ -390,19 +311,6 @@ void GnssHal::injectLocation(jint gnssLocationFlags, jdouble latitudeDegrees,
                              jfloat speedAccuracyMetersPerSecond, jfloat bearingAccuracyDegrees,
                              jlong timestamp, jint elapsedRealtimeFlags, jlong elapsedRealtimeNanos,
                              jdouble elapsedRealtimeUncertaintyNanos) {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        GnssLocationAidl location =
-                createGnssLocation(gnssLocationFlags, latitudeDegrees, longitudeDegrees,
-                                   altitudeMeters, speedMetersPerSec, bearingDegrees,
-                                   horizontalAccuracyMeters, verticalAccuracyMeters,
-                                   speedAccuracyMetersPerSecond, bearingAccuracyDegrees, timestamp,
-                                   elapsedRealtimeFlags, elapsedRealtimeNanos,
-                                   elapsedRealtimeUncertaintyNanos);
-        auto status = gnssHalAidl->injectLocation(location);
-        checkAidlStatus(status, "IGnssAidl injectLocation() failed.");
-        return;
-    }
-
     if (gnssHal == nullptr) {
         return;
     }
@@ -419,19 +327,6 @@ void GnssHal::injectBestLocation(jint gnssLocationFlags, jdouble latitudeDegrees
                                  jlong timestamp, jint elapsedRealtimeFlags,
                                  jlong elapsedRealtimeNanos,
                                  jdouble elapsedRealtimeUncertaintyNanos) {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        GnssLocationAidl location =
-                createGnssLocation(gnssLocationFlags, latitudeDegrees, longitudeDegrees,
-                                   altitudeMeters, speedMetersPerSec, bearingDegrees,
-                                   horizontalAccuracyMeters, verticalAccuracyMeters,
-                                   speedAccuracyMetersPerSecond, bearingAccuracyDegrees, timestamp,
-                                   elapsedRealtimeFlags, elapsedRealtimeNanos,
-                                   elapsedRealtimeUncertaintyNanos);
-        auto status = gnssHalAidl->injectBestLocation(location);
-        checkAidlStatus(status, "IGnssAidl injectBestLocation() failed.");
-        return;
-    }
-
     if (gnssHal_V2_0 != nullptr) {
         GnssLocation_V2_0 location =
                 createGnssLocation_V2_0(gnssLocationFlags, latitudeDegrees, longitudeDegrees,
@@ -461,13 +356,7 @@ void GnssHal::injectBestLocation(jint gnssLocationFlags, jdouble latitudeDegrees
 }
 
 std::unique_ptr<AGnssInterface> GnssHal::getAGnssInterface() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        sp<IAGnssAidl> agnssAidl;
-        auto status = gnssHalAidl->getExtensionAGnss(&agnssAidl);
-        if (checkAidlStatus(status, "Unable to get a handle to AGnss interface.")) {
-            return std::make_unique<gnss::AGnss>(agnssAidl);
-        }
-    } else if (gnssHal_V2_0 != nullptr) {
+    if (gnssHal_V2_0 != nullptr) {
         auto agnss_V2_0 = gnssHal_V2_0->getExtensionAGnss_2_0();
         if (checkHidlReturn(agnss_V2_0, "Unable to get a handle to AGnss_V2_0")) {
             return std::make_unique<gnss::AGnss_V2_0>(agnss_V2_0);
@@ -482,13 +371,7 @@ std::unique_ptr<AGnssInterface> GnssHal::getAGnssInterface() {
 }
 
 std::unique_ptr<AGnssRilInterface> GnssHal::getAGnssRilInterface() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        sp<IAGnssRilAidl> agnssRilAidl;
-        auto status = gnssHalAidl->getExtensionAGnssRil(&agnssRilAidl);
-        if (checkAidlStatus(status, "Unable to get a handle to AGnssRil interface.")) {
-            return std::make_unique<gnss::AGnssRil>(agnssRilAidl);
-        }
-    } else if (gnssHal_V2_0 != nullptr) {
+    if (gnssHal_V2_0 != nullptr) {
         auto agnssRil_V2_0 = gnssHal_V2_0->getExtensionAGnssRil_2_0();
         if (checkHidlReturn(agnssRil_V2_0, "Unable to get a handle to AGnssRil_V2_0")) {
             return std::make_unique<gnss::AGnssRil_V2_0>(agnssRil_V2_0);
@@ -503,14 +386,7 @@ std::unique_ptr<AGnssRilInterface> GnssHal::getAGnssRilInterface() {
 }
 
 std::unique_ptr<GnssNavigationMessageInterface> GnssHal::getGnssNavigationMessageInterface() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        sp<hardware::gnss::IGnssNavigationMessageInterface> gnssNavigationMessage;
-        auto status = gnssHalAidl->getExtensionGnssNavigationMessage(&gnssNavigationMessage);
-        if (checkAidlStatus(status,
-                            "Unable to get a handle to GnssNavigationMessage AIDL interface.")) {
-            return std::make_unique<gnss::GnssNavigationMessageAidl>(gnssNavigationMessage);
-        }
-    } else if (gnssHal != nullptr) {
+    if (gnssHal != nullptr) {
         auto gnssNavigationMessage = gnssHal->getExtensionGnssNavigationMessage();
         if (checkHidlReturn(gnssNavigationMessage,
                             "Unable to get a handle to GnssNavigationMessage interface.")) {
@@ -564,13 +440,6 @@ std::unique_ptr<GnssDebugInterface> GnssHal::getGnssDebugInterface() {
     // Allow all causal combinations between IGnss.hal and IGnssDebug.hal. That means,
     // 2.0@IGnss can be paired with {1.0, 2.0}@IGnssDebug
     // 1.0@IGnss is paired with 1.0@IGnssDebug
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        sp<IGnssDebugAidl> gnssDebugAidl;
-        auto status = gnssHalAidl->getExtensionGnssDebug(&gnssDebugAidl);
-        if (checkAidlStatus(status, "Unable to get a handle to GnssDebug interface.")) {
-            return std::make_unique<gnss::GnssDebug>(gnssDebugAidl);
-        }
-    }
     if (gnssHal_V2_0 != nullptr) {
         auto gnssDebug_V2_0 = gnssHal_V2_0->getExtensionGnssDebug_2_0();
         if (checkHidlReturn(gnssDebug_V2_0, "Unable to get a handle to GnssDebug_V2_0.")) {
@@ -623,13 +492,7 @@ std::unique_ptr<GnssConfigurationInterface> GnssHal::getGnssConfigurationInterfa
 }
 
 std::unique_ptr<GnssGeofenceInterface> GnssHal::getGnssGeofenceInterface() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        sp<hardware::gnss::IGnssGeofence> gnssGeofence;
-        auto status = gnssHalAidl->getExtensionGnssGeofence(&gnssGeofence);
-        if (checkAidlStatus(status, "Unable to get a handle to GnssGeofence AIDL interface.")) {
-            return std::make_unique<gnss::GnssGeofenceAidl>(gnssGeofence);
-        }
-    } else if (gnssHal != nullptr) {
+    if (gnssHal != nullptr) {
         auto gnssGeofencing = gnssHal->getExtensionGnssGeofencing();
         if (checkHidlReturn(gnssGeofencing, "Unable to get a handle to GnssGeofencing")) {
             return std::make_unique<gnss::GnssGeofenceHidl>(gnssGeofencing);
@@ -639,13 +502,6 @@ std::unique_ptr<GnssGeofenceInterface> GnssHal::getGnssGeofenceInterface() {
 }
 
 std::unique_ptr<GnssBatchingInterface> GnssHal::getGnssBatchingInterface() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        sp<android::hardware::gnss::IGnssBatching> gnssBatchingAidl;
-        auto status = gnssHalAidl->getExtensionGnssBatching(&gnssBatchingAidl);
-        if (checkAidlStatus(status, "Unable to get a handle to GnssBatching interface.")) {
-            return std::make_unique<gnss::GnssBatching>(gnssBatchingAidl);
-        }
-    }
     if (gnssHal_V2_0 != nullptr) {
         auto gnssBatching_V2_0 = gnssHal_V2_0->getExtensionGnssBatching_2_0();
         if (checkHidlReturn(gnssBatching_V2_0, "Unable to get a handle to GnssBatching_V2_0")) {
@@ -662,17 +518,6 @@ std::unique_ptr<GnssBatchingInterface> GnssHal::getGnssBatchingInterface() {
 }
 
 std::unique_ptr<MeasurementCorrectionsInterface> GnssHal::getMeasurementCorrectionsInterface() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        sp<android::hardware::gnss::measurement_corrections::IMeasurementCorrectionsInterface>
-                gnssMeasurementCorrectionsAidl;
-        auto status =
-                gnssHalAidl->getExtensionMeasurementCorrections(&gnssMeasurementCorrectionsAidl);
-        if (checkAidlStatus(status,
-                            "Unable to get a handle to GnssVisibilityControl AIDL interface.")) {
-            return std::make_unique<gnss::MeasurementCorrectionsIface_Aidl>(
-                    gnssMeasurementCorrectionsAidl);
-        }
-    }
     if (gnssHal_V2_1 != nullptr) {
         auto gnssCorrections = gnssHal_V2_1->getExtensionMeasurementCorrections_1_1();
         if (checkHidlReturn(gnssCorrections,
@@ -693,15 +538,7 @@ std::unique_ptr<MeasurementCorrectionsInterface> GnssHal::getMeasurementCorrecti
 }
 
 std::unique_ptr<GnssVisibilityControlInterface> GnssHal::getGnssVisibilityControlInterface() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        sp<android::hardware::gnss::visibility_control::IGnssVisibilityControl>
-                gnssVisibilityControlAidl;
-        auto status = gnssHalAidl->getExtensionGnssVisibilityControl(&gnssVisibilityControlAidl);
-        if (checkAidlStatus(status,
-                            "Unable to get a handle to GnssVisibilityControl AIDL interface.")) {
-            return std::make_unique<gnss::GnssVisibilityControlAidl>(gnssVisibilityControlAidl);
-        }
-    } else if (gnssHal_V2_0 != nullptr) {
+    if (gnssHal_V2_0 != nullptr) {
         auto gnssVisibilityControlHidl = gnssHal_V2_0->getExtensionVisibilityControl();
         if (checkHidlReturn(gnssVisibilityControlHidl,
                             "Unable to get a handle to GnssVisibilityControl HIDL interface")) {
@@ -712,13 +549,7 @@ std::unique_ptr<GnssVisibilityControlInterface> GnssHal::getGnssVisibilityContro
 }
 
 std::unique_ptr<GnssAntennaInfoInterface> GnssHal::getGnssAntennaInfoInterface() {
-    if (gnssHalAidl != nullptr && gnssHalAidl->getInterfaceVersion() >= 2) {
-        sp<IGnssAntennaInfoAidl> gnssAntennaInfoAidl;
-        auto status = gnssHalAidl->getExtensionGnssAntennaInfo(&gnssAntennaInfoAidl);
-        if (checkAidlStatus(status, "Unable to get a handle to GnssAntennaInfo interface.")) {
-            return std::make_unique<gnss::GnssAntennaInfoAidl>(gnssAntennaInfoAidl);
-        }
-    } else if (gnssHal_V2_1 != nullptr) {
+    if (gnssHal_V2_1 != nullptr) {
         auto gnssAntennaInfo_V2_1 = gnssHal_V2_1->getExtensionGnssAntennaInfo();
         if (checkHidlReturn(gnssAntennaInfo_V2_1,
                             "Unable to get a handle to GnssAntennaInfo_V2_1")) {
