diff --git a/media/libstagefright/Android.bp b/media/libstagefright/Android.bp
index d084f10590..30686eeaf0 100644
--- a/media/libstagefright/Android.bp
+++ b/media/libstagefright/Android.bp
@@ -202,6 +202,7 @@ cc_library_shared {
         "liblog",
         "libui",
         "libutils",
+        "libcutils",
     ],
 
     export_include_dirs: [
diff --git a/media/libstagefright/SurfaceUtils.cpp b/media/libstagefright/SurfaceUtils.cpp
index 74432a6c30..337ec51072 100644
--- a/media/libstagefright/SurfaceUtils.cpp
+++ b/media/libstagefright/SurfaceUtils.cpp
@@ -22,10 +22,12 @@
 #include <media/hardware/VideoAPI.h>
 #include <media/stagefright/SurfaceUtils.h>
 #include <gui/Surface.h>
+#include <cutils/properties.h>
 
 extern "C" int android_get_application_target_sdk_version();
 
 namespace android {
+char enableSoftDecode[PROPERTY_VALUE_MAX];
 
 status_t setNativeWindowSizeFormatAndUsage(
         ANativeWindow *nativeWindow /* nonnull */,
@@ -53,10 +55,19 @@ status_t setNativeWindowSizeFormatAndUsage(
         return err;
     }
 
-    err = native_window_set_buffers_format(nativeWindow, format);
-    if (err != NO_ERROR) {
-        ALOGE("native_window_set_buffers_format failed: %s (%d)", strerror(-err), -err);
-        return err;
+    if (property_get("ro.hardware.omxsoftdecode", enableSoftDecode, "0") && strcmp(enableSoftDecode, "1") != 0) {
+        // 未配置ro.hardware.omxsoftdecode属性，或者值为0时走va硬解
+        err = native_window_set_buffers_format(nativeWindow, HAL_PIXEL_FORMAT_YCBCR_420_888);
+        if (err != NO_ERROR) {
+            ALOGE("native_window_set_buffers_format failed: %s (%d)", strerror(-err), -err);
+            return err;
+        }
+    } else {
+        err = native_window_set_buffers_format(nativeWindow, format);
+        if (err != NO_ERROR) {
+            ALOGE("native_window_set_buffers_format failed: %s (%d)", strerror(-err), -err);
+            return err;
+        }
     }
 
     int transform = 0;
diff --git a/media/libstagefright/colorconversion/SoftwareRenderer.cpp b/media/libstagefright/colorconversion/SoftwareRenderer.cpp
index 240931591d..bb015b751c 100644
--- a/media/libstagefright/colorconversion/SoftwareRenderer.cpp
+++ b/media/libstagefright/colorconversion/SoftwareRenderer.cpp
@@ -130,6 +130,7 @@ void SoftwareRenderer::resetFormatIfChanged(
     // hardware has YUV12 and RGBA8888 support, so convert known formats
     {
         switch (mColorFormat) {
+#if 0 // kbox_hacked
             case OMX_COLOR_FormatYUV420Planar:
             case OMX_COLOR_FormatYUV420SemiPlanar:
             case OMX_TI_COLOR_FormatYUV420PackedSemiPlanar:
@@ -139,6 +140,7 @@ void SoftwareRenderer::resetFormatIfChanged(
                 bufHeight = (mCropHeight + 1) & ~1;
                 break;
             }
+#endif
             case OMX_COLOR_Format24bitRGB888:
             {
                 halFormat = HAL_PIXEL_FORMAT_RGB_888;
@@ -154,6 +156,7 @@ void SoftwareRenderer::resetFormatIfChanged(
                 bufHeight = (mCropHeight + 1) & ~1;
                 break;
             }
+#if 0 // kbox_hacked
             case OMX_COLOR_FormatYUV420Planar16:
             {
                 if (((dataSpace & HAL_DATASPACE_STANDARD_MASK) == HAL_DATASPACE_STANDARD_BT2020)
@@ -170,6 +173,7 @@ void SoftwareRenderer::resetFormatIfChanged(
                 bufHeight = (mCropHeight + 1) & ~1;
                 break;
             }
+#endif
             default:
             {
                 break;
diff --git a/media/libstagefright/omx/Android.bp b/media/libstagefright/omx/Android.bp
index 630817c7e6..23d1b1aa86 100644
--- a/media/libstagefright/omx/Android.bp
+++ b/media/libstagefright/omx/Android.bp
@@ -183,7 +183,6 @@ cc_defaults {
     name: "libstagefright_softomx-defaults",
     // TODO (b/316432618) Software OMX codecs are no longer used, disable building them till
     // this code is removed completely.
-    enabled: false,
     vendor_available: true,
 
     cflags: [
diff --git a/media/libstagefright/omx/OMXStore.cpp b/media/libstagefright/omx/OMXStore.cpp
index b2d5a70ffa..d00536623d 100644
--- a/media/libstagefright/omx/OMXStore.cpp
+++ b/media/libstagefright/omx/OMXStore.cpp
@@ -29,6 +29,7 @@
 #include <fcntl.h>
 
 #include <sstream>
+#include <cutils/properties.h>
 
 namespace android {
 
@@ -60,9 +61,14 @@ OMXStore::OMXStore() {
 OMXStore::~OMXStore() {
     clearPlugins();
 }
-
+char enableSoftDecode[PROPERTY_VALUE_MAX];
 void OMXStore::addVendorPlugin() {
-    addPlugin("libstagefrighthw.so");
+    if (property_get("ro.hardware.omxsoftdecode", enableSoftDecode, "0") && strcmp(enableSoftDecode, "1") != 0) {
+        // 未配置ro.hardware.omxsoftdecode属性，或者值为0时走va硬解
+        addPlugin("libstagefrighthw_va.so");
+    } else {
+        addPlugin("libstagefrighthw.so");
+    }
 }
 
 void OMXStore::addPlatformPlugin() {
@@ -161,7 +167,6 @@ void OMXStore::addPlugin(OMXPluginBase *plugin) {
                 }
             }
             if (skip) {
-                continue;
             }
         }
 
diff --git a/services/mediacodec/Android.bp b/services/mediacodec/Android.bp
index 506b3bca67..a2f17c22b9 100644
--- a/services/mediacodec/Android.bp
+++ b/services/mediacodec/Android.bp
@@ -95,6 +95,31 @@ cc_binary {
         "android.hidl.memory@1.0",
     ],
 
+    runtime_libs: [
+        "libstagefright_soft_aacdec",
+        "libstagefright_soft_aacenc",
+        "libstagefright_soft_amrdec",
+        "libstagefright_soft_amrnbenc",
+        "libstagefright_soft_amrwbenc",
+        "libstagefright_soft_avcdec",
+        "libstagefright_soft_avcenc",
+        "libstagefright_soft_flacdec",
+        "libstagefright_soft_flacenc",
+        "libstagefright_soft_g711dec",
+        "libstagefright_soft_gsmdec",
+        "libstagefright_soft_hevcdec",
+        "libstagefright_soft_mp3dec",
+        "libstagefright_soft_mpeg2dec",
+        "libstagefright_soft_mpeg4dec",
+        "libstagefright_soft_mpeg4enc",
+        "libstagefright_soft_opusdec",
+        "libstagefright_soft_rawdec",
+        "libstagefright_soft_vorbisdec",
+        "libstagefright_soft_vpxdec",
+        "libstagefright_soft_vpxenc",
+        "libstagefright_softomx_plugin",
+    ],
+
     // OMX interfaces force this to stay in 32-bit mode;
     compile_multilib: "32",
 
