diff --git a/src/amd/vulkan/radv_image.c b/src/amd/vulkan/radv_image.c
index 118ff61e091..e77fa7a895d 100644
--- a/src/amd/vulkan/radv_image.c
+++ b/src/amd/vulkan/radv_image.c
@@ -243,6 +243,11 @@ static bool
 radv_use_dcc_for_image_early(struct radv_device *device, struct radv_image *image, const VkImageCreateInfo *pCreateInfo,
                              VkFormat format, bool *sign_reinterpret)
 {
+   // 规避无限暖暖黑块问题，dcc特性对miplevel较大的纹理处理有问题，10是根据无限暖暖实测确定的阈值
+   if (pCreateInfo->mipLevels > 10) {
+      return false;
+   }
+
    const struct radv_physical_device *pdev = radv_device_physical(device);
    const struct radv_instance *instance = radv_physical_device_instance(pdev);
 
diff --git a/src/amd/vulkan/radv_physical_device.c b/src/amd/vulkan/radv_physical_device.c
index 23668eb3bf0..9aa83595c32 100644
--- a/src/amd/vulkan/radv_physical_device.c
+++ b/src/amd/vulkan/radv_physical_device.c
@@ -637,7 +637,7 @@ radv_physical_device_get_supported_extensions(const struct radv_physical_device
       .EXT_line_rasterization = true,
       .EXT_load_store_op_none = true,
       .EXT_map_memory_placed = true,
-      .EXT_memory_budget = true,
+      .EXT_memory_budget = false, // 解决安兔兔GPU跑分crash问题
       .EXT_memory_priority = true,
       .EXT_mesh_shader = radv_taskmesh_enabled(pdev),
       .EXT_multi_draw = true,
diff --git a/src/gallium/drivers/radeonsi/si_get.c b/src/gallium/drivers/radeonsi/si_get.c
index 39b1f022d2c..d78f497bb6d 100644
--- a/src/gallium/drivers/radeonsi/si_get.c
+++ b/src/gallium/drivers/radeonsi/si_get.c
@@ -698,7 +698,7 @@ static int si_get_video_param(struct pipe_screen *screen, enum pipe_video_profil
          if (codec != PIPE_VIDEO_FORMAT_UNKNOWN && QUERYABLE_KERNEL)
             return KERNEL_ENC_CAP(codec, max_height);
          else
-            return (sscreen->info.family < CHIP_TONGA) ? 1152 : 2304;
+            return (sscreen->info.family < CHIP_TONGA) ? 1152 : 4096; // 解决分辨率高大于2304时vaCreateContext失败问题
       case PIPE_VIDEO_CAP_PREFERED_FORMAT:
          if (profile == PIPE_VIDEO_PROFILE_HEVC_MAIN_10)
             return PIPE_FORMAT_P010;
diff --git a/src/mesa/main/mtypes.h b/src/mesa/main/mtypes.h
index 9924eeee613..75f62a369e0 100644
--- a/src/mesa/main/mtypes.h
+++ b/src/mesa/main/mtypes.h
@@ -3226,6 +3226,7 @@ struct gl_dispatch
  */
 struct gl_context
 {
+   bool FlushSwitch; // 解决优酷弹幕花屏问题
    /** State possibly shared with other contexts in the address space */
    struct gl_shared_state *Shared;
 
diff --git a/src/mesa/main/teximage.c b/src/mesa/main/teximage.c
index 6191c6fd67c..0388fbb2742 100644
--- a/src/mesa/main/teximage.c
+++ b/src/mesa/main/teximage.c
@@ -3359,6 +3359,9 @@ teximage_err(struct gl_context *ctx, GLboolean compressed, GLuint dims,
              GLint border, GLenum format, GLenum type,
              GLsizei imageSize, const GLvoid *pixels)
 {
+   if (ctx->FlushSwitch) {
+      _mesa_Flush();
+   }
    teximage(ctx, compressed, dims, NULL, target, level, internalFormat, width, height,
             depth, border, format, type, imageSize, pixels, false);
 }
@@ -3920,6 +3923,10 @@ texsubimage_err(struct gl_context *ctx, GLuint dims, GLenum target, GLint level,
    texture_sub_image(ctx, dims, texObj, texImage, target, level,
                      xoffset, yoffset, zoffset, width, height, depth,
                      format, type, pixels);
+   
+   if (ctx->FlushSwitch) {
+      _mesa_Flush();
+   }
 }
 
 
diff --git a/src/mesa/state_tracker/st_context.c b/src/mesa/state_tracker/st_context.c
index a13657b25d6..54b501979df 100644
--- a/src/mesa/state_tracker/st_context.c
+++ b/src/mesa/state_tracker/st_context.c
@@ -847,6 +847,7 @@ st_create_context(gl_api api, struct pipe_context *pipe,
       return NULL;
    }
 
+   ctx->FlushSwitch = shareCtx;
    st_debug_init();
 
    if (pipe->screen->get_disk_shader_cache)
