diff --git a/src/amd/vulkan/meson.build b/src/amd/vulkan/meson.build
index 5976bef8b85..3263857e3f6 100644
--- a/src/amd/vulkan/meson.build
+++ b/src/amd/vulkan/meson.build
@@ -35,6 +35,9 @@ radv_entrypoints = custom_target(
 )
 
 libradv_files = files(
+  '../../mesa/main/texcompress_etc.h',
+  '../../mesa/main/texcompress_etc.c',
+  '../../mesa/main/formats.c',
   'bvh/bvh.h',
   'layers/radv_ctx_roll_layer.c',
   'layers/radv_metro_exodus.c',
@@ -227,7 +230,7 @@ libvulkan_radeon = shared_library(
   [libradv_files, radv_entrypoints, sha1_h, radix_sort_spv, bvh_spv, radv_annotate_layer],
   vs_module_defs : vulkan_api_def,
   include_directories : [
-    inc_include, inc_src, inc_amd, inc_amd_common, inc_amd_common_llvm, inc_util,
+    inc_include, inc_src, inc_amd, inc_amd_common, inc_amd_common_llvm, inc_util, inc_mesa
   ],
   link_with : [
     libamd_common, libamd_common_llvm, libamdgpu_addrlib,
diff --git a/src/amd/vulkan/meta/radv_meta_blit.c b/src/amd/vulkan/meta/radv_meta_blit.c
index ffc84cb0b1e..e06931731a5 100644
--- a/src/amd/vulkan/meta/radv_meta_blit.c
+++ b/src/amd/vulkan/meta/radv_meta_blit.c
@@ -8,6 +8,7 @@
 #include "radv_meta.h"
 #include "vk_command_pool.h"
 #include "vk_common_entrypoints.h"
+#include "log/log.h"
 
 static VkResult create_pipeline(struct radv_device *device, VkImageAspectFlagBits aspect, enum glsl_sampler_dim tex_dim,
                                 VkFormat format, VkPipeline *pipeline);
@@ -181,6 +182,11 @@ get_pipeline(struct radv_device *device, const struct radv_image_view *src_iview
    switch (src_image->vk.aspects) {
    case VK_IMAGE_ASPECT_COLOR_BIT: {
       fs_key = radv_format_meta_fs_key(device, dst_image->vk.format);
+      if (fs_key >= NUM_META_FS_KEYS) {
+         ALOGE("Invalid fs key. if sys.vmi.vk.texturecompress is enabled, disable it and restart app.");
+         return VK_ERROR_FORMAT_NOT_SUPPORTED;
+      }
+
       format = radv_fs_key_format_exemplars[fs_key];
 
       switch (src_image->vk.image_type) {
diff --git a/src/amd/vulkan/meta/radv_meta_clear.c b/src/amd/vulkan/meta/radv_meta_clear.c
index 34c23a81d0d..a1f3205c2c7 100644
--- a/src/amd/vulkan/meta/radv_meta_clear.c
+++ b/src/amd/vulkan/meta/radv_meta_clear.c
@@ -16,6 +16,7 @@
 #include "vk_shader_module.h"
 
 #include "ac_formats.h"
+#include "log/log.h"
 
 enum { DEPTH_CLEAR_SLOW, DEPTH_CLEAR_FAST };
 
@@ -224,6 +225,11 @@ get_color_pipeline(struct radv_device *device, uint32_t samples, uint32_t frag_o
 {
    struct radv_meta_state *state = &device->meta_state;
    const uint32_t fs_key = radv_format_meta_fs_key(device, format);
+   // fs_key 最大值为 NUM_META_FS_KEYS - 1，数组索引越界时需要报告异常并返回。
+   if (fs_key >= NUM_META_FS_KEYS) {
+      ALOGE("Invalid fs key. if sys.vmi.vk.texturecompress is enabled, disable it and restart app.");
+      return VK_ERROR_FORMAT_NOT_SUPPORTED;
+   }
    const uint32_t samples_log2 = ffs(samples) - 1;
    VkResult result = VK_SUCCESS;
    VkPipeline *pipeline;
diff --git a/src/amd/vulkan/meta/radv_meta_copy.c b/src/amd/vulkan/meta/radv_meta_copy.c
index 79fa24e205e..7043214878b 100644
--- a/src/amd/vulkan/meta/radv_meta_copy.c
+++ b/src/amd/vulkan/meta/radv_meta_copy.c
@@ -9,6 +9,11 @@
 #include "radv_sdma.h"
 #include "vk_format.h"
 
+#include "main/formats.h"
+#include "main/texcompress_etc.h"
+#include "util/format/texcompress_bptc_tmp.h"
+#include "vk_common_entrypoints.h"
+
 static VkFormat
 vk_format_for_size(int bs)
 {
@@ -101,6 +106,152 @@ transfer_copy_buffer_image(struct radv_cmd_buffer *cmd_buffer, struct radv_buffe
    radv_sdma_copy_buffer_image(device, cs, &buf, &img, extent, to_image);
 }
 
+// 函数返回 image 所占的字节数
+static size_t
+radv_get_texel_count(const VkBufferImageCopy2* pRegions, struct radv_image* image)
+{
+   if (image == NULL) {
+      return 0;
+   }
+   const VkExtent3D bufferExtent = {
+      .width  = pRegions->bufferRowLength ? pRegions->bufferRowLength : pRegions->imageExtent.width,
+      .height = pRegions->bufferImageHeight ? pRegions->bufferImageHeight : pRegions->imageExtent.height,
+   };
+   const VkExtent3D buf_extent_el = vk_image_extent_to_elements(&image->vk, bufferExtent);
+   size_t w = buf_extent_el.width;
+   size_t h = buf_extent_el.height; 
+   int bs = vk_format_get_blocksize(image->vk.format);   
+   return w * h * bs;
+}
+
+static mesa_format
+radv_translate_etc_to_mesa_format(VkFormat format)
+{
+   switch (format) {
+   case VK_FORMAT_ETC2_R8G8B8_UNORM_BLOCK:
+      return MESA_FORMAT_ETC2_RGB8;
+   case VK_FORMAT_ETC2_R8G8B8_SRGB_BLOCK:
+      return MESA_FORMAT_ETC2_SRGB8;
+   case VK_FORMAT_ETC2_R8G8B8A1_UNORM_BLOCK:
+      return MESA_FORMAT_ETC2_RGB8_PUNCHTHROUGH_ALPHA1;
+   case VK_FORMAT_ETC2_R8G8B8A1_SRGB_BLOCK:
+      return MESA_FORMAT_ETC2_SRGB8_PUNCHTHROUGH_ALPHA1;
+   case VK_FORMAT_ETC2_R8G8B8A8_UNORM_BLOCK:
+      return MESA_FORMAT_ETC2_RGBA8_EAC;
+   case VK_FORMAT_ETC2_R8G8B8A8_SRGB_BLOCK:
+      return MESA_FORMAT_ETC2_SRGB8_ALPHA8_EAC;
+   default:
+      return MESA_FORMAT_ETC2_RGB8;
+   }
+}
+
+static mesa_format
+radv_translate_vkformat_to_mesa_format(VkFormat format)
+{
+   switch (format) {
+   case VK_FORMAT_R8G8B8_UNORM:
+      return MESA_FORMAT_RGB_UNORM8;
+   case VK_FORMAT_R8G8B8A8_UNORM:
+      return MESA_FORMAT_R8G8B8A8_UNORM;
+   case VK_FORMAT_R8G8B8_SRGB:
+      return MESA_FORMAT_BGR_SRGB8;
+   case VK_FORMAT_R8G8B8A8_SRGB:
+      return MESA_FORMAT_R8G8B8A8_SRGB;
+   case VK_FORMAT_BC7_UNORM_BLOCK:
+      return MESA_FORMAT_BPTC_RGBA_UNORM;
+   case VK_FORMAT_BC7_SRGB_BLOCK:
+      return MESA_FORMAT_BPTC_SRGB_ALPHA_UNORM;
+   default:
+      return MESA_FORMAT_RGB_UNORM8;
+   }
+}
+
+static void
+prepare_bufferMemory(struct radv_device* device, struct radv_buffer *buffer,
+                     struct radv_image *image, const VkBufferImageCopy2 *region,
+                     void** pDstData, VkBuffer* dstBuffer, VkDeviceMemory* memory_h, bool isDecodeBuffer) {
+      int dstBufferSize = radv_get_texel_count(region, image);
+
+      if (isDecodeBuffer && image->isNeedSoftEncode) {
+         VkFormat tempFormat = image->vk.format;
+         image->vk.format = image->unpackETCFormat;
+         dstBufferSize = radv_get_texel_count(region, image);
+         image->vk.format = tempFormat;
+      }
+      if(VK_SUCCESS != radv_CreateBuffer(radv_device_to_handle(device), &(VkBufferCreateInfo) {
+				.sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
+				.flags = buffer->vk.create_flags,
+				.size = dstBufferSize,
+				.usage = buffer->vk.usage,
+				.sharingMode = VK_SHARING_MODE_EXCLUSIVE
+            }, NULL, dstBuffer)) {
+			return;
+		}
+
+      VK_FROM_HANDLE(radv_buffer, _dstbuffer, *dstBuffer);
+      if (isDecodeBuffer) {
+         buffer->unpack_buffer_for_ct = _dstbuffer;
+      } else {
+         buffer->compressed_buffer_for_BC = _dstbuffer;
+      }
+
+		VkMemoryRequirements2 memoryRequirements2 = {
+            .sType = VK_STRUCTURE_TYPE_MEMORY_REQUIREMENTS_2,
+         };
+      vk_common_GetBufferMemoryRequirements2(radv_device_to_handle(device),
+         &(VkBufferMemoryRequirementsInfo2) {
+            .buffer = *dstBuffer,
+         }, &memoryRequirements2);
+		int memory_type_index = -1;
+      const struct radv_physical_device *pdev = radv_device_physical(device);
+		for (int i = 0; i < pdev->memory_properties.memoryTypeCount; ++i) {
+			bool is_local = !!(pdev->memory_properties.memoryTypes[i].propertyFlags & (VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT));
+			if (is_local) {
+				memory_type_index = i;
+				break;
+			}
+		}
+
+      // 分配设备内存，返回设备内存地址 memory_h
+		if(VK_SUCCESS != radv_AllocateMemory(radv_device_to_handle(device),
+			&(VkMemoryAllocateInfo) {
+				.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
+				.pNext = NULL,
+				.allocationSize = memoryRequirements2.memoryRequirements.size,
+				.memoryTypeIndex = memory_type_index,
+		  }, NULL, memory_h)) {
+			return;
+		}
+
+      VK_FROM_HANDLE(radv_device_memory, _mem, *memory_h);
+      if (isDecodeBuffer) {
+         buffer->unpack_mem_for_ct = _mem;
+      } else {
+         buffer->compressed_mem_for_BC = _mem;
+      }
+
+      // 将 buffer 绑定到设备内存
+		if(VK_SUCCESS != radv_BindBufferMemory2(radv_device_to_handle(device), 1,
+         &(VkBindBufferMemoryInfo) {
+            .sType = VK_STRUCTURE_TYPE_BIND_BUFFER_MEMORY_INFO,
+            .buffer = *dstBuffer,
+            .memory = *memory_h,
+            .memoryOffset = 0,
+         })) {
+			return;
+		}
+      VkMemoryMapInfoKHR pMemoryMapInfo = {
+         .sType = VK_STRUCTURE_TYPE_MEMORY_MAP_INFO_KHR,
+         .pNext = NULL,
+         .flags = 0,
+         .memory = *memory_h,
+         .offset = 0,
+         .size = dstBufferSize,
+      };
+		// 将设备内存 memory_h 映射到 pDstData
+		radv_MapMemory2KHR(radv_device_to_handle(device), &pMemoryMapInfo, pDstData);
+}
+
 static void
 copy_buffer_to_image(struct radv_cmd_buffer *cmd_buffer, struct radv_buffer *buffer, struct radv_image *image,
                      VkImageLayout layout, const VkBufferImageCopy2 *region)
@@ -111,6 +262,17 @@ copy_buffer_to_image(struct radv_cmd_buffer *cmd_buffer, struct radv_buffer *buf
    }
 
    struct radv_device *device = radv_cmd_buffer_device(cmd_buffer);
+
+   VkBuffer dstBuffer;
+   struct radv_buffer* radvDstBuffer = buffer;
+   VkDeviceMemory memory_h;
+   void* pDstData = NULL;
+   void* pSrcData = NULL;
+
+   VkBuffer dstBCnBuffer;
+   VkDeviceMemory memory_h_BCn;
+   void* pDstBCnData = NULL;
+
    struct radv_meta_saved_state saved_state;
    bool cs;
 
@@ -128,6 +290,19 @@ copy_buffer_to_image(struct radv_cmd_buffer *cmd_buffer, struct radv_buffer *buf
                   (cs ? RADV_META_SAVE_COMPUTE_PIPELINE : RADV_META_SAVE_GRAPHICS_PIPELINE) | RADV_META_SAVE_CONSTANTS |
                      RADV_META_SAVE_DESCRIPTORS | RADV_META_SUSPEND_PREDICATING);
 
+   pSrcData = radv_buffer_map(device->ws, buffer->bo) + buffer->offset;
+   // 申请软解 buffer 内存
+   if (image->isNeedSoftDecode) {
+      prepare_bufferMemory(device, buffer, image, region, &pDstData, &dstBuffer, &memory_h, true);
+   }
+   // 申请软编 buffer 内存
+   if (image->isNeedSoftEncode) {
+      prepare_bufferMemory(device, buffer, image, region, &pDstBCnData, &dstBCnBuffer, &memory_h_BCn, false);
+   }
+
+   void* realDstData = pDstData;
+   void* realDstBcData = pDstBCnData;
+
    /**
     * From the Vulkan 1.0.6 spec: 18.3 Copying Data Between Images
     *    extent is the size in texels of the source image to copy in width,
@@ -173,12 +348,58 @@ copy_buffer_to_image(struct radv_cmd_buffer *cmd_buffer, struct radv_buffer *buf
       img_bsurf.format = vk_format_for_size(vk_format_get_blocksize(img_bsurf.format));
    }
 
+   const VkExtent3D bufferExtent = {
+      .width  = region->bufferRowLength ? region->bufferRowLength : region->imageExtent.width,
+      .height = region->bufferImageHeight ? region->bufferImageHeight : region->imageExtent.height,
+   };
+   const VkExtent3D buf_extent_el = vk_image_extent_to_elements(&image->vk, bufferExtent);
+
+   uint32_t offset = region->bufferOffset;
+   mesa_format unpackETCFormat;
+   if (image->isNeedSoftDecode) {
+      // translate vkFormat to MESA format
+      unpackETCFormat = radv_translate_vkformat_to_mesa_format(image->unpackETCFormat);
+      mesa_format srcFormat = radv_translate_etc_to_mesa_format(image->srcFormat);
+      bool bgra = false;
+      void * realSrcData = pSrcData;
+      realSrcData += region->bufferOffset;
+      unsigned dst_stride = _mesa_format_row_stride(unpackETCFormat, region->imageExtent.width);
+      unsigned src_stride =_mesa_format_row_stride(srcFormat, bufferExtent.width);
+      _mesa_unpack_etc2_format(realDstData, dst_stride, realSrcData, src_stride, bufferExtent.width, bufferExtent.height, srcFormat, bgra);
+      if (image->isNeedSoftEncode) {
+         VK_FROM_HANDLE(radv_buffer, _tmp, dstBuffer);
+         radvDstBuffer = _tmp;
+         offset = realDstData - pDstData;
+      }
+   }
+   if (image->isNeedSoftEncode) {
+      mesa_format srcFormat;
+      void * realSrcData;
+      if (image->isNeedSoftDecode) {
+         srcFormat = unpackETCFormat;
+         realSrcData = realDstData;
+      } else {
+         srcFormat = radv_translate_vkformat_to_mesa_format(image->srcFormat);
+         realSrcData = pSrcData;
+         realSrcData += region->bufferOffset;
+      }
+      mesa_format dstFormat = radv_translate_vkformat_to_mesa_format(image->vk.format);
+      unsigned dst_stride = _mesa_format_row_stride(dstFormat, region->imageExtent.width);
+      unsigned src_stride = _mesa_format_row_stride(srcFormat, bufferExtent.width);
+
+      compress_rgba_unorm(bufferExtent.width, bufferExtent.height, realSrcData, src_stride, realDstBcData, dst_stride);
+
+      VK_FROM_HANDLE(radv_buffer, _tmp, dstBCnBuffer);
+      radvDstBuffer = _tmp;
+      offset = realDstBcData - pDstBCnData;
+   }
+
    const struct vk_image_buffer_layout buf_layout = vk_image_buffer_copy_layout(&image->vk, region);
    struct radv_meta_blit2d_buffer buf_bsurf = {
       .bs = img_bsurf.bs,
       .format = img_bsurf.format,
-      .buffer = buffer,
-      .offset = region->bufferOffset,
+      .buffer = radvDstBuffer,
+      .offset = offset,
       .pitch = buf_layout.row_stride_B / buf_layout.element_size_B,
    };
 
@@ -207,12 +428,40 @@ copy_buffer_to_image(struct radv_cmd_buffer *cmd_buffer, struct radv_buffer *buf
        * re-binding it to different backing memory.
        */
       buf_bsurf.offset += buf_layout.image_stride_B;
+      if (image->isNeedSoftDecode) {
+         realDstData += buf_extent_el.width * buf_extent_el.height * buf_bsurf.bs;
+      }
+      if (image->isNeedSoftEncode) {
+         realDstBcData += buf_extent_el.width * buf_extent_el.height * buf_bsurf.bs;
+      }
       img_bsurf.layer++;
       if (image->vk.image_type == VK_IMAGE_TYPE_3D)
          slice_3d++;
       else
          slice_array++;
    }
+   pSrcData = NULL;
+   if (image->isNeedSoftDecode || image->isNeedSoftEncode) {
+      device->ws->buffer_unmap(device->ws, buffer->bo, false);
+   }
+   VkMemoryUnmapInfoKHR pMemoryUnmapInfo = {
+      .sType = VK_STRUCTURE_TYPE_MEMORY_UNMAP_INFO_KHR,
+      .pNext = NULL,
+      .flags = 0,
+   };
+   if (image->isNeedSoftDecode) {
+      pMemoryUnmapInfo.memory = memory_h;
+      radv_UnmapMemory2KHR(radv_device_to_handle(device), &pMemoryUnmapInfo);
+   }
+   if (image->isNeedSoftEncode) {
+      pMemoryUnmapInfo.memory = memory_h_BCn;
+      radv_UnmapMemory2KHR(radv_device_to_handle(device), &pMemoryUnmapInfo);
+      // 压缩纹理的原始纹理由ETC2纹理解压而来因此需要释放掉内存
+      if (buffer->unpack_mem_for_ct) {
+         radv_free_memory(device, NULL, buffer->unpack_mem_for_ct);
+         buffer->unpack_mem_for_ct = NULL;
+      }
+   }
 
    radv_meta_restore(&saved_state, cmd_buffer);
 }
diff --git a/src/amd/vulkan/radv_buffer.c b/src/amd/vulkan/radv_buffer.c
index 4207d20115e..e9602be8868 100644
--- a/src/amd/vulkan/radv_buffer.c
+++ b/src/amd/vulkan/radv_buffer.c
@@ -86,6 +86,12 @@ radv_create_buffer(struct radv_device *device, const VkBufferCreateInfo *pCreate
    buffer->bo_va = 0;
    buffer->range = 0;
 
+   buffer->unpack_buffer_for_ct = NULL;
+   buffer->unpack_mem_for_ct = NULL;
+
+   buffer->compressed_buffer_for_BC = NULL;
+   buffer->compressed_mem_for_BC = NULL;
+
    uint64_t replay_address = 0;
    const VkBufferOpaqueCaptureAddressCreateInfo *replay_info =
       vk_find_struct_const(pCreateInfo->pNext, BUFFER_OPAQUE_CAPTURE_ADDRESS_CREATE_INFO);
@@ -137,6 +143,26 @@ radv_DestroyBuffer(VkDevice _device, VkBuffer _buffer, const VkAllocationCallbac
    if (!buffer)
       return;
 
+   if (buffer->unpack_mem_for_ct) {
+      radv_free_memory(device, NULL, buffer->unpack_mem_for_ct);
+      buffer->unpack_mem_for_ct = NULL;
+   }
+   
+   if (buffer->unpack_buffer_for_ct) {
+      radv_destroy_buffer(device, NULL, buffer->unpack_buffer_for_ct);
+      buffer->unpack_buffer_for_ct = NULL;
+   }
+   
+   if (buffer->compressed_mem_for_BC) {
+      radv_free_memory(device, NULL, buffer->compressed_mem_for_BC);
+      buffer->compressed_mem_for_BC = NULL;
+   }
+   
+   if (buffer->compressed_buffer_for_BC) {
+      radv_destroy_buffer(device, NULL, buffer->compressed_buffer_for_BC);
+      buffer->compressed_buffer_for_BC = NULL;
+   }
+
    radv_destroy_buffer(device, pAllocator, buffer);
 }
 
diff --git a/src/amd/vulkan/radv_buffer.h b/src/amd/vulkan/radv_buffer.h
index dd21c62ff09..5d226d0601c 100644
--- a/src/amd/vulkan/radv_buffer.h
+++ b/src/amd/vulkan/radv_buffer.h
@@ -23,6 +23,14 @@ struct radv_buffer {
    /* Set when bound */
    struct radeon_winsys_bo *bo;
    VkDeviceSize offset;
+
+   // 存储解压的buffer
+   struct radv_buffer* unpack_buffer_for_ct;
+   struct radv_device_memory* unpack_mem_for_ct;
+   // 存储软压缩的buffer
+   struct radv_buffer* compressed_buffer_for_BC;
+   struct radv_device_memory* compressed_mem_for_BC;
+
    uint64_t bo_va;
    uint64_t range;
 };
diff --git a/src/amd/vulkan/radv_image.c b/src/amd/vulkan/radv_image.c
index 9e64c76c412..c193132570f 100644
--- a/src/amd/vulkan/radv_image.c
+++ b/src/amd/vulkan/radv_image.c
@@ -32,6 +32,8 @@
 #include "vk_render_pass.h"
 #include "vk_util.h"
 
+#include <cutils/properties.h>
+
 #include "gfx10_format_table.h"
 
 static unsigned
@@ -1319,6 +1321,89 @@ radv_select_modifier(const struct radv_device *dev, VkFormat format,
    unreachable("App specified an invalid modifier");
 }
 
+static char isEnableSoftEncode[PROPERTY_VALUE_MAX];
+static bool
+radv_need_soft_decode(VkFormat format) {
+   if (property_get("sys.vmi.vk.texturecompress", isEnableSoftEncode, "0") && strcmp(isEnableSoftEncode, "1") != 0) {
+      return false;
+   }
+   switch (format) {
+   case VK_FORMAT_ETC2_R8G8B8_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8_SRGB_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A1_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A1_SRGB_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A8_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A8_SRGB_BLOCK:
+      return true;
+   default:
+      return false;
+   }
+}
+
+#include "log/log.h"
+bool
+radv_need_soft_encode(const VkImageCreateInfo *pCreateInfo) {
+   if (property_get("sys.vmi.vk.texturecompress", isEnableSoftEncode, "0") && strcmp(isEnableSoftEncode, "1") != 0) {
+      return false;
+   }
+   // 过滤掉对稀疏纹理的压缩，对该部分纹理进行压缩时可能造成GPU reset。 
+   if (pCreateInfo->flags & (VK_IMAGE_CREATE_SPARSE_BINDING_BIT | VK_IMAGE_CREATE_SPARSE_RESIDENCY_BIT |
+      VK_IMAGE_CREATE_SPARSE_ALIASED_BIT)) {
+      return false;
+   }
+   // 首先检查图像的用途，防止压缩 swapchain 的 images；并只对 VK_IMAGE_TYPE_2D 类型纹理进行压缩
+   if (!(pCreateInfo->usage & VK_IMAGE_USAGE_SAMPLED_BIT) || pCreateInfo->imageType != VK_IMAGE_TYPE_2D) {
+   	return false;
+   }  
+   // 对指定格式纹理进行压缩
+   switch (pCreateInfo->format) {
+   case VK_FORMAT_R8G8B8A8_SRGB:
+   case VK_FORMAT_R8G8B8A8_UNORM:
+   {
+      const VkImageAspectFlags aspects = vk_format_aspects(pCreateInfo->format);
+      if (aspects & VK_IMAGE_ASPECT_COLOR_BIT) {
+         // 如果是颜色附件，不让他压缩成BC格式，否则会出现找不到管线进行clear操作
+         return false;
+      }
+   }
+   case VK_FORMAT_ETC2_R8G8B8_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8_SRGB_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A1_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A1_SRGB_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A8_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A8_SRGB_BLOCK:
+      break;
+   default:
+      return false;
+   }  
+   // 检查图像大小
+   if (pCreateInfo->extent.width >= 256 && pCreateInfo->extent.height >= 256) {
+      return true;
+   }
+   return false;
+}
+
+VkFormat
+radv_translate_rgb2BC(VkFormat format)
+{
+   switch (format) {
+   case VK_FORMAT_R8G8B8_SRGB:
+   case VK_FORMAT_R8G8B8A8_SRGB:
+   case VK_FORMAT_ETC2_R8G8B8_SRGB_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A1_SRGB_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A8_SRGB_BLOCK:
+      return VK_FORMAT_BC7_SRGB_BLOCK;
+   case VK_FORMAT_R8G8B8_UNORM:
+   case VK_FORMAT_R8G8B8A8_UNORM:
+   case VK_FORMAT_ETC2_R8G8B8_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A1_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A8_UNORM_BLOCK:
+      return VK_FORMAT_BC7_UNORM_BLOCK;
+   default:
+      return format;
+   }
+}
+
 VkResult
 radv_image_create(VkDevice _device, const struct radv_image_create_info *create_info,
                   const VkAllocationCallbacks *alloc, VkImage *pImage, bool is_internal)
@@ -1340,6 +1425,17 @@ radv_image_create(VkDevice _device, const struct radv_image_create_info *create_
 
    unsigned plane_count = radv_get_internal_plane_count(pdev, format);
 
+   const VkExternalMemoryImageCreateInfo *external_info =
+      vk_find_struct_const(pCreateInfo->pNext, EXTERNAL_MEMORY_IMAGE_CREATE_INFO);
+   bool needSoftEncode = !external_info && radv_need_soft_encode(pCreateInfo);
+   bool needSoftDecode = needSoftEncode && radv_need_soft_decode(format);
+   VkFormat softDecodeFormat = radv_translate_etc(format);
+   VkFormat softEncodeFormat = radv_translate_rgb2BC(format);
+   // plane_count 影响纹理数据拷贝的过程，需要根据最终格式进行判断
+   if (needSoftEncode) {
+      plane_count = vk_format_get_plane_count(softEncodeFormat);
+   }
+
    const size_t image_struct_size = sizeof(*image) + sizeof(struct radv_image_plane) * plane_count;
 
    image = vk_zalloc2(&device->vk.alloc, alloc, image_struct_size, 8, VK_SYSTEM_ALLOCATION_SCOPE_OBJECT);
@@ -1348,9 +1444,24 @@ radv_image_create(VkDevice _device, const struct radv_image_create_info *create_
 
    vk_image_init(&device->vk, &image->vk, pCreateInfo);
 
+   image->isNeedSoftDecode = needSoftDecode;
+   image->isNeedSoftEncode = needSoftEncode;
+   image->srcFormat = format;
+   image->unpackETCFormat = format;
+
    image->plane_count = vk_format_get_plane_count(format);
    image->disjoint = image->plane_count > 1 && pCreateInfo->flags & VK_IMAGE_CREATE_DISJOINT_BIT;
 
+   //满足要求的 ETC2 纹理使用软解
+   if (needSoftDecode) {
+      image->unpackETCFormat = softDecodeFormat;
+      image->vk.format = softDecodeFormat;
+   }
+   // 软解的ETC2、原始RGB/A纹理使用软编
+   if (needSoftEncode) {
+      image->vk.format = softEncodeFormat;
+   }
+
    image->exclusive = pCreateInfo->sharingMode == VK_SHARING_MODE_EXCLUSIVE;
    if (pCreateInfo->sharingMode == VK_SHARING_MODE_CONCURRENT) {
       for (uint32_t i = 0; i < pCreateInfo->queueFamilyIndexCount; ++i)
@@ -1364,9 +1475,6 @@ radv_image_create(VkDevice _device, const struct radv_image_create_info *create_
       image->queue_family_mask &= ~(1u << RADV_QUEUE_SPARSE);
    }
 
-   const VkExternalMemoryImageCreateInfo *external_info =
-      vk_find_struct_const(pCreateInfo->pNext, EXTERNAL_MEMORY_IMAGE_CREATE_INFO);
-
    image->shareable = external_info;
 
    if (mod_list)
diff --git a/src/amd/vulkan/radv_image.h b/src/amd/vulkan/radv_image.h
index 5840a87f07b..f80ba8c64c1 100644
--- a/src/amd/vulkan/radv_image.h
+++ b/src/amd/vulkan/radv_image.h
@@ -50,6 +50,11 @@ struct radv_image {
    bool dcc_sign_reinterpret;
    bool support_comp_to_single;
 
+   bool isNeedSoftEncode;// 是否需要使用 软编
+   bool isNeedSoftDecode;// 是否需要使用 软解
+	VkFormat srcFormat; //用于保存纹理的原始格式
+	VkFormat unpackETCFormat;//用于保存ETC纹理解压后的格式
+
    struct radv_image_binding bindings[3];
    bool tc_compatible_cmask;
 
@@ -308,6 +313,23 @@ radv_image_get_iterate256(const struct radv_device *device, struct radv_image *i
    return pdev->info.gfx_level >= GFX10 && radv_image_is_tc_compat_htile(image) && image->vk.samples > 1;
 }
 
+static VkFormat
+radv_translate_etc(VkFormat format)
+{
+   switch (format) {
+   case VK_FORMAT_ETC2_R8G8B8_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A1_UNORM_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A8_UNORM_BLOCK:
+      return VK_FORMAT_R8G8B8A8_UNORM;
+   case VK_FORMAT_ETC2_R8G8B8_SRGB_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A1_SRGB_BLOCK:
+   case VK_FORMAT_ETC2_R8G8B8A8_SRGB_BLOCK:
+      return  VK_FORMAT_R8G8B8A8_SRGB;
+   default:
+      return format;
+   }
+}
+
 bool radv_are_formats_dcc_compatible(const struct radv_physical_device *pdev, const void *pNext, VkFormat format,
                                      VkImageCreateFlags flags, bool *sign_reinterpret);
 
diff --git a/src/amd/vulkan/radv_image_view.c b/src/amd/vulkan/radv_image_view.c
index 5b1b4349ab0..8b6b5f94a10 100644
--- a/src/amd/vulkan/radv_image_view.c
+++ b/src/amd/vulkan/radv_image_view.c
@@ -533,6 +533,10 @@ radv_image_view_init(struct radv_image_view *iview, struct radv_device *device,
       plane_count = vk_format_get_plane_count(iview->vk.format);
    }
 
+   if (image->isNeedSoftDecode) {
+      iview->vk.format = radv_translate_etc(iview->vk.format);
+   }
+
    /* when the view format is emulated, redirect the view to the hidden plane 1 */
    if (radv_is_format_emulated(pdev, iview->vk.format)) {
       assert(radv_is_format_emulated(pdev, image->vk.format));
@@ -662,10 +666,24 @@ radv_CreateImageView(VkDevice _device, const VkImageViewCreateInfo *pCreateInfo,
    if (view == NULL)
       return vk_error(device, VK_ERROR_OUT_OF_HOST_MEMORY);
 
-   radv_image_view_init(view, device, pCreateInfo, &(struct radv_image_view_extra_create_info){.from_client = true});
+   // 纹理为 RGB 纹理时，需要修改 view 的 format
+   VkImageViewCreateInfo *pCreateInfoTemp;
+   pCreateInfoTemp =
+      vk_alloc2(&device->vk.alloc, pAllocator, sizeof(*pCreateInfo), 8, VK_SYSTEM_ALLOCATION_SCOPE_OBJECT);
+   if (pCreateInfoTemp == NULL)
+      return vk_error(device, VK_ERROR_OUT_OF_HOST_MEMORY);
+
+   memcpy(pCreateInfoTemp, pCreateInfo, sizeof(VkImageViewCreateInfo));
+   VK_FROM_HANDLE(radv_image, image, pCreateInfoTemp->image);
+   if (image->isNeedSoftEncode) {
+      pCreateInfoTemp->format = image->vk.format;
+   }
+   radv_image_view_init(view, device, pCreateInfoTemp, &(struct radv_image_view_extra_create_info){.from_client = true});
 
    *pView = radv_image_view_to_handle(view);
 
+   vk_free2(&device->vk.alloc, pAllocator, pCreateInfoTemp);
+
    return VK_SUCCESS;
 }
 
diff --git a/src/mesa/state_tracker/st_context.c b/src/mesa/state_tracker/st_context.c
index 54b501979df..5f9810959b0 100644
--- a/src/mesa/state_tracker/st_context.c
+++ b/src/mesa/state_tracker/st_context.c
@@ -69,6 +69,7 @@
 #include "util/thread_sched.h"
 #include "cso_cache/cso_context.h"
 #include "compiler/glsl/glsl_parser_extras.h"
+#include <cutils/properties.h>
 
 DEBUG_GET_ONCE_BOOL_OPTION(mesa_mvp_dp4, "MESA_MVP_DP4", false)
 
@@ -429,6 +430,7 @@ st_have_perfquery(struct st_context *ctx)
           pipe->get_intel_perf_query_data;
 }
 
+static char isEnableHardEncode[PROPERTY_VALUE_MAX];
 static struct st_context *
 st_create_context_priv(struct gl_context *ctx, struct pipe_context *pipe,
                        const struct st_config_options *options)
@@ -535,6 +537,8 @@ st_create_context_priv(struct gl_context *ctx, struct pipe_context *pipe,
                                                    PIPE_TEXTURE_2D, 0, 0,
                                                    PIPE_BIND_SAMPLER_VIEW);
    st->transcode_astc = options->transcode_astc &&
+                        property_get("sys.vmi.gl.texturecompress", isEnableHardEncode, "0") &&
+                        strcmp(isEnableHardEncode, "1") == 0 &&
                         screen->is_format_supported(screen, PIPE_FORMAT_DXT5_SRGBA,
                                                     PIPE_TEXTURE_2D, 0, 0,
                                                     PIPE_BIND_SAMPLER_VIEW) &&
