diff --git a/libs/binder/Binder.cpp b/libs/binder/Binder.cpp
index 53bd08d420..3516dcde3a 100644
--- a/libs/binder/Binder.cpp
+++ b/libs/binder/Binder.cpp
@@ -522,6 +522,7 @@ void BBinder::setRequestingSid(bool requestingSid)
                         "setRequestingSid() should not be called after a binder object "
                         "is parceled/sent to another process");
 
+    requestingSid = false; // kbox_hacked
     Extras* e = mExtras.load(std::memory_order_acquire);
 
     if (!e) {
diff --git a/libs/binder/IPCThreadState.cpp b/libs/binder/IPCThreadState.cpp
index 6698d0c0cd..6707c95c0a 100644
--- a/libs/binder/IPCThreadState.cpp
+++ b/libs/binder/IPCThreadState.cpp
@@ -1441,7 +1441,7 @@ status_t IPCThreadState::executeCommand(int32_t cmd)
             clearPropagateWorkSource();
 
             mCallingPid = tr.sender_pid;
-            mCallingSid = reinterpret_cast<const char*>(tr_secctx.secctx);
+            mCallingSid = "KBOX_HACKED";
             mCallingUid = tr.sender_euid;
             mHasExplicitIdentity = false;
             mLastTransactionBinderFlags = tr.flags;
diff --git a/libs/binder/ProcessState.cpp b/libs/binder/ProcessState.cpp
index 5e7f1510bc..83bae2c75f 100644
--- a/libs/binder/ProcessState.cpp
+++ b/libs/binder/ProcessState.cpp
@@ -225,12 +225,15 @@ bool ProcessState::becomeContextManager()
 {
     std::unique_lock<std::mutex> _l(mLock);
 
+#if 0
     flat_binder_object obj {
         .flags = FLAT_BINDER_FLAG_TXN_SECURITY_CTX,
     };
 
     int result = ioctl(mDriverFD, BINDER_SET_CONTEXT_MGR_EXT, &obj);
+#endif
 
+    int result = 1;
     // fallback to original method
     if (result != 0) {
         android_errorWriteLog(0x534e4554, "121035042");
diff --git a/libs/ui/GraphicBufferAllocator.cpp b/libs/ui/GraphicBufferAllocator.cpp
index 1ebe5973fa..375019124c 100644
--- a/libs/ui/GraphicBufferAllocator.cpp
+++ b/libs/ui/GraphicBufferAllocator.cpp
@@ -143,8 +143,9 @@ auto GraphicBufferAllocator::allocate(const AllocationRequest& request) -> Alloc
             return result;
         }
         // If there's no additional options, fall back to previous allocate version
+        const uint64_t usage = (request.usage & ~static_cast<uint64_t>((1 << 10) | (1 << 13)));
         result.status = mAllocator->allocate(request.requestorName, request.width, request.height,
-                                             request.format, request.layerCount, request.usage,
+                                             request.format, request.layerCount, usage,
                                              &result.stride, &result.handle, request.importBuffer);
     }
 
diff --git a/services/surfaceflinger/Scheduler/RefreshRateSelector.cpp b/services/surfaceflinger/Scheduler/RefreshRateSelector.cpp
index ab9014e418..8bce49cb41 100644
--- a/services/surfaceflinger/Scheduler/RefreshRateSelector.cpp
+++ b/services/surfaceflinger/Scheduler/RefreshRateSelector.cpp
@@ -1427,6 +1427,8 @@ auto RefreshRateSelector::setPolicy(const PolicyVariant& policy) -> SetPolicyRes
             return SetPolicyResult::Unchanged;
         }
 
+        mDisplayManagerPolicy.primaryRanges.render.max = Fps::fromValue(120.f);
+
         constructAvailableRefreshRates();
 
         displayId = getActiveModeLocked().modePtr->getPhysicalDisplayId();
diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index 65a0ed3065..0fd9262b3b 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -2924,7 +2924,8 @@ CompositeResultsPerDisplay SurfaceFlinger::composite(
     // TODO(b/160583065): Enable skip validation when SF caches all client composition layers.
     const bool hasGpuUseOrReuse =
             coverage.any(CompositionCoverage::Gpu | CompositionCoverage::GpuReuse);
-    mScheduler->modulateVsync({}, &VsyncModulator::onDisplayRefresh, hasGpuUseOrReuse);
+    // fix bug: 解决配置60fps实际帧率只有54、55问题。此处会引入额外的offset，导致vsync周期偏大
+    // mScheduler->modulateVsync({}, &VsyncModulator::onDisplayRefresh, hasGpuUseOrReuse);
 
     mLayersWithQueuedFrames.clear();
     mLayersIdsWithQueuedFrames.clear();
diff --git a/vulkan/libvulkan/driver.cpp b/vulkan/libvulkan/driver.cpp
index 01436db4ae..3ea9b59f3c 100644
--- a/vulkan/libvulkan/driver.cpp
+++ b/vulkan/libvulkan/driver.cpp
@@ -165,6 +165,23 @@ const std::array<const char*, 2> HAL_SUBNAME_KEY_PROPERTIES = {{
 constexpr int LIB_DL_FLAGS = RTLD_LOCAL | RTLD_NOW;
 constexpr char RO_VULKAN_APEX_PROPERTY[] = "ro.vulkan.apex";
 
+const std::array<const char*, 1> DISABLE_VULKAN_FOR_APPLICATION = {
+    "com.miHoYo.hkrpg"
+};
+
+static void get_proc_name(char* applicationName, int fReadSize)
+{
+    pid_t pid = getpid();
+    char path[256] = { 0 };
+    sprintf(path, "/proc/%d/cmdline", pid);
+    FILE* cmdline = fopen(path, "r");
+
+    if(cmdline) {
+        fread(applicationName, fReadSize, 1, cmdline);
+        fclose(cmdline);
+    }
+}
+
 // LoadDriver returns:
 // * 0 when succeed, or
 // * -ENOENT when fail to open binary libraries, or
@@ -174,6 +191,16 @@ int LoadDriver(android_namespace_t* library_namespace,
                const char* ns_name,
                const hwvulkan_module_t** module) {
     ATRACE_CALL();
+    // 对名单中的应用禁用vulkan
+    const uint32_t app_name_max_len = 256;
+    char current_process[app_name_max_len] = { 0 };
+    get_proc_name(current_process, app_name_max_len);
+    for (auto app : DISABLE_VULKAN_FOR_APPLICATION) {
+        if (strcmp(app, current_process) == 0) {
+            ALOGI("\"%s\" disable vulkan api", app);
+            return -ENOENT;
+        }
+    }
 
     void* so = nullptr;
     for (auto key : HAL_SUBNAME_KEY_PROPERTIES) {
