diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index 429a6abb2..8248045ba 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -1457,6 +1457,25 @@ config ARM64_ILP32
 	  is an ABI where long and pointers are 32bits but it uses the AARCH64
 	  instruction set.
 
+config KBOX
+	bool "Kernel support for Kbox"
+	select STAGING
+	select ANDROID
+	select ASHMEM
+	select ANDROID_BINDER_IPC
+	select ANDROID_BINDERFS
+	select ANDROID_ALARM
+	select ANON_VMA_NAME
+	select TANGO32
+	select VFAT_FS
+	select INPUT_UINPUT
+   select DMABUF_HEAPS
+   select DMABUF_HEAPS_SYSTEM
+   select DMABUF_HEAPS_CMA
+	default y
+	help
+	  This option enables kernel support for Kbox and selects dependent kernel options.
+
 menuconfig AARCH32_EL0
 	bool "Kernel support for 32-bit EL0"
 	depends on ARM64_4K_PAGES || EXPERT
-- 
2.33.0

