diff --git a/include/linux/memcontrol.h b/include/linux/memcontrol.h
index 24794b56e..1fdd1438c 100644
--- a/include/linux/memcontrol.h
+++ b/include/linux/memcontrol.h
@@ -1047,6 +1047,8 @@ void unlock_page_memcg(struct page *page);
 
 void __mod_memcg_state(struct mem_cgroup *memcg, int idx, int val);
 
+unsigned long memcg_page_state(struct mem_cgroup *memcg, int idx);
+
 /* idx can be of type enum memcg_stat_item or node_stat_item */
 static inline void mod_memcg_state(struct mem_cgroup *memcg,
 				   int idx, int val)
diff --git a/kernel/sys.c b/kernel/sys.c
old mode 100644
new mode 100755
index 6b276cd25..ea4c1d0e1
--- a/kernel/sys.c
+++ b/kernel/sys.c
@@ -70,6 +70,7 @@
 #include <generated/utsrelease.h>
 
 #include <linux/uaccess.h>
+#include <linux/kbox_kernel_switch.h>
 #include <asm/io.h>
 #include <asm/unistd.h>
 
@@ -2667,6 +2668,20 @@ SYSCALL_DEFINE3(getcpu, unsigned __user *, cpup, unsigned __user *, nodep,
 	return err ? -EFAULT : 0;
 }
 
+static void si_meminfo_cgroup(struct sysinfo *val)
+{
+        struct mem_cgroup *memcg = mem_cgroup_from_task(current);
+
+        val->totalram = memcg->memory.max;
+        val->sharedram = memcg_page_state(memcg, NR_SHMEM);
+        val->freeram = val->totalram - page_counter_read(&memcg->memory);
+        val->bufferram = memcg_page_state(memcg, NR_FILE_PAGES);
+        val->totalhigh = totalhigh_pages();
+        val->freehigh = nr_free_highpages();
+        val->mem_unit = PAGE_SIZE;
+}
+
+
 /**
  * do_sysinfo - fill in sysinfo struct
  * @info: pointer to buffer to fill
@@ -2687,7 +2702,14 @@ static int do_sysinfo(struct sysinfo *info)
 
 	info->procs = nr_threads;
 
+if (kbox_enable) {
+	if(mem_cgroup_from_task(current)->memory.max == PAGE_COUNTER_MAX)
+		si_meminfo(info);
+	else
+		si_meminfo_cgroup(info);
+} else {
 	si_meminfo(info);
+}
 	si_swapinfo(info);
 
 	/*
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index c9ffc793e..7406f367b 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -726,7 +726,7 @@ void __mod_memcg_state(struct mem_cgroup *memcg, int idx, int val)
 }
 
 /* idx can be of type enum memcg_stat_item or node_stat_item. */
-static unsigned long memcg_page_state(struct mem_cgroup *memcg, int idx)
+unsigned long memcg_page_state(struct mem_cgroup *memcg, int idx)
 {
 	long x = READ_ONCE(memcg->vmstats.state[idx]);
 #ifdef CONFIG_SMP
