diff --git a/src/com/android/gallery3d/glrenderer/TiledTexture.java b/src/com/android/gallery3d/glrenderer/TiledTexture.java
index 78cb2f2de..b6ac47459 100644
--- a/src/com/android/gallery3d/glrenderer/TiledTexture.java
+++ b/src/com/android/gallery3d/glrenderer/TiledTexture.java
@@ -113,6 +113,7 @@ public class TiledTexture implements Texture {
         public int offsetX;
         public int offsetY;
         public Bitmap bitmap;
+        public Bitmap sourceBitmap;
         public Tile nextFreeTile;
         public int contentWidth;
         public int contentHeight;
@@ -134,6 +135,8 @@ public class TiledTexture implements Texture {
             Bitmap localBitmapRef = bitmap;
             bitmap = null;
 
+            if (localBitmapRef == null) localBitmapRef = sourceBitmap;
+
             if (localBitmapRef != null) {
                 int x = BORDER_SIZE - offsetX;
                 int y = BORDER_SIZE - offsetY;
@@ -147,6 +150,8 @@ public class TiledTexture implements Texture {
                 if (y > 0) sCanvas.drawLine(0, y - 1, TILE_SIZE, y - 1, sPaint);
                 if (r < CONTENT_SIZE) sCanvas.drawLine(r, 0, r, TILE_SIZE, sPaint);
                 if (b < CONTENT_SIZE) sCanvas.drawLine(0, b, TILE_SIZE, b, sPaint);
+            } else {
+                sCanvas.drawColor(Color.TRANSPARENT, Mode.SRC);
             }
 
             return sUploadBitmap;
@@ -161,6 +166,7 @@ public class TiledTexture implements Texture {
     private static void freeTile(Tile tile) {
         tile.invalidateContent();
         tile.bitmap = null;
+        tile.sourceBitmap = null;
         synchronized (sFreeTileLock) {
             tile.nextFreeTile = sFreeTileHead;
             sFreeTileHead = tile;
@@ -210,6 +216,7 @@ public class TiledTexture implements Texture {
                 tile.offsetX = x;
                 tile.offsetY = y;
                 tile.bitmap = bitmap;
+                tile.sourceBitmap = bitmap;
                 tile.setSize(
                         Math.min(CONTENT_SIZE, mWidth - x),
                         Math.min(CONTENT_SIZE, mHeight - y));
