diff --git a/fs_mgr/libfstab/boot_config.cpp b/fs_mgr/libfstab/boot_config.cpp
index b21495ecd..f566f1e01 100644
--- a/fs_mgr/libfstab/boot_config.cpp
+++ b/fs_mgr/libfstab/boot_config.cpp
@@ -154,13 +154,15 @@ bool GetKernelCmdlineFromString(const std::string& cmdline, const std::string& k
 
 void ImportKernelCmdline(const std::function<void(std::string, std::string)>& fn) {
     std::string cmdline;
-    android::base::ReadFileToString("/proc/cmdline", &cmdline);
+    android::base::ReadFileToString("/proc/self/cmdline", &cmdline); // kbox_hacked
+    std::replace(cmdline.begin(), cmdline.end(), '\0', ' '); // kbox_hacked
     ImportKernelCmdlineFromString(android::base::Trim(cmdline), fn);
 }
 
 bool GetKernelCmdline(const std::string& key, std::string* out) {
     std::string cmdline;
-    android::base::ReadFileToString("/proc/cmdline", &cmdline);
+    android::base::ReadFileToString("/proc/self/cmdline", &cmdline); // kbox_hacked
+    std::replace(cmdline.begin(), cmdline.end(), '\0', ' '); // kbox_hacked
     return GetKernelCmdlineFromString(android::base::Trim(cmdline), key, out);
 }
 
diff --git a/init/Android.bp b/init/Android.bp
index 18a79d6c4..dd3e70267 100644
--- a/init/Android.bp
+++ b/init/Android.bp
@@ -247,6 +247,10 @@ cc_defaults {
 cc_library_static {
     name: "libinit",
     defaults: ["libinit_defaults"],
+    cflags: [
+       "-Wno-unused-argument",
+       "-Wno-unused-function",
+    ],
 }
 
 cc_library_static {
diff --git a/init/devices.cpp b/init/devices.cpp
index f2bb9d276..c759bb3da 100644
--- a/init/devices.cpp
+++ b/init/devices.cpp
@@ -444,6 +444,15 @@ std::vector<std::string> DeviceHandler::GetBlockDeviceSymlinks(const Uevent& uev
     return links;
 }
 
+void DeviceHandler::chinit_dev(const char *path)
+{
+    std::vector<std::string> links;
+    auto[mode, uid, gid] = GetDevicePermissions(path, links);
+    // NOTICE("chinit_dev %s %d %d 0%o\n", path, uid, gid, mode);
+    chown(path, uid, gid);
+    chmod(path, mode);
+}
+
 static void RemoveDeviceMapperLinks(const std::string& devpath) {
     std::vector<std::string> dirs = {
             "/dev/block/mapper",
@@ -473,6 +482,12 @@ static void RemoveDeviceMapperLinks(const std::string& devpath) {
 
 void DeviceHandler::HandleDevice(const std::string& action, const std::string& devpath, bool block,
                                  int major, int minor, const std::vector<std::string>& links) const {
+ 
+    std::string temp = devpath.substr(0, 11);
+    if ((temp.compare("/dev/input/")) == 0) {
+        return;
+    }
+
     if (action == "add") {
         MakeDevice(devpath, block, major, minor, links);
     }
@@ -609,7 +624,14 @@ DeviceHandler::DeviceHandler(std::vector<Permissions> dev_permissions,
       subsystems_(std::move(subsystems)),
       boot_devices_(std::move(boot_devices)),
       skip_restorecon_(skip_restorecon),
-      sysfs_mount_point_("/sys") {}
+      sysfs_mount_point_("/sys") {
+    chinit_dev("/dev/binder");
+    chinit_dev("/dev/hwbinder");
+    chinit_dev("/dev/vndbinder");
+    chinit_dev("/dev/ashmem");
+    chinit_dev("/dev/uinput");
+    chinit_dev("/dev/dri/renderD128");
+}
 
 DeviceHandler::DeviceHandler()
     : DeviceHandler(std::vector<Permissions>{}, std::vector<SysfsPermissions>{},
diff --git a/init/devices.h b/init/devices.h
index 6da123259..610e883f0 100644
--- a/init/devices.h
+++ b/init/devices.h
@@ -142,6 +142,7 @@ class DeviceHandler : public UeventHandler {
     void MakeDevice(const std::string& path, bool block, int major, int minor,
                     const std::vector<std::string>& links) const;
     std::vector<std::string> GetBlockDeviceSymlinks(const Uevent& uevent) const;
+    void chinit_dev(const char *path);
     void HandleDevice(const std::string& action, const std::string& devpath, bool block, int major,
                       int minor, const std::vector<std::string>& links) const;
     void FixupSysPermissions(const std::string& upath, const std::string& subsystem) const;
diff --git a/init/first_stage_init.cpp b/init/first_stage_init.cpp
index e06a64593..8485b3dc1 100644
--- a/init/first_stage_init.cpp
+++ b/init/first_stage_init.cpp
@@ -36,6 +36,7 @@
 #include <android-base/chrono_utils.h>
 #include <android-base/file.h>
 #include <android-base/logging.h>
+#include <android-base/strings.h>
 #include <android-base/stringprintf.h>
 #include <android/avf_cc_flags.h>
 #include <fs_mgr.h>
@@ -348,10 +349,11 @@ int FirstStageMain(int argc, char** argv) {
     CHECKCALL(setenv("PATH", _PATH_DEFPATH, 1));
     // Get the basic filesystem setup we need put together in the initramdisk
     // on / and then we'll let the rc file figure out the rest.
-    CHECKCALL(mount("tmpfs", "/dev", "tmpfs", MS_NOSUID, "mode=0755"));
     CHECKCALL(mkdir("/dev/pts", 0755));
     CHECKCALL(mkdir("/dev/socket", 0755));
     CHECKCALL(mkdir("/dev/dm-user", 0755));
+    mount("/system/etc", "/etc", "none", MS_BIND, NULL); // cgroup fix
+    unshare(CLONE_NEWCGROUP);
     CHECKCALL(mount("devpts", "/dev/pts", "devpts", 0, NULL));
 #define MAKE_STR(x) __STRING(x)
     CHECKCALL(mount("proc", "/proc", "proc", 0, "hidepid=2,gid=" MAKE_STR(AID_READPROC)));
@@ -420,7 +422,6 @@ int FirstStageMain(int argc, char** argv) {
         for (const auto& [error_string, error_errno] : errors) {
             LOG(ERROR) << error_string << " " << strerror(error_errno);
         }
-        LOG(FATAL) << "Init encountered errors starting first stage, aborting";
     }
 
     LOG(INFO) << "init first stage started!";
@@ -528,6 +529,7 @@ int FirstStageMain(int argc, char** argv) {
         if (!fsm) {
             fsm = CreateFirstStageMount(cmdline);
         }
+        #if 0
         if (!fsm) {
             LOG(FATAL) << "FirstStageMount not available";
         }
@@ -537,8 +539,9 @@ int FirstStageMain(int argc, char** argv) {
         }
 
         if (!fsm->DoFirstStageMount()) {
-            LOG(FATAL) << "Failed to mount required partitions early ...";
+            LOG(ERROR) << "Failed to mount required partitions early ...";
         }
+        #endif
     }
 
     struct stat new_root_info {};
@@ -557,12 +560,22 @@ int FirstStageMain(int argc, char** argv) {
            1);
 
     const char* path = "/system/bin/init";
-    const char* args[] = {path, "selinux_setup", nullptr};
+    std::vector<const char *> args = {path, "second_stage"};
+    std::string init_cmdline;
+    android::base::ReadFileToString("/proc/self/cmdline", &init_cmdline);
+    std::replace(init_cmdline.begin(), init_cmdline.end(), '\0', ' ');
+    auto cmd_vector = android::base::Split(android::base::Trim(init_cmdline), " ");
+    int i = 0;
+    for (const auto& entry : cmd_vector) {
+        if (i++ == 0) continue; // ignore first arg '/init'
+        args.push_back(entry.c_str());
+    }
+    args.push_back(nullptr);
     auto fd = open("/dev/kmsg", O_WRONLY | O_CLOEXEC);
     dup2(fd, STDOUT_FILENO);
     dup2(fd, STDERR_FILENO);
     close(fd);
-    execv(path, const_cast<char**>(args));
+    execv(path, const_cast<char**>(args.data()));
 
     // execv() only returns if an error happened, in which case we
     // panic and never fall through this conditional.
diff --git a/init/init.cpp b/init/init.cpp
index 6c8089926..5404733b8 100644
--- a/init/init.cpp
+++ b/init/init.cpp
@@ -41,6 +41,7 @@
 #include <optional>
 #include <thread>
 #include <vector>
+#include <string>
 
 #include <android-base/chrono_utils.h>
 #include <android-base/file.h>
@@ -109,6 +110,9 @@ using android::base::unique_fd;
 using android::fs_mgr::AvbHandle;
 using android::snapshot::SnapshotManager;
 
+#define CPUINFO_PATH "/proc/cpuinfo"
+#define CPUTMP_PATH "/system/cpuinfo_tmp"
+
 namespace android {
 namespace init {
 
@@ -364,7 +368,8 @@ void PropertyChanged(const std::string& name, const std::string& value) {
     // In non-thermal-shutdown case, 'shutdown' trigger will be fired to let device specific
     // commands to be executed.
     if (name == "sys.powerctl") {
-        trigger_shutdown(value);
+        // trigger_shutdown(value);
+        LOG(INFO) << "kbox don't reboot";
     }
 
     if (property_triggers_enabled) {
@@ -832,6 +837,49 @@ static void UmountSecondStageRes() {
     }
 }
 
+static int PrepareFile() {
+    std::ifstream sourceFile(CPUINFO_PATH);
+
+    if (!sourceFile.is_open()) {
+        LOG(ERROR) << "can not open file: " << CPUINFO_PATH;
+        return -1;
+    }
+
+    std::ofstream destFile(CPUTMP_PATH);
+    if (!destFile.is_open()) {
+        sourceFile.close();
+        LOG(ERROR) << "can not create tmp file: " << CPUTMP_PATH;
+        return -1;
+    }
+
+    std::string line;
+    int oneCpuInfoStep = 0;
+    int totalLine = 0;
+    int currentLine = 0;
+
+    while (std::getline(sourceFile, line)) {
+        oneCpuInfoStep++;
+        destFile << line << std::endl;
+        if (line.empty()) {
+            break;
+        }
+    }
+
+    totalLine = oneCpuInfoStep * 7 - 1;
+    while (std::getline(sourceFile, line)) {
+        currentLine++;
+        if (currentLine >= 1 && currentLine <= totalLine) {
+            destFile << line << std::endl;
+            continue;
+        }
+        break;
+    }
+
+    sourceFile.close();
+    destFile.close();
+    return 0;
+}
+
 static void MountExtraFilesystems() {
 #define CHECKCALL(x) \
     if ((x) != 0) PLOG(FATAL) << #x " failed.";
@@ -849,6 +897,11 @@ static void MountExtraFilesystems() {
     // /linkerconfig is used to keep generated linker configuration
     CHECKCALL(mount("tmpfs", "/linkerconfig", "tmpfs", MS_NOEXEC | MS_NOSUID | MS_NODEV,
                     "mode=0755,uid=0,gid=0"));
+
+    // /sys/devices/system/cpu cpumock目录挂载
+    // cpuinfo修改挂载
+    (PrepareFile());
+
 #undef CHECKCALL
 }
 
@@ -937,10 +990,10 @@ int SecondStageMain(int argc, char** argv) {
     trigger_shutdown = [](const std::string& command) { shutdown_state.TriggerShutdown(command); };
 
     SetStdioToDevNull(argv);
-    InitKernelLogging(argv);
+    // InitKernelLogging(argv);
     LOG(INFO) << "init second stage started!";
 
-    SelinuxSetupKernelLogging();
+    // SelinuxSetupKernelLogging();
 
     // Update $PATH in the case the second stage init is newer than first stage init, where it is
     // first set.
@@ -997,8 +1050,8 @@ int SecondStageMain(int argc, char** argv) {
     MountExtraFilesystems();
 
     // Now set up SELinux for second stage.
-    SelabelInitialize();
-    SelinuxRestoreContext();
+    // SelabelInitialize();
+    // SelinuxRestoreContext();
 
     Epoll epoll;
     if (auto result = epoll.Open(); !result.ok()) {
diff --git a/init/ipconfigstore/Android.bp b/init/ipconfigstore/Android.bp
new file mode 100644
index 000000000..34ed932dd
--- /dev/null
+++ b/init/ipconfigstore/Android.bp
@@ -0,0 +1,8 @@
+cc_binary {
+    name: "ipconfigstore",
+    srcs: ["*.cc"],
+    cflags: ["-Wall", "-Werror"],
+    shared_libs: ["libbase"],
+    vendor: true,
+}
+
diff --git a/init/ipconfigstore/data.cc b/init/ipconfigstore/data.cc
new file mode 100644
index 000000000..b87e77069
--- /dev/null
+++ b/init/ipconfigstore/data.cc
@@ -0,0 +1,65 @@
+#include <string.h>
+#include <strings.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <ctype.h>
+
+#include "data.h"
+
+uint16_t convertBigEndianUInt16(uint16_t value)
+{
+	union { uint16_t value; unsigned char data[2]; } aux = { 0x4142 };
+
+	if (aux.data[0] == 0x41)
+	{
+		return value;
+	}
+
+	aux.data[0] = (value >> 8) & 0xff;
+	aux.data[1] = value & 0xff;
+
+	return aux.value;
+}
+
+uint32_t convertBigEndianUInt32(uint32_t value)
+{
+	union { uint32_t value; unsigned char data[4]; } aux = { 0x41424344 };
+
+	if (aux.data[0] == 0x41)
+	{
+		return value;
+	}
+
+	aux.data[0] = (value >> 24) & 0xff;
+	aux.data[1] = (value >> 16) & 0xff;
+	aux.data[2] = (value >> 8) & 0xff;
+	aux.data[3] = value & 0xff;
+
+	return aux.value;
+}
+
+bool writePackedString(const std::string& str, FILE *stream)
+{
+    const char *string = str.c_str();
+	size_t stringLength = strlen(string);
+
+	if (!writePackedUInt16(stringLength, stream))
+	{
+		return false;
+	}
+
+	return fwrite(string, stringLength, 1, stream) == 1;
+}
+
+bool writePackedUInt16(uint16_t value, FILE *stream)
+{
+	uint16_t buffer = convertBigEndianUInt16(value);
+	return fwrite(&buffer, sizeof buffer, 1, stream) == 1;
+}
+
+bool writePackedUInt32(uint32_t value, FILE *stream)
+{
+	uint32_t buffer = convertBigEndianUInt32(value);
+	return fwrite(&buffer, sizeof buffer, 1, stream) == 1;
+}
+
diff --git a/init/ipconfigstore/data.h b/init/ipconfigstore/data.h
new file mode 100644
index 000000000..12c3a1af9
--- /dev/null
+++ b/init/ipconfigstore/data.h
@@ -0,0 +1,13 @@
+#pragma once
+
+#include <cstdint>
+#include <string>
+
+#include <stdio.h>
+
+uint16_t convertBigEndianUInt16(uint16_t value);
+uint32_t convertBigEndianUInt32(uint32_t value);
+
+bool writePackedString(const std::string& str, FILE *stream);
+bool writePackedUInt16(uint16_t value, FILE *stream);
+bool writePackedUInt32(uint32_t value, FILE *stream);
diff --git a/init/ipconfigstore/main.cc b/init/ipconfigstore/main.cc
new file mode 100644
index 000000000..ecbc0ea18
--- /dev/null
+++ b/init/ipconfigstore/main.cc
@@ -0,0 +1,166 @@
+#include <sys/types.h>
+#include <ifaddrs.h>
+#include <sys/socket.h>
+#include <netinet/in.h>
+#include <net/if.h>
+#include <sys/ioctl.h>
+#include <errno.h>
+#include <string.h>
+#include <net/route.h>
+#include <unistd.h>
+#include <stdio.h>
+#include <arpa/inet.h>
+
+#include <set>
+
+#include <android-base/properties.h>
+
+#include "data.h"
+
+using namespace android::base;
+
+struct ipconfig {
+    uint32_t mask;
+    char ipv4[16];
+    char gateway[16];
+};
+
+static int get_gateway(char *dev, char *ret) {
+    FILE *fp;
+    char buf[256]; // 128 is enough for linux
+    char iface[16];
+    unsigned long dest_addr, gate_addr;
+    fp = fopen("/proc/net/route", "r");
+    if (fp == NULL) return -1;
+    /* Skip title line */
+    fgets(buf, sizeof(buf), fp);
+    while (fgets(buf, sizeof(buf), fp)) {
+        if (sscanf(buf, "%s\t%lX\t%lX", iface, &dest_addr, &gate_addr) != 3 || dest_addr != 0 || strcmp(dev, iface)) continue;
+        inet_ntop(AF_INET, &gate_addr, ret, INET_ADDRSTRLEN);
+        break;
+    }
+
+    fclose(fp);
+    return 0;
+}
+
+static int bitcount(uint32_t n)
+{
+    int count=0;
+    while (n) {
+        count++;
+        n &= (n - 1);
+    }
+    return count;
+}
+
+static int get_conf(struct ipconfig *conf) {
+    struct ifaddrs *ifAddrStruct;
+    void *tmpAddrPtr=NULL;
+    getifaddrs(&ifAddrStruct);
+    while (ifAddrStruct != NULL) {
+        if (ifAddrStruct->ifa_addr->sa_family==AF_INET && !strcmp("eth0", ifAddrStruct->ifa_name)) {
+            tmpAddrPtr=&((struct sockaddr_in *)ifAddrStruct->ifa_addr)->sin_addr;
+            inet_ntop(AF_INET, tmpAddrPtr, conf->ipv4, INET_ADDRSTRLEN);
+            conf->mask = bitcount(((struct sockaddr_in *)ifAddrStruct->ifa_netmask)->sin_addr.s_addr);
+            break;
+        }
+        ifAddrStruct=ifAddrStruct->ifa_next;
+    }
+    freeifaddrs(ifAddrStruct);
+    get_gateway((char *) "eth0", conf->gateway);
+    return 0;
+}
+
+static void write_dns(FILE *fp) {
+    std::set<std::string> dnsList;
+    auto ndns = GetIntProperty("ro.boot.kbox_net_ndns", 0);
+    for (int i = 1; i <= ndns; ++i) {
+        dnsList.insert(GetProperty("ro.boot.kbox_net_dns" + std::to_string(i), ""));
+    }
+    if (dnsList.empty()) dnsList.insert("8.8.8.8");
+
+    for (auto& dns: dnsList) {
+        writePackedString("dns", fp);
+        writePackedString(dns.c_str(), fp);
+    }
+}
+
+static void write_proxy(FILE *fp) {
+    // static | pac | none | unassigned
+    std::string proxy_type = GetProperty("ro.boot.kbox_net_proxy_type", "");
+    if ("static" == proxy_type) {
+        writePackedString("proxySettings", fp);
+        writePackedString("STATIC", fp);
+
+        writePackedString("proxyHost", fp);
+        writePackedString(GetProperty("ro.boot.kbox_net_proxy_host", "").c_str(), fp);
+
+        writePackedString("proxyPort", fp);
+        writePackedUInt32(GetIntProperty("ro.boot.kbox_net_proxy_port", 3128), fp);
+
+        writePackedString("exclusionList", fp);
+        writePackedString(GetProperty("ro.boot.kbox_net_proxy_exclude_list", "").c_str(), fp);
+    } else if ("pac" == proxy_type) {
+        writePackedString("proxySettings", fp);
+        writePackedString("PAC", fp);
+
+        writePackedString("proxyPac", fp);
+        writePackedString(GetProperty("ro.boot.kbox_net_proxy_pac", "").c_str(), fp);
+    } else if ("none" == proxy_type) {
+        writePackedString("proxySettings", fp);
+        writePackedString("NONE", fp);
+    } else {
+        // ignored
+    }
+}
+
+static int write_conf(char *file_path, struct ipconfig *conf, uint32_t v) {
+    FILE *fp = fopen(file_path, "w+");
+
+    writePackedUInt32(v, fp); // version
+
+    writePackedString("ipAssignment", fp);
+    writePackedString("STATIC", fp);
+
+    writePackedString("linkAddress", fp);
+    writePackedString(conf->ipv4, fp);
+    writePackedUInt32(conf->mask, fp);
+
+    writePackedString("gateway", fp);
+    writePackedUInt32(1, fp); // Default route (dest).
+    writePackedString("0.0.0.0", fp);
+    writePackedUInt32(0, fp);
+    writePackedUInt32(1, fp); // Have a gateway.
+    writePackedString(conf->gateway, fp);
+
+    write_dns(fp);
+
+    write_proxy(fp);
+
+    writePackedString("id", fp);
+    if (v == 2) writePackedUInt32(0, fp);
+    else writePackedString("eth0", fp);
+
+    writePackedString("eos", fp);
+
+    fclose(fp);
+    return 0;
+}
+
+int main(int argc, char **argv) {
+    (void)argc;
+    (void)argv;
+
+    uint32_t v = 3;
+    // use V2 for Android 8.1
+    if (GetIntProperty("ro.build.version.sdk", 0) <= 27) v = 2;
+
+    struct ipconfig conf;
+    get_conf(&conf);
+    printf("ipconfig: ipv4: %s, mask: %i, gateway: %s", conf.ipv4, conf.mask, conf.gateway);
+    write_conf((char *)"/data/misc/ethernet/ipconfig.txt", &conf, v);
+    write_conf((char *)"/data/misc/apexdata/com.android.tethering/misc/ethernet/ipconfig.txt", &conf, v);
+    return 0;
+}
+
diff --git a/init/property_service.cpp b/init/property_service.cpp
index f2606e3c5..62f9c34ad 100644
--- a/init/property_service.cpp
+++ b/init/property_service.cpp
@@ -1224,6 +1224,8 @@ void PropertyLoadBootDefaults() {
     load_properties_from_file("/odm_dlkm/etc/build.prop", nullptr, &properties);
     load_properties_from_partition("odm", /* support_legacy_path_until */ 28);
     load_properties_from_partition("product", /* support_legacy_path_until */ 30);
+    load_properties_from_file("/kbox_prop/default.prop", nullptr, &properties);
+    load_properties_from_file("/kbox_prop/build.prop", nullptr, &properties);
 
     if (access(kDebugRamdiskProp, R_OK) == 0) {
         LOG(INFO) << "Loading " << kDebugRamdiskProp;
@@ -1351,7 +1353,7 @@ static void ExportKernelBootProps() {
         { "ro.boot.mode",       "ro.bootmode",   "unknown", },
         { "ro.boot.baseband",   "ro.baseband",   "unknown", },
         { "ro.boot.bootloader", "ro.bootloader", "unknown", },
-        { "ro.boot.hardware",   "ro.hardware",   "unknown", },
+        { "ro.boot.hardware",   "ro.hardware",   "kbox", },
         { "ro.boot.revision",   "ro.revision",   "0", },
             // clang-format on
     };
@@ -1393,6 +1395,8 @@ static void ProcessKernelCmdline() {
     android::fs_mgr::ImportKernelCmdline([&](const std::string& key, const std::string& value) {
         if (StartsWith(key, ANDROIDBOOT_PREFIX)) {
             InitPropertySet("ro.boot." + key.substr(ANDROIDBOOT_PREFIX.size()), value);
+        } else if (StartsWith(key, "ro.")) {
+            InitPropertySet(key, value);
         }
     });
 }
diff --git a/init/selinux.cpp b/init/selinux.cpp
index 01af2b64d..8c49f9e4a 100644
--- a/init/selinux.cpp
+++ b/init/selinux.cpp
@@ -106,6 +106,7 @@ EnforcingStatus StatusFromProperty() {
 }
 
 bool IsEnforcing() {
+    kbox_hack_p(false);
     if (ALLOW_PERMISSIVE_SELINUX) {
         return StatusFromProperty() == SELINUX_ENFORCING;
     }
@@ -686,14 +687,14 @@ void LoadSelinuxPolicyAndroid() {
 
 int SetupSelinux(char** argv) {
     SetStdioToDevNull(argv);
-    InitKernelLogging(argv);
+    // InitKernelLogging(argv);
 
     if (REBOOT_BOOTLOADER_ON_PANIC) {
         InstallRebootSignalHandlers();
     }
 
     boot_clock::time_point start_time = boot_clock::now();
-
+#if 0
     SelinuxSetupKernelLogging();
 
     // TODO(b/287206497): refactor into different headers to only include what we need.
@@ -721,7 +722,7 @@ int SetupSelinux(char** argv) {
     if (selinux_android_restorecon("/system/bin/init", 0) == -1) {
         PLOG(FATAL) << "restorecon failed of /system/bin/init failed";
     }
-
+#endif
     setenv(kEnvSelinuxStartedAt, std::to_string(start_time.time_since_epoch().count()).c_str(), 1);
 
     const char* path = "/system/bin/init";
diff --git a/init/service.cpp b/init/service.cpp
index d76a5d5e0..981f4ee82 100644
--- a/init/service.cpp
+++ b/init/service.cpp
@@ -76,6 +76,7 @@ namespace android {
 namespace init {
 
 static Result<std::string> ComputeContextFromExecutable(const std::string& service_path) {
+    kbox_hack_p("KBOX_HACKED");
     std::string computed_context;
 
     char* raw_con = nullptr;
diff --git a/init/subcontext.cpp b/init/subcontext.cpp
index 6a095fb7b..8166f4461 100644
--- a/init/subcontext.cpp
+++ b/init/subcontext.cpp
@@ -346,6 +346,7 @@ Result<std::vector<std::string>> Subcontext::ExpandArgs(const std::vector<std::s
 }
 
 void InitializeSubcontext() {
+    kbox_hack();
     if (IsMicrodroid()) {
         LOG(INFO) << "Not using subcontext for microdroid";
         return;
diff --git a/libcutils/android_reboot.cpp b/libcutils/android_reboot.cpp
index e0def711d..b50b7f896 100644
--- a/libcutils/android_reboot.cpp
+++ b/libcutils/android_reboot.cpp
@@ -23,7 +23,12 @@
 
 #define TAG "android_reboot"
 
+#ifndef kbox_hack_p
+#define kbox_hack_p(p) if (true) return p
+#endif
+
 int android_reboot(unsigned cmd, int /*flags*/, const char* arg) {
+    kbox_hack_p(0);
     int ret;
     const char* restart_cmd = NULL;
     char* prop_value;
diff --git a/libprocessgroup/processgroup.cpp b/libprocessgroup/processgroup.cpp
index 83a2258bf..37bf761d0 100644
--- a/libprocessgroup/processgroup.cpp
+++ b/libprocessgroup/processgroup.cpp
@@ -273,6 +273,7 @@ static bool RemoveEmptyUidCgroups(const std::string& uid_path) {
 }
 
 void removeAllEmptyProcessGroups() {
+    kbox_hack();
     LOG(VERBOSE) << "removeAllEmptyProcessGroups()";
 
     std::vector<std::string> cgroups;
diff --git a/libsysutils/src/NetlinkEvent.cpp b/libsysutils/src/NetlinkEvent.cpp
index 55bbe46e1..b7288c227 100644
--- a/libsysutils/src/NetlinkEvent.cpp
+++ b/libsysutils/src/NetlinkEvent.cpp
@@ -159,12 +159,14 @@ NetlinkEvent::~NetlinkEvent() {
 
 void NetlinkEvent::dump() {
     int i;
-
+    std::string buffer ;
     for (i = 0; i < NL_PARAMS_MAX; i++) {
         if (!mParams[i])
             break;
-        SLOGD("NL param '%s'\n", mParams[i]);
+        buffer += std::to_string(i) + ":" + mParams[i] ;
     }
+    SLOGE("NL param '%s'\n",buffer.c_str());
+    SLOGE(" 下一个 \n");
 }
 
 /*
diff --git a/rootdir/init.rc b/rootdir/init.rc
index 1acd63774..22bcd0a42 100644
--- a/rootdir/init.rc
+++ b/rootdir/init.rc
@@ -265,10 +265,10 @@ on init
     mkdir /dev/binderfs
     mount binder binder /dev/binderfs stats=global
     chmod 0755 /dev/binderfs
+    # exec -- /vendor/bin/binder_alloc /dev/binderfs/binder-control binder hwbinder vndbinder
 
     # Mount fusectl
-    mount fusectl none /sys/fs/fuse/connections
-
+    # mount fusectl none /sys/fs/fuse/connections
     symlink /dev/binderfs/binder /dev/binder
     symlink /dev/binderfs/hwbinder /dev/hwbinder
     symlink /dev/binderfs/vndbinder /dev/vndbinder
@@ -588,7 +588,6 @@ on post-fs
 
     # Once everything is setup, no need to modify /.
     # The bind+remount combination allows this to work in containers.
-    mount rootfs rootfs / remount bind ro nodev
 
     # Mount default storage into root namespace
     mount none /mnt/user/0 /storage bind rec
@@ -1095,7 +1094,7 @@ on post-fs-data
     # sys.memfd_use set to false by default, which keeps it disabled
     # until it is confirmed that apps and vendor processes don't make
     # IOCTLs on ashmem fds any more.
-    setprop sys.use_memfd false
+    setprop sys.use_memfd true
 
     # Set fscklog permission
     chown root system /dev/fscklogs/log
@@ -1269,6 +1268,12 @@ on property:sys.boot_completed=1
     # Setup per_boot directory so other .rc could start to use it on boot_completed
     exec - system system -- /bin/rm -rf /data/per_boot
     mkdir /data/per_boot 0700 system system encryption=Require key=per_boot_ref
+    mount none /data/local/cpu /sys/devices/system/cpu bind rec
+    mount none /system/cpuinfo_tmp /proc/cpuinfo bind rec
+    mount none /lxcfs-proc/meminfo /proc/meminfo bind rec
+    mount none /lxcfs-proc/stat /proc/stat bind rec
+    mount none /lxcfs-proc/swaps /proc/swaps bind rec
+    mount none /lxcfs-proc/uptime /proc/uptime bind rec
 
 # system server cannot write to /proc/sys files,
 # and chown/chmod does not work for /proc/sys/ entries.
diff --git a/rootdir/init.usb.rc b/rootdir/init.usb.rc
index b30d6d02c..4dd943149 100644
--- a/rootdir/init.usb.rc
+++ b/rootdir/init.usb.rc
@@ -27,9 +27,10 @@ on property:vendor.sys.usb.adb.disabled=*
 # overridden in `on boot` action of init.hardware.rc.
 on init
     setprop sys.usb.configfs 0
+    start adbd
 
 # Used to disable USB when switching states
-on property:sys.usb.config=none && property:sys.usb.configfs=0
+on property:sys.usb.config=none && property:sys.usb.configfs=0 && false
     stop adbd
     write /sys/class/android_usb/android0/enable 0
     write /sys/class/android_usb/android0/bDeviceClass 0
@@ -39,6 +40,7 @@ on property:sys.usb.config=none && property:sys.usb.configfs=0
 # This is the fallback configuration if the
 # USB manager fails to set a standard configuration
 on property:sys.usb.config=adb && property:sys.usb.configfs=0
+    stop adbd
     write /sys/class/android_usb/android0/enable 0
     write /sys/class/android_usb/android0/idVendor 18d1
     write /sys/class/android_usb/android0/idProduct 4EE7
@@ -107,7 +109,7 @@ on property:sys.usb.config=accessory,audio_source,adb && property:sys.usb.config
 # Used to set USB configuration at boot and to switch the configuration
 # when changing the default configuration
 on boot && property:persist.sys.usb.config=*
-    setprop sys.usb.config ${persist.sys.usb.config}
+    setprop sys.usb.config adb
 
 #
 # USB type C
