diff --git a/loader/Loader.cpp b/loader/Loader.cpp

--- a/loader/Loader.cpp
+++ b/loader/Loader.cpp
@@ -1001,9 +1001,11 @@ void legacyBpfLoader() {
             ALOGE("If this triggers reliably, you're probably missing kernel options or patches.");
             ALOGE("If this triggers randomly, you might be hitting some memory allocation "
                   "problems or startup script race.");
+            #if 0
             ALOGE("--- DO NOT EXPECT SYSTEM TO BOOT SUCCESSFULLY ---");
             sleep(20);
             exit(121);
+            #endif
         }
     }
 }

