diff --git a/src/amd/vulkan/meson.build b/src/amd/vulkan/meson.build
index 71c8b16d4fc..691ebd9d616 100644
--- a/src/amd/vulkan/meson.build
+++ b/src/amd/vulkan/meson.build
@@ -70,6 +70,7 @@ libradv_files = files(
   'radv_meta_fmask_copy.c',
   'radv_meta_fmask_expand.c',
   'radv_meta_astc_decode.c',
+  'radv_meta_rgba_encode.c',
   'radv_meta_resolve.c',
   'radv_meta_resolve_cs.c',
   'radv_meta_resolve_fs.c',
diff --git a/src/amd/vulkan/radv_image.c b/src/amd/vulkan/radv_image.c
index 3fd4ea777a8..dfd7bdca894 100644
--- a/src/amd/vulkan/radv_image.c
+++ b/src/amd/vulkan/radv_image.c
@@ -38,6 +38,7 @@
 #include <cutils/properties.h>
 
 #include "gfx10_format_table.h"
+#include "vk_texcompress_bcn.h"
 
 static const VkImageUsageFlagBits RADV_IMAGE_USAGE_WRITE_BITS =
    VK_IMAGE_USAGE_TRANSFER_DST_BIT | VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT |
@@ -546,6 +547,66 @@ radv_patch_image_from_extra_info(struct radv_device *device, struct radv_image *
 
 extern bool radv_is_format_emulated(const struct radv_physical_device *physical_device, VkFormat format);
 
+static char isEnableHardEncode[PROPERTY_VALUE_MAX];
+VkFormat
+radv_translate_rgba2dxt5(VkFormat format)
+{
+   switch (format) {
+   case VK_FORMAT_R8G8B8A8_UNORM:
+      return VK_FORMAT_BC3_UNORM_BLOCK;
+   case VK_FORMAT_R8G8B8A8_SRGB:
+      return VK_FORMAT_BC3_SRGB_BLOCK;
+   default:
+      return format;
+   }
+}
+
+static bool
+radv_need_hard_encode(const struct radv_physical_device *pdev, const VkImageCreateInfo *pCreateInfo)
+{
+   if (property_get("sys.vmi.vk.texturecompress", isEnableHardEncode, "0") && strcmp(isEnableHardEncode, "2") != 0) {
+      return false;
+   }
+
+   // 暂不支持对图像数组进行压缩
+   if (pCreateInfo->arrayLayers > 1) {
+      return false;
+   }
+
+   if (!radv_is_format_emulated(pdev, pCreateInfo->format)) {
+      return false;
+   }
+
+   const enum util_format_layout format_layout = vk_format_description(pCreateInfo->format)->layout;
+   if (format_layout != UTIL_FORMAT_LAYOUT_ASTC &&
+       vk_texcompress_etc2_emulation_format(pCreateInfo->format) != VK_FORMAT_R8G8B8A8_UNORM &&
+       vk_texcompress_etc2_emulation_format(pCreateInfo->format) != VK_FORMAT_R8G8B8A8_SRGB) {
+      return false;
+   }
+
+   // 过滤掉对稀疏纹理的压缩，对该部分纹理进行压缩时可能造成GPU reset。 
+   if (pCreateInfo->flags & (VK_IMAGE_CREATE_SPARSE_BINDING_BIT | VK_IMAGE_CREATE_SPARSE_RESIDENCY_BIT |
+      VK_IMAGE_CREATE_SPARSE_ALIASED_BIT)) {
+      return false;
+   }
+
+	// 检查图像的用途，防止压缩 swapchain 的 images；并只对 VK_IMAGE_TYPE_2D 类型纹理进行压缩
+	if (!(pCreateInfo->usage & VK_IMAGE_USAGE_SAMPLED_BIT) || pCreateInfo->imageType != VK_IMAGE_TYPE_2D) {
+		return false;
+	}
+
+   // 图像太大内存申请不了
+	if (pCreateInfo->extent.width >= 3960 && pCreateInfo->extent.height >= 3960) {
+	   return false;
+	}
+
+	// 检查图像大小
+	if (pCreateInfo->extent.width >= 256 && pCreateInfo->extent.height >= 256) {
+	   return true;
+	}
+	return false;
+}
+
 static VkFormat
 radv_image_get_plane_format(const struct radv_physical_device *pdev, const struct radv_image *image,
                             unsigned plane)
@@ -553,10 +614,15 @@ radv_image_get_plane_format(const struct radv_physical_device *pdev, const struc
    if (radv_is_format_emulated(pdev, image->vk_format)) {
       if (plane == 0)
          return image->vk_format;
-      if (vk_format_description(image->vk_format)->layout == UTIL_FORMAT_LAYOUT_ASTC)
+      if (vk_format_description(image->vk_format)->layout == UTIL_FORMAT_LAYOUT_ASTC) {
+         if (image->isNeedHardEncode)
+            return radv_translate_rgba2dxt5(vk_texcompress_astc_emulation_format(image->vk_format));
          return vk_texcompress_astc_emulation_format(image->vk_format);
-      else
+      } else {
+         if (image->isNeedHardEncode)
+            return radv_translate_rgba2dxt5(vk_texcompress_etc2_emulation_format(image->vk_format));
          return vk_texcompress_etc2_emulation_format(image->vk_format);
+      }
    }
    return vk_format_get_plane_format(image->vk_format, plane);
 }
@@ -1710,6 +1776,12 @@ radv_destroy_image(struct radv_device *device, const VkAllocationCallbacks *pAll
    vk_free2(&device->vk.alloc, pAllocator, image);
 }
 
+void
+radv_internal_image_finish(struct radv_image *image)
+{
+   vk_object_base_finish(&image->base);
+}
+
 static void
 radv_image_print_info(struct radv_device *device, struct radv_image *image)
 {
@@ -1890,9 +1962,11 @@ radv_image_create(VkDevice _device, const struct radv_image_create_info *create_
       vk_find_struct_const(pCreateInfo->pNext, EXTERNAL_MEMORY_IMAGE_CREATE_INFO);
    assert(pCreateInfo->sType == VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO);
 
+   bool needHardEncode = radv_need_hard_encode(device->physical_device, pCreateInfo);
+
    unsigned plane_count = radv_get_internal_plane_count(device->physical_device, format);
 
-   bool needSoftEncode = !external_info && radv_need_soft_encode(pCreateInfo);
+   bool needSoftEncode = !external_info && radv_need_soft_encode(pCreateInfo) && !needHardEncode;
    bool needSoftDecode = needSoftEncode && radv_need_soft_decode(format);
    VkFormat softDecodeFormat = radv_translate_etc(format);
    VkFormat softEncodeFormat = radv_translate_rgb2BC(format);
@@ -1936,6 +2010,7 @@ radv_image_create(VkDevice _device, const struct radv_image_create_info *create_
    image->usage = pCreateInfo->usage;
    image->flags = pCreateInfo->flags;
    image->plane_count = vk_format_get_plane_count(format);
+   image->isNeedHardEncode = needHardEncode;
 
    //满足要求的 ETC2 纹理使用软解
    if (needSoftDecode) {
diff --git a/src/amd/vulkan/radv_meta.c b/src/amd/vulkan/radv_meta.c
index b1080ef8935..9b5a73cd1be 100644
--- a/src/amd/vulkan/radv_meta.c
+++ b/src/amd/vulkan/radv_meta.c
@@ -601,8 +601,14 @@ radv_device_init_meta(struct radv_device *device)
    if (result != VK_SUCCESS)
       goto fail_astc_decode;
 
+   result = radv_device_init_meta_rgba_encode_state(device);
+   if (result != VK_SUCCESS)
+      goto fail_bcn_encode;
+
    return VK_SUCCESS;
 
+fail_bcn_encode:
+   radv_device_finish_meta_rgba_encode_state(device);
 fail_astc_decode:
    radv_device_finish_meta_astc_decode_state(device);
 fail_etc_decode:
@@ -644,6 +650,7 @@ fail_clear:
 void
 radv_device_finish_meta(struct radv_device *device)
 {
+   radv_device_finish_meta_rgba_encode_state(device);
    radv_device_finish_meta_etc_decode_state(device);
    radv_device_finish_meta_astc_decode_state(device);
    radv_device_finish_accel_struct_build_state(device);
diff --git a/src/amd/vulkan/radv_meta.h b/src/amd/vulkan/radv_meta.h
index d8ab75481d9..9702bfc9c7b 100644
--- a/src/amd/vulkan/radv_meta.h
+++ b/src/amd/vulkan/radv_meta.h
@@ -107,6 +107,9 @@ void radv_device_finish_meta_etc_decode_state(struct radv_device *device);
 VkResult radv_device_init_meta_astc_decode_state(struct radv_device *device, bool on_demand);
 void radv_device_finish_meta_astc_decode_state(struct radv_device *device);
 
+VkResult radv_device_init_meta_rgba_encode_state(struct radv_device *device);
+void radv_device_finish_meta_rgba_encode_state(struct radv_device *device);
+
 void radv_meta_save(struct radv_meta_saved_state *saved_state, struct radv_cmd_buffer *cmd_buffer,
                     uint32_t flags);
 
@@ -232,6 +235,9 @@ void radv_meta_decode_etc(struct radv_cmd_buffer *cmd_buffer, struct radv_image
 void radv_meta_decode_astc(struct radv_cmd_buffer *cmd_buffer, struct radv_image *image, VkImageLayout layout,
                            const VkImageSubresourceLayers *subresource, VkOffset3D offset, VkExtent3D extent);
 
+void radv_meta_bcn_encoded(struct radv_cmd_buffer *cmd_buffer, struct radv_image *image, VkImageLayout layout,
+                           const VkImageSubresourceLayers *subresource, VkOffset3D offset, VkExtent3D extent);
+
 /**
  * Return whether the bound pipeline is the FMASK decompress pass.
  */
diff --git a/src/amd/vulkan/radv_meta_astc_decode.c b/src/amd/vulkan/radv_meta_astc_decode.c
index 93a63692c09..006394c639b 100644
--- a/src/amd/vulkan/radv_meta_astc_decode.c
+++ b/src/amd/vulkan/radv_meta_astc_decode.c
@@ -166,8 +166,15 @@ radv_meta_decode_astc(struct radv_cmd_buffer *cmd_buffer, struct radv_image *ima
    struct radv_image_view src_iview, dst_iview;
    image_view_init(device, image, VK_FORMAT_R32G32B32A32_UINT, VK_IMAGE_ASPECT_COLOR_BIT, subresource->mipLevel,
                    subresource->baseArrayLayer, vk_image_subresource_layer_count(vkImage, subresource), &src_iview);
-   image_view_init(device, image, VK_FORMAT_R8G8B8A8_UINT, VK_IMAGE_ASPECT_PLANE_1_BIT, subresource->mipLevel,
-                   subresource->baseArrayLayer, vk_image_subresource_layer_count(vkImage, subresource), &dst_iview);
+   if (image->isNeedHardEncode) {
+      RADV_FROM_HANDLE(radv_image, store_image, image->staging_images->image_rgba);
+      cmd_buffer->state.flush_bits |= radv_dst_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT, store_image);
+      image_view_init(device, store_image, VK_FORMAT_R8G8B8A8_UINT, VK_IMAGE_ASPECT_COLOR_BIT, subresource->mipLevel,
+         subresource->baseArrayLayer, vk_image_subresource_layer_count(vkImage, subresource), &dst_iview);
+   } else {
+      image_view_init(device, image, VK_FORMAT_R8G8B8A8_UINT, VK_IMAGE_ASPECT_PLANE_1_BIT, subresource->mipLevel,
+         subresource->baseArrayLayer, vk_image_subresource_layer_count(vkImage, subresource), &dst_iview);
+   }
 
    VkExtent3D extent_copy = {
       .width = extent.width,
diff --git a/src/amd/vulkan/radv_meta_copy.c b/src/amd/vulkan/radv_meta_copy.c
index b418d3f2ab9..a0a6b41cce9 100644
--- a/src/amd/vulkan/radv_meta_copy.c
+++ b/src/amd/vulkan/radv_meta_copy.c
@@ -26,6 +26,7 @@
 #include "main/formats.h"
 #include "main/texcompress_etc.h"
 #include "main/texcompress_bptc_tmp.h"
+#include "vk_texcompress_bcn.h"
 
 static VkExtent3D
 meta_image_block_size(const struct radv_image *image)
@@ -471,6 +472,74 @@ copy_buffer_to_image(struct radv_cmd_buffer *cmd_buffer, struct radv_buffer *buf
    radv_meta_restore(&saved_state, cmd_buffer);
 }
 
+void radv_create_staging_images(struct radv_image *dst_image, struct radv_cmd_buffer *cmd_buffer)
+{
+   struct vk_texcompress_bcn_image *staging_images =
+      vk_alloc(&cmd_buffer->device->vk.alloc, sizeof(struct vk_texcompress_bcn_image), 8, VK_SYSTEM_ALLOCATION_SCOPE_COMMAND);
+   dst_image->staging_images = staging_images;
+
+   uint32_t size = 0;
+   VkExtent3D extent = (VkExtent3D){dst_image->info.width, dst_image->info.height, 1};
+   VkFormat format = VK_FORMAT_R8G8B8A8_UNORM;
+   vk_texcompress_create_image(&cmd_buffer->device->vk, NULL, &staging_images->image_mem,
+                               &staging_images->image_rgba, extent, format, 0);
+   RADV_FROM_HANDLE(radv_image, _image_rgba, dst_image->staging_images->image_rgba);
+   size += _image_rgba->size;
+
+   extent = (VkExtent3D){(dst_image->info.width + 3) / 4, (dst_image->info.height + 3) / 4, 1};
+   format = VK_FORMAT_R32G32_UINT;
+   vk_texcompress_create_image(&cmd_buffer->device->vk, NULL, &staging_images->image_mem,
+                               &staging_images->image_bc1, extent, format, 0);
+   RADV_FROM_HANDLE(radv_image, _image_bc1, dst_image->staging_images->image_bc1);
+   size += _image_bc1->size;
+
+   extent = (VkExtent3D){(dst_image->info.width + 3) / 4, (dst_image->info.height + 3) / 4, 1};
+   format = VK_FORMAT_R32G32_UINT;
+   vk_texcompress_create_image(&cmd_buffer->device->vk, NULL, &staging_images->image_mem,
+                               &staging_images->image_bc4, extent, format, 0);
+   RADV_FROM_HANDLE(radv_image, _image_bc4, dst_image->staging_images->image_bc4);
+   size += _image_bc4->size;
+
+   extent = (VkExtent3D){dst_image->info.width, dst_image->info.height, 1};
+   format = VK_FORMAT_R8G8B8A8_UNORM;
+   vk_texcompress_create_image(&cmd_buffer->device->vk, NULL, &staging_images->image_mem,
+      &staging_images->image_dummy, extent, format, size);
+   RADV_FROM_HANDLE(radv_image, _image_dummy, dst_image->staging_images->image_dummy);
+
+   bind_memory(&cmd_buffer->device->vk, &staging_images->image_rgba, &staging_images->image_mem, 0);
+   bind_memory(&cmd_buffer->device->vk, &staging_images->image_bc1, &staging_images->image_mem, _image_rgba->size);
+   bind_memory(&cmd_buffer->device->vk, &staging_images->image_bc4, &staging_images->image_mem, _image_bc1->size);
+
+   cmd_buffer->vk.is_need_hard_encode = true;
+
+   vk_texcompress_insert_head(&cmd_buffer->vk.staging_image_list_head, dst_image->staging_images);
+}
+
+void radv_copy_buffer_hard_encode(struct radv_image *dst_image,
+                                  const enum util_format_layout format_layout,
+                                  struct radv_cmd_buffer *cmd_buffer,
+                                  const VkCopyBufferToImageInfo2 *pCopyBufferToImageInfo)
+{
+   radv_create_staging_images(dst_image, cmd_buffer);
+   for (unsigned r = 0; r < pCopyBufferToImageInfo->regionCount; r++) {
+      if (format_layout == UTIL_FORMAT_LAYOUT_ASTC) {
+         radv_meta_decode_astc(cmd_buffer, dst_image, pCopyBufferToImageInfo->dstImageLayout,
+                               &pCopyBufferToImageInfo->pRegions[r].imageSubresource,
+                               pCopyBufferToImageInfo->pRegions[r].imageOffset,
+                               pCopyBufferToImageInfo->pRegions[r].imageExtent);
+      } else {
+         radv_meta_decode_etc(cmd_buffer, dst_image, pCopyBufferToImageInfo->dstImageLayout,
+                              &pCopyBufferToImageInfo->pRegions[r].imageSubresource,
+                              pCopyBufferToImageInfo->pRegions[r].imageOffset,
+                              pCopyBufferToImageInfo->pRegions[r].imageExtent);
+      }
+      radv_meta_bcn_encoded(cmd_buffer, dst_image, pCopyBufferToImageInfo->dstImageLayout,
+                            &pCopyBufferToImageInfo->pRegions[r].imageSubresource,
+                            pCopyBufferToImageInfo->pRegions[r].imageOffset,
+                            pCopyBufferToImageInfo->pRegions[r].imageExtent);
+   }
+}
+
 extern bool
 radv_is_format_emulated(const struct radv_physical_device *physical_device, VkFormat format);
 
@@ -495,6 +564,12 @@ radv_CmdCopyBufferToImage2(VkCommandBuffer commandBuffer,
          radv_dst_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT, dst_image);
 
       const enum util_format_layout format_layout = vk_format_description(dst_image->vk_format)->layout;
+
+      if (dst_image->isNeedHardEncode) {
+         radv_copy_buffer_hard_encode(dst_image, format_layout, cmd_buffer, pCopyBufferToImageInfo);
+         return;
+      }
+
       for (unsigned r = 0; r < pCopyBufferToImageInfo->regionCount; r++) {
          if (format_layout == UTIL_FORMAT_LAYOUT_ASTC) {
             radv_meta_decode_astc(cmd_buffer, dst_image, pCopyBufferToImageInfo->dstImageLayout,
@@ -847,6 +922,36 @@ copy_image(struct radv_cmd_buffer *cmd_buffer, struct radv_image *src_image,
    radv_meta_restore(&saved_state, cmd_buffer);
 }
 
+void radv_copy_image_hard_encode(struct radv_image *dst_image,
+                                  const enum util_format_layout format_layout,
+                                  struct radv_cmd_buffer *cmd_buffer,
+                                  const VkCopyImageInfo2 *pCopyImageInfo)
+{
+   radv_create_staging_images(dst_image, cmd_buffer);
+   RADV_FROM_HANDLE(radv_image, src_image, pCopyImageInfo->srcImage);
+   for (unsigned r = 0; r < pCopyImageInfo->regionCount; r++) {
+      VkExtent3D dst_extent = pCopyImageInfo->pRegions[r].extent;
+      if (src_image->vk_format != dst_image->vk_format) {
+         dst_extent.width = dst_extent.width / vk_format_get_blockwidth(src_image->vk_format) *
+                            vk_format_get_blockwidth(dst_image->vk_format);
+         dst_extent.height = dst_extent.height / vk_format_get_blockheight(src_image->vk_format) *
+                             vk_format_get_blockheight(dst_image->vk_format);
+      }
+      if (format_layout == UTIL_FORMAT_LAYOUT_ASTC) {
+         radv_meta_decode_astc(cmd_buffer, dst_image, pCopyImageInfo->dstImageLayout,
+                               &pCopyImageInfo->pRegions[r].dstSubresource, pCopyImageInfo->pRegions[r].dstOffset,
+                               dst_extent);
+      } else {
+         radv_meta_decode_etc(cmd_buffer, dst_image, pCopyImageInfo->dstImageLayout,
+                              &pCopyImageInfo->pRegions[r].dstSubresource, pCopyImageInfo->pRegions[r].dstOffset,
+                              dst_extent);
+      }
+      radv_meta_bcn_encoded(cmd_buffer, dst_image, pCopyImageInfo->dstImageLayout,
+                            &pCopyImageInfo->pRegions[r].dstSubresource, pCopyImageInfo->pRegions[r].dstOffset,
+                            dst_extent);
+   }
+}
+
 VKAPI_ATTR void VKAPI_CALL
 radv_CmdCopyImage2(VkCommandBuffer commandBuffer, const VkCopyImageInfo2 *pCopyImageInfo)
 {
@@ -866,6 +971,12 @@ radv_CmdCopyImage2(VkCommandBuffer commandBuffer, const VkCopyImageInfo2 *pCopyI
          radv_dst_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT, dst_image);
 
       const enum util_format_layout format_layout = vk_format_description(dst_image->vk_format)->layout;
+
+      if (dst_image->isNeedHardEncode) {
+         radv_copy_image_hard_encode(dst_image, format_layout, cmd_buffer, pCopyImageInfo);
+         return;
+      }
+
       for (unsigned r = 0; r < pCopyImageInfo->regionCount; r++) {
          VkExtent3D dst_extent = pCopyImageInfo->pRegions[r].extent;
          if (src_image->vk_format != dst_image->vk_format) {
diff --git a/src/amd/vulkan/radv_meta_etc_decode.c b/src/amd/vulkan/radv_meta_etc_decode.c
index d5d002c63a0..3685218d915 100644
--- a/src/amd/vulkan/radv_meta_etc_decode.c
+++ b/src/amd/vulkan/radv_meta_etc_decode.c
@@ -801,23 +801,45 @@ radv_meta_decode_etc(struct radv_cmd_buffer *cmd_buffer, struct radv_image *imag
       store_format = VK_FORMAT_R8G8B8A8_UNORM;
    }
    struct radv_image_view dest_iview;
-   radv_image_view_init(
-      &dest_iview, cmd_buffer->device,
-      &(VkImageViewCreateInfo){
-         .sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO,
-         .image = radv_image_to_handle(image),
-         .viewType = radv_meta_get_view_type(image),
-         .format = store_format,
-         .subresourceRange =
-            {
-               .aspectMask = VK_IMAGE_ASPECT_PLANE_1_BIT,
-               .baseMipLevel = subresource->mipLevel,
-               .levelCount = 1,
-               .baseArrayLayer = 0,
-               .layerCount = subresource->baseArrayLayer + subresource->layerCount,
-            },
-      },
-      NULL);
+   if (image->isNeedHardEncode) {
+      RADV_FROM_HANDLE(radv_image, store_image, image->staging_images->image_rgba);
+      cmd_buffer->state.flush_bits |= radv_dst_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT, store_image);
+      radv_image_view_init(
+         &dest_iview, cmd_buffer->device,
+         &(VkImageViewCreateInfo){
+            .sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO,
+            .image = radv_image_to_handle(store_image),
+            .viewType = radv_meta_get_view_type(store_image),
+            .format = store_format,
+            .subresourceRange =
+               {
+                  .aspectMask = VK_IMAGE_ASPECT_COLOR_BIT,
+                  .baseMipLevel = subresource->mipLevel,
+                  .levelCount = 1,
+                  .baseArrayLayer = 0,
+                  .layerCount = subresource->baseArrayLayer + subresource->layerCount,
+               },
+         },
+         NULL);
+   } else {
+      radv_image_view_init(
+         &dest_iview, cmd_buffer->device,
+         &(VkImageViewCreateInfo){
+            .sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO,
+            .image = radv_image_to_handle(image),
+            .viewType = radv_meta_get_view_type(image),
+            .format = store_format,
+            .subresourceRange =
+               {
+                  .aspectMask = VK_IMAGE_ASPECT_PLANE_1_BIT,
+                  .baseMipLevel = subresource->mipLevel,
+                  .levelCount = 1,
+                  .baseArrayLayer = 0,
+                  .layerCount = subresource->baseArrayLayer + subresource->layerCount,
+               },
+         },
+         NULL);
+   }
 
    decode_etc(cmd_buffer, &src_iview, &dest_iview, &(VkOffset3D){offset.x, offset.y, base_slice},
               &(VkExtent3D){extent.width, extent.height, slice_count});
diff --git a/src/amd/vulkan/radv_meta_rgba_encode.c b/src/amd/vulkan/radv_meta_rgba_encode.c
new file mode 100644
index 00000000000..b144767143d
--- /dev/null
+++ b/src/amd/vulkan/radv_meta_rgba_encode.c
@@ -0,0 +1,312 @@
+/* Copyright (c) 2017-2023 Hans-Kristian Arntzen
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <assert.h>
+#include <stdbool.h>
+
+#include "radv_meta.h"
+#include "sid.h"
+#include "vk_common_entrypoints.h"
+#include "vk_format.h"
+#include "vk_texcompress_bcn.h"
+#include "radv_debug.h"
+
+struct radv_dispatch_info {
+   /**
+    * Determine the layout of the grid (in block units) to be used.
+    */
+   uint32_t blocks[3];
+
+   /**
+    * A starting offset for the grid. If unaligned is set, the offset
+    * must still be aligned.
+    */
+   uint32_t offsets[3];
+
+   /**
+    * Whether it's an unaligned compute dispatch.
+    */
+   bool unaligned;
+
+   /**
+    * Whether waves must be launched in order.
+    */
+   bool ordered;
+
+   /**
+    * Indirect compute parameters resource.
+    */
+   struct radeon_winsys_bo *indirect;
+   uint64_t va;
+};
+
+VkResult
+radv_device_init_meta_rgba_encode_state(struct radv_device *device)
+{
+   const struct radv_physical_device *pdev = device->physical_device;
+   struct radv_meta_state *state = &device->meta_state;
+
+   return vk_texcompress_bcn_init(&device->vk, &state->alloc, radv_pipeline_cache_to_handle(&state->cache), state->bcn_encoded);
+}
+
+void
+radv_device_finish_meta_rgba_encode_state(struct radv_device *device)
+{
+   struct radv_meta_state *state = &device->meta_state;
+   struct vk_texcompress_bcn_state **bcns = state->bcn_encoded;
+
+   if (bcns)
+      vk_texcompress_bcn_finish(&device->vk, &state->alloc, state->bcn_encoded);
+}
+
+extern void
+radv_compute_dispatch(struct radv_cmd_buffer *cmd_buffer, const struct radv_dispatch_info *info);
+
+static VkImageViewType
+get_view_type(const struct radv_image *image)
+{
+   switch (image->type) {
+   case VK_IMAGE_TYPE_2D:
+      return VK_IMAGE_VIEW_TYPE_2D_ARRAY;
+   case VK_IMAGE_TYPE_3D:
+      return VK_IMAGE_VIEW_TYPE_3D;
+   default:
+      unreachable("bad VkImageViewType");
+   }
+}
+
+static void
+image_view_init(struct radv_device *device, struct radv_image *image, VkFormat format, VkImageAspectFlags aspectMask,
+                uint32_t baseMipLevel, uint32_t baseArrayLayer, uint32_t layerCount, VkComponentMapping *component,
+                struct radv_image_view *iview)
+{
+   VkImageViewCreateInfo iview_create_info = {
+      .sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO,
+      .image = radv_image_to_handle(image),
+      .viewType = get_view_type(image),
+      .format = format,
+      .subresourceRange =
+         {
+            .aspectMask = aspectMask,
+            .baseMipLevel = baseMipLevel,
+            .levelCount = 1,
+            .baseArrayLayer = 0,
+            .layerCount = baseArrayLayer + layerCount,
+         },
+   };
+   if (component) {
+      iview_create_info.components = *component;
+   }
+
+   radv_image_view_init(iview, device, &iview_create_info, NULL);
+}
+
+static void
+encode_bc1(struct radv_cmd_buffer *cmd_buffer, VkImageLayout layout,
+           const VkExtent3D *extent, const VkImageSubresourceLayers *subresource,
+           struct radv_image *src_image, struct radv_image *dst_image)
+{
+   struct radv_device *device = cmd_buffer->device;
+   struct radv_meta_state *state = &device->meta_state;
+
+   struct radv_image_view src_iview, dst_iview;
+   VkComponentMapping component = {
+      .r = VK_COMPONENT_SWIZZLE_R,
+      .g = VK_COMPONENT_SWIZZLE_G,
+      .b = VK_COMPONENT_SWIZZLE_B,
+      .a = VK_COMPONENT_SWIZZLE_A
+   };
+   image_view_init(device, src_image, VK_FORMAT_R8G8B8A8_UNORM, VK_IMAGE_ASPECT_COLOR_BIT, subresource->mipLevel,
+                   subresource->baseArrayLayer, 1, &component, &src_iview);
+   image_view_init(device, dst_image, VK_FORMAT_R16G16B16A16_UINT, VK_IMAGE_ASPECT_COLOR_BIT, subresource->mipLevel,
+                   subresource->baseArrayLayer, 1, NULL, &dst_iview);
+
+   cmd_buffer->state.flush_bits |=
+      RADV_CMD_FLAG_CS_PARTIAL_FLUSH | RADV_CMD_FLAG_PS_PARTIAL_FLUSH |
+      radv_src_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_WRITE_BIT, src_iview.image) |
+      radv_dst_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT, dst_iview.image);
+
+   struct vk_texcompress_bcn_write_descriptor_set write_desc_set;
+   vk_texcompress_bc1_fill_write_descriptor_sets(state->bcn_encoded[BC1_ENCODE], &write_desc_set,
+                                                  radv_image_view_to_handle(&src_iview), layout,
+                                                  radv_image_view_to_handle(&dst_iview));
+
+   radv_meta_push_descriptor_set(cmd_buffer, VK_PIPELINE_BIND_POINT_COMPUTE, state->bcn_encoded[BC1_ENCODE]->p_layout,
+                                 0, /* set number */
+                                 3, write_desc_set.descriptor_set);
+
+   VkPipeline pipeline =
+      vk_texcompress_rgba_get_encode_pipeline(&device->vk, &state->alloc,
+                                              state->bcn_encoded[BC1_ENCODE],
+                                              radv_pipeline_cache_to_handle(&state->cache), BC1_ENCODE);
+   if (pipeline == VK_NULL_HANDLE)
+      return;
+
+   radv_CmdBindPipeline(radv_cmd_buffer_to_handle(cmd_buffer), VK_PIPELINE_BIND_POINT_COMPUTE, pipeline);
+
+   const unsigned num_refinements = 1;
+   const unsigned push_constants[2] = {num_refinements};
+   radv_CmdPushConstants(radv_cmd_buffer_to_handle(cmd_buffer), state->bcn_encoded[BC1_ENCODE]->p_layout,
+                         VK_SHADER_STAGE_COMPUTE_BIT, 0, sizeof(push_constants), &push_constants);
+
+   struct radv_dispatch_info info = {
+      .blocks[0] = DIV_ROUND_UP(extent->width, 32),
+      .blocks[1] = DIV_ROUND_UP(extent->height, 32),
+      .blocks[2] = 1,
+   };
+   radv_compute_dispatch(cmd_buffer, &info);
+   radv_image_view_finish(&src_iview);
+   radv_image_view_finish(&dst_iview);
+}
+
+static void
+encode_bc4(struct radv_cmd_buffer *cmd_buffer, VkImageLayout layout,
+           const VkExtent3D *extent, const VkImageSubresourceLayers *subresource,
+           struct radv_image *src_image, struct radv_image *dst_image)
+{
+   struct radv_device *device = cmd_buffer->device;
+   struct radv_meta_state *state = &device->meta_state;
+
+   struct radv_image_view src_iview, dst_iview;
+   VkComponentMapping component = {
+      .r = VK_COMPONENT_SWIZZLE_A,
+      .g = VK_COMPONENT_SWIZZLE_ZERO,
+      .b = VK_COMPONENT_SWIZZLE_ZERO,
+      .a = VK_COMPONENT_SWIZZLE_ONE
+   };
+   image_view_init(device, src_image, VK_FORMAT_R8G8B8A8_UNORM, VK_IMAGE_ASPECT_COLOR_BIT, subresource->mipLevel,
+                   subresource->baseArrayLayer, 1, &component, &src_iview);
+   image_view_init(device, dst_image, VK_FORMAT_R16G16B16A16_UINT, VK_IMAGE_ASPECT_COLOR_BIT, subresource->mipLevel,
+                   subresource->baseArrayLayer, 1, NULL, &dst_iview);
+
+   cmd_buffer->state.flush_bits |=
+      RADV_CMD_FLAG_CS_PARTIAL_FLUSH | RADV_CMD_FLAG_PS_PARTIAL_FLUSH |
+      radv_src_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_WRITE_BIT, src_iview.image) |
+      radv_dst_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT, dst_iview.image);
+
+   struct vk_texcompress_bc4_write_descriptor_set write_desc_set;
+   vk_texcompress_bc4_fill_write_descriptor_sets(state->bcn_encoded[BC4_ENCODE], &write_desc_set,
+                                                  radv_image_view_to_handle(&src_iview), layout,
+                                                  radv_image_view_to_handle(&dst_iview));
+
+   radv_meta_push_descriptor_set(cmd_buffer, VK_PIPELINE_BIND_POINT_COMPUTE, state->bcn_encoded[BC4_ENCODE]->p_layout,
+                                 0, /* set number */
+                                 2, write_desc_set.descriptor_set);
+
+   VkPipeline pipeline =
+      vk_texcompress_rgba_get_encode_pipeline(&device->vk, &state->alloc,
+                                              state->bcn_encoded[BC4_ENCODE],
+                                              radv_pipeline_cache_to_handle(&state->cache), BC4_ENCODE);
+   if (pipeline == VK_NULL_HANDLE)
+      return;
+
+   radv_CmdBindPipeline(radv_cmd_buffer_to_handle(cmd_buffer), VK_PIPELINE_BIND_POINT_COMPUTE, pipeline);
+
+   const unsigned alpha_channel_id = 0;  // r通道
+   const unsigned use_norm = 0;
+   const unsigned push_constants[2] = {alpha_channel_id, use_norm};
+   radv_CmdPushConstants(radv_cmd_buffer_to_handle(cmd_buffer), state->bcn_encoded[BC4_ENCODE]->p_layout,
+                         VK_SHADER_STAGE_COMPUTE_BIT, 0, sizeof(push_constants), &push_constants);
+
+   struct radv_dispatch_info info = {
+      .blocks[0] = 1,
+      .blocks[1] = DIV_ROUND_UP(extent->width, 16),
+      .blocks[2] = DIV_ROUND_UP(extent->height, 16),
+   };
+   radv_compute_dispatch(cmd_buffer, &info);
+   radv_image_view_finish(&src_iview);
+   radv_image_view_finish(&dst_iview);
+}
+
+static void
+encode_bc3(struct radv_cmd_buffer *cmd_buffer,
+           const VkExtent3D *extent, const VkImageSubresourceLayers *subresource,
+           struct radv_image *bc1_image, struct radv_image *bc4_image, struct radv_image *dst_image)
+{
+   struct radv_device *device = cmd_buffer->device;
+   struct radv_meta_state *state = &device->meta_state;
+
+   struct radv_image_view bc1_iview, bc4_iview, dst_iview;
+   VkComponentMapping component = {
+      .r = VK_COMPONENT_SWIZZLE_R,
+      .g = VK_COMPONENT_SWIZZLE_G,
+      .b = VK_COMPONENT_SWIZZLE_ZERO,
+      .a = VK_COMPONENT_SWIZZLE_ONE
+   };
+   image_view_init(device, bc1_image, VK_FORMAT_R32G32_UINT, VK_IMAGE_ASPECT_COLOR_BIT, subresource->mipLevel,
+                   subresource->baseArrayLayer, 1, &component, &bc1_iview);
+   image_view_init(device, bc4_image, VK_FORMAT_R32G32_UINT, VK_IMAGE_ASPECT_COLOR_BIT, subresource->mipLevel,
+                   subresource->baseArrayLayer, 1, &component, &bc4_iview);
+   image_view_init(device, dst_image, VK_FORMAT_R32G32B32A32_UINT, VK_IMAGE_ASPECT_PLANE_1_BIT, subresource->mipLevel,
+                   subresource->baseArrayLayer, 1, NULL, &dst_iview);
+
+   cmd_buffer->state.flush_bits |=
+      RADV_CMD_FLAG_CS_PARTIAL_FLUSH | RADV_CMD_FLAG_PS_PARTIAL_FLUSH |
+      radv_src_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_WRITE_BIT, bc1_iview.image) |
+      radv_src_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_WRITE_BIT, bc4_iview.image) |
+      radv_dst_access_flush(cmd_buffer, VK_ACCESS_TRANSFER_READ_BIT | VK_ACCESS_TRANSFER_WRITE_BIT, dst_iview.image);
+
+   struct vk_texcompress_bcn_write_descriptor_set write_desc_set;
+   vk_texcompress_bc3_fill_write_descriptor_sets(state->bcn_encoded[BC3_ENCODE], &write_desc_set,
+                                                 radv_image_view_to_handle(&bc1_iview), radv_image_view_to_handle(&bc4_iview),
+                                                 radv_image_view_to_handle(&dst_iview));
+
+   radv_meta_push_descriptor_set(cmd_buffer, VK_PIPELINE_BIND_POINT_COMPUTE, state->bcn_encoded[BC3_ENCODE]->p_layout,
+                                 0, /* set number */
+                                 3, write_desc_set.descriptor_set);
+
+   VkPipeline pipeline =
+      vk_texcompress_rgba_get_encode_pipeline(&device->vk, &state->alloc,
+                                              state->bcn_encoded[BC3_ENCODE],
+                                              radv_pipeline_cache_to_handle(&state->cache), BC3_ENCODE);
+   if (pipeline == VK_NULL_HANDLE)
+      return;
+
+   radv_CmdBindPipeline(radv_cmd_buffer_to_handle(cmd_buffer), VK_PIPELINE_BIND_POINT_COMPUTE, pipeline);
+
+   struct radv_dispatch_info info = {
+      .blocks[0] = DIV_ROUND_UP(extent->width, 32),
+      .blocks[1] = DIV_ROUND_UP(extent->height, 32),
+      .blocks[2] = 1,
+   };
+   radv_compute_dispatch(cmd_buffer, &info);
+   radv_image_view_finish(&bc1_iview);
+   radv_image_view_finish(&bc4_iview);
+   radv_image_view_finish(&dst_iview);
+}
+
+void
+radv_meta_bcn_encoded(struct radv_cmd_buffer *cmd_buffer, struct radv_image *image, VkImageLayout layout,
+                      const VkImageSubresourceLayers *subresource, VkOffset3D offset, VkExtent3D extent)
+{
+   struct radv_device *device = cmd_buffer->device;
+   struct radv_meta_saved_state saved_state;
+   radv_meta_save(&saved_state, cmd_buffer,
+                  RADV_META_SAVE_COMPUTE_PIPELINE | RADV_META_SAVE_CONSTANTS | RADV_META_SAVE_DESCRIPTORS);
+
+   VkImage _image = radv_image_to_handle(image);
+   VK_FROM_HANDLE(vk_image, vkImage, _image);
+
+   extent = radv_sanitize_image_extent(image->type, extent);
+   offset = radv_sanitize_image_offset(image->type, offset);
+
+   VkExtent3D extent_copy = {
+      .width = extent.width,
+      .height = extent.height,
+      .depth = 1,
+   };
+
+   RADV_FROM_HANDLE(radv_image, rgba_image, image->staging_images->image_rgba);
+   RADV_FROM_HANDLE(radv_image, image1, image->staging_images->image_bc1);
+   RADV_FROM_HANDLE(radv_image, image4, image->staging_images->image_bc4);
+
+   encode_bc1(cmd_buffer, layout, &extent_copy, subresource, rgba_image, image1);
+
+   encode_bc4(cmd_buffer, layout, &extent_copy, subresource, rgba_image, image4);
+
+   encode_bc3(cmd_buffer, &extent_copy, subresource, image1, image4, image);
+
+   radv_meta_restore(&saved_state, cmd_buffer);
+}
diff --git a/src/amd/vulkan/radv_private.h b/src/amd/vulkan/radv_private.h
index 472858dd401..04a670d714d 100644
--- a/src/amd/vulkan/radv_private.h
+++ b/src/amd/vulkan/radv_private.h
@@ -98,6 +98,7 @@ typedef uint32_t xcb_window_t;
 
 #include "vk_texcompress_astc.h"
 #include "wsi_common.h"
+#include "vk_texcompress_bcn.h"
 
 #ifdef __cplusplus
 extern "C"
@@ -681,6 +682,8 @@ struct radv_meta_state {
    } etc_decode;
 
    struct vk_texcompress_astc_state *astc_decode;
+
+   struct vk_texcompress_bcn_state *bcn_encoded[VK_TEXCOMPRESS_BCN_NUM_COMPUTE_PIPELINES];
 };
 
 #define RADV_NUM_HW_CTX (RADEON_CTX_PRIORITY_REALTIME + 1)
@@ -2090,8 +2093,12 @@ struct radv_image {
    bool dcc_sign_reinterpret;
    bool support_comp_to_single;
 
-   bool isNeedSoftEncode;// 是否需要使用 软编
-   bool isNeedSoftDecode;// 是否需要使用 软解
+   bool isNeedSoftEncode;  // 是否需要使用 软编
+   bool isNeedSoftDecode;  // 是否需要使用 软解
+
+   bool isNeedHardEncode;  // 是否需要使用 硬编
+   struct vk_texcompress_bcn_image *staging_images;  // 硬编的中间资源
+
 	VkFormat srcFormat; //用于保存纹理的原始格式
 	VkFormat unpackETCFormat;//用于保存ETC纹理解压后的格式
 
@@ -2474,6 +2481,8 @@ void radv_image_view_init(struct radv_image_view *view, struct radv_device *devi
                           const struct radv_image_view_extra_create_info *extra_create_info);
 void radv_image_view_finish(struct radv_image_view *iview);
 
+void radv_internal_image_finish(struct radv_image *image);
+
 VkFormat radv_get_aspect_format(struct radv_image *image, VkImageAspectFlags mask);
 
 struct radv_sampler_ycbcr_conversion_state {
diff --git a/src/compiler/glsl/bc1.glsl b/src/compiler/glsl/bc1.glsl
index 251635d7b69..627b2e5419c 100644
--- a/src/compiler/glsl/bc1.glsl
+++ b/src/compiler/glsl/bc1.glsl
@@ -1,3 +1,4 @@
+#version 310 es
 /*
  * Copyright 2020-2022 Matias N. Goldberg
  * Copyright 2022 Intel Corporation
@@ -21,7 +22,6 @@
  * DEALINGS IN THE SOFTWARE.
  */
 
-#version 310 es
 
 #if defined(GL_ES) && GL_ES == 1
 	// Desktop GLSL allows the const keyword for either compile-time or
@@ -31,10 +31,31 @@
 	#define const
 #endif
 
-%s // include "CrossPlatformSettings_piece_all.glsl"
-
 #define FLT_MAX 340282346638528859811704183484516925440.0f
 
+#ifdef VULKAN
+
+#extension GL_GOOGLE_include_directive : require
+#include "CrossPlatformSettings_piece_all.glsl"
+
+layout(push_constant, std430) uniform pc {
+	uint p_numRefinements;
+};
+
+layout ( set = 0, binding = 0 ) uniform sampler2D srcTex;
+
+layout (std430, set = 0, binding = 1) readonly restrict buffer globalBuffer
+{
+	float2 c_oMatch5[256];
+	float2 c_oMatch6[256];
+};
+
+layout (rgba16ui, set = 0, binding = 2 ) uniform restrict writeonly mediump uimage2D dstTexture;
+
+#else
+
+%s // include "CrossPlatformSettings_piece_all.glsl"
+
 layout( location = 0 ) uniform uint p_numRefinements;
 
 uniform sampler2D srcTex;
@@ -47,6 +68,8 @@ layout( std430, binding = 1 ) readonly restrict buffer globalBuffer
 	float2 c_oMatch6[256];
 };
 
+#endif
+
 layout( local_size_x = 8,  //
 		local_size_y = 8,  //
 		local_size_z = 1 ) in;
diff --git a/src/compiler/glsl/bc4.glsl b/src/compiler/glsl/bc4.glsl
index 2486fa4ae0c..0b9d236d624 100644
--- a/src/compiler/glsl/bc4.glsl
+++ b/src/compiler/glsl/bc4.glsl
@@ -1,3 +1,5 @@
+#version 310 es
+
 /*
  * Copyright 2020-2022 Matias N. Goldberg
  * Copyright 2022 Intel Corporation
@@ -21,8 +23,6 @@
  * DEALINGS IN THE SOFTWARE.
  */
 
-#version 310 es
-
 #if defined(GL_ES) && GL_ES == 1
 	// Desktop GLSL allows the const keyword for either compile-time or
 	// run-time constants. GLSL ES only allows the keyword for compile-time
@@ -33,20 +33,37 @@
 
 #define __sharedOnlyBarrier memoryBarrierShared();barrier();
 
-%s // include "CrossPlatformSettings_piece_all.glsl"
+#ifdef VULKAN
 
-shared float2 g_minMaxValues[4u * 4u * 4u];
-shared uint2 g_mask[4u * 4u];
+#extension GL_GOOGLE_include_directive : require
+#include "CrossPlatformSettings_piece_all.glsl"
 
-layout( location = 0 ) uniform uint2 params;
+layout(push_constant, std430) uniform pc {
+	uint2 params;
+};
 
-#define p_channelIdx params.x
-#define p_useSNorm params.y
+layout ( set = 0, binding = 0 ) uniform sampler2D srcTex;
+
+layout (rgba16ui, set = 0, binding = 1) uniform restrict writeonly mediump uimage2D dstTexture;
+
+#else
+
+%s // include "CrossPlatformSettings_piece_all.glsl"
 
 uniform sampler2D srcTex;
 
 layout( rgba16ui ) uniform restrict writeonly mediump uimage2D dstTexture;
 
+layout( location = 0 ) uniform uint2 params;
+
+#endif
+
+shared float2 g_minMaxValues[4u * 4u * 4u];
+shared uint2 g_mask[4u * 4u];
+
+#define p_channelIdx params.x
+#define p_useSNorm params.y
+
 layout( local_size_x = 4,  //
 		local_size_y = 4,  //
 		local_size_z = 4 ) in;
diff --git a/src/compiler/glsl/etc2_rgba_stitch.glsl b/src/compiler/glsl/etc2_rgba_stitch.glsl
index f90accb82e1..8d50c469956 100644
--- a/src/compiler/glsl/etc2_rgba_stitch.glsl
+++ b/src/compiler/glsl/etc2_rgba_stitch.glsl
@@ -1,3 +1,5 @@
+#version 310 es
+
 /*
  * Copyright 2020-2022 Matias N. Goldberg
  * Copyright 2022 Intel Corporation
@@ -25,18 +27,30 @@
 // This compute shader merely stitches them together to form the final result
 // It's also used by RG11 driver to stitch two R11 into one RG11
 
-#version 310 es
+#ifdef VULKAN
 
-%s // include "CrossPlatformSettings_piece_all.glsl"
+#extension GL_GOOGLE_include_directive : require
+#include "CrossPlatformSettings_piece_all.glsl"
 
-layout( local_size_x = 8,  //
-		local_size_y = 8,  //
-		local_size_z = 1 ) in;
+layout ( binding = 0 ) uniform highp usampler2D srcRGB;
+layout ( binding = 1 ) uniform highp usampler2D srcAlpha;
+
+layout ( rgba32ui, binding = 2 ) uniform restrict writeonly mediump uimage2D dstTexture;
+
+#else
+
+%s // include "CrossPlatformSettings_piece_all.glsl"
 
 layout( binding = 0 ) uniform highp usampler2D srcRGB;
 layout( binding = 1 ) uniform highp usampler2D srcAlpha;
 layout( rgba32ui ) uniform restrict writeonly highp uimage2D dstTexture;
 
+#endif
+
+layout( local_size_x = 8,  //
+		local_size_y = 8,  //
+		local_size_z = 1 ) in;
+
 void main()
 {
 	uint2 etcRgb = OGRE_Load2D( srcRGB, int2( gl_GlobalInvocationID.xy ), 0 ).xy;
diff --git a/src/compiler/meson.build b/src/compiler/meson.build
index e18af3bf335..85b19b84bec 100644
--- a/src/compiler/meson.build
+++ b/src/compiler/meson.build
@@ -24,6 +24,14 @@ inc_spirv = include_directories('spirv')
 
 astc_decoder_glsl_file = files('glsl/astc_decoder.glsl')
 
+bc3_encode_head_dir = meson.current_source_dir() / 'glsl'
+
+bc1_glsl_file = files('glsl/bc1.glsl')
+
+bc4_glsl_file = files('glsl/bc4.glsl')
+
+etc2_rgba_stitch_glsl_file = files('glsl/etc2_rgba_stitch.glsl')
+
 files_libcompiler = files(
   'builtin_type_macros.h',
   'glsl_types.cpp',
diff --git a/src/util/bc1_tables.h b/src/util/bc1_tables.h
new file mode 100644
index 00000000000..654ccdadb78
--- /dev/null
+++ b/src/util/bc1_tables.h
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2020-2022 Matias N. Goldberg
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+/* These tables have been generated with:
+
+		#define STB_DXT_IMPLEMENTATION
+		#include "stb_dxt.h"
+		#include <stdio.h>
+
+		int main()
+		{
+			stb__InitDXT();
+
+			printf( "static unsigned char stb__OMatch5[256][2] = {\n" );
+			for( size_t i = 0u; i < 256u; ++i )
+			{
+				printf( "{ %i, %i }", stb__OMatch5[i][0], stb__OMatch5[i][1] );
+				if( i != 255u )
+					printf( ",\n" );
+			}
+			printf( "\n};\n" );
+
+			printf( "static unsigned char stb__OMatch6[256][2] = {\n" );
+			for( size_t i = 0u; i < 256u; ++i )
+			{
+				printf( "{ %i, %i }", stb__OMatch6[i][0], stb__OMatch6[i][1] );
+				if( i != 255u )
+					printf( ",\n" );
+			}
+			printf( "\n};\n" );
+
+			return 0;
+		}
+
+	Note that stb__OMatch5[i][0] = max and stb__OMatch5[i][1] = min
+*/
+static unsigned char stb__OMatch5[256][2] = {
+	{ 0, 0 },   { 0, 0 },   { 0, 1 },   { 0, 1 },   { 1, 0 },   { 1, 0 },   { 1, 0 },   { 1, 1 },
+	{ 1, 1 },   { 2, 0 },   { 2, 0 },   { 0, 4 },   { 2, 1 },   { 2, 1 },   { 2, 1 },   { 3, 0 },
+	{ 3, 0 },   { 3, 0 },   { 3, 1 },   { 1, 5 },   { 3, 2 },   { 3, 2 },   { 4, 0 },   { 4, 0 },
+	{ 4, 1 },   { 4, 1 },   { 4, 2 },   { 4, 2 },   { 4, 2 },   { 3, 5 },   { 5, 1 },   { 5, 1 },
+	{ 5, 2 },   { 4, 4 },   { 5, 3 },   { 5, 3 },   { 5, 3 },   { 6, 2 },   { 6, 2 },   { 6, 2 },
+	{ 6, 3 },   { 5, 5 },   { 6, 4 },   { 6, 4 },   { 4, 8 },   { 7, 3 },   { 7, 3 },   { 7, 3 },
+	{ 7, 4 },   { 7, 4 },   { 7, 4 },   { 7, 5 },   { 5, 9 },   { 7, 6 },   { 7, 6 },   { 8, 4 },
+	{ 8, 4 },   { 8, 5 },   { 8, 5 },   { 8, 6 },   { 8, 6 },   { 8, 6 },   { 7, 9 },   { 9, 5 },
+	{ 9, 5 },   { 9, 6 },   { 8, 8 },   { 9, 7 },   { 9, 7 },   { 9, 7 },   { 10, 6 },  { 10, 6 },
+	{ 10, 6 },  { 10, 7 },  { 9, 9 },   { 10, 8 },  { 10, 8 },  { 8, 12 },  { 11, 7 },  { 11, 7 },
+	{ 11, 7 },  { 11, 8 },  { 11, 8 },  { 11, 8 },  { 11, 9 },  { 9, 13 },  { 11, 10 }, { 11, 10 },
+	{ 12, 8 },  { 12, 8 },  { 12, 9 },  { 12, 9 },  { 12, 10 }, { 12, 10 }, { 12, 10 }, { 11, 13 },
+	{ 13, 9 },  { 13, 9 },  { 13, 10 }, { 12, 12 }, { 13, 11 }, { 13, 11 }, { 13, 11 }, { 14, 10 },
+	{ 14, 10 }, { 14, 10 }, { 14, 11 }, { 13, 13 }, { 14, 12 }, { 14, 12 }, { 12, 16 }, { 15, 11 },
+	{ 15, 11 }, { 15, 11 }, { 15, 12 }, { 15, 12 }, { 15, 12 }, { 15, 13 }, { 13, 17 }, { 15, 14 },
+	{ 15, 14 }, { 16, 12 }, { 16, 12 }, { 16, 13 }, { 16, 13 }, { 16, 14 }, { 16, 14 }, { 16, 14 },
+	{ 15, 17 }, { 17, 13 }, { 17, 13 }, { 17, 14 }, { 16, 16 }, { 17, 15 }, { 17, 15 }, { 17, 15 },
+	{ 18, 14 }, { 18, 14 }, { 18, 14 }, { 18, 15 }, { 17, 17 }, { 18, 16 }, { 18, 16 }, { 16, 20 },
+	{ 19, 15 }, { 19, 15 }, { 19, 15 }, { 19, 16 }, { 19, 16 }, { 19, 16 }, { 19, 17 }, { 17, 21 },
+	{ 19, 18 }, { 19, 18 }, { 20, 16 }, { 20, 16 }, { 20, 17 }, { 20, 17 }, { 20, 18 }, { 20, 18 },
+	{ 20, 18 }, { 19, 21 }, { 21, 17 }, { 21, 17 }, { 21, 18 }, { 20, 20 }, { 21, 19 }, { 21, 19 },
+	{ 21, 19 }, { 22, 18 }, { 22, 18 }, { 22, 18 }, { 22, 19 }, { 21, 21 }, { 22, 20 }, { 22, 20 },
+	{ 20, 24 }, { 23, 19 }, { 23, 19 }, { 23, 19 }, { 23, 20 }, { 23, 20 }, { 23, 20 }, { 23, 21 },
+	{ 21, 25 }, { 23, 22 }, { 23, 22 }, { 24, 20 }, { 24, 20 }, { 24, 21 }, { 24, 21 }, { 24, 22 },
+	{ 24, 22 }, { 24, 22 }, { 23, 25 }, { 25, 21 }, { 25, 21 }, { 25, 22 }, { 24, 24 }, { 25, 23 },
+	{ 25, 23 }, { 25, 23 }, { 26, 22 }, { 26, 22 }, { 26, 22 }, { 26, 23 }, { 25, 25 }, { 26, 24 },
+	{ 26, 24 }, { 24, 28 }, { 27, 23 }, { 27, 23 }, { 27, 23 }, { 27, 24 }, { 27, 24 }, { 27, 24 },
+	{ 27, 25 }, { 25, 29 }, { 27, 26 }, { 27, 26 }, { 28, 24 }, { 28, 24 }, { 28, 25 }, { 28, 25 },
+	{ 28, 26 }, { 28, 26 }, { 28, 26 }, { 27, 29 }, { 29, 25 }, { 29, 25 }, { 29, 26 }, { 28, 28 },
+	{ 29, 27 }, { 29, 27 }, { 29, 27 }, { 30, 26 }, { 30, 26 }, { 30, 26 }, { 30, 27 }, { 29, 29 },
+	{ 30, 28 }, { 30, 28 }, { 30, 28 }, { 31, 27 }, { 31, 27 }, { 31, 27 }, { 31, 28 }, { 31, 28 },
+	{ 31, 28 }, { 31, 29 }, { 31, 29 }, { 31, 30 }, { 31, 30 }, { 31, 30 }, { 31, 31 }, { 31, 31 }
+};
+
+static unsigned char stb__OMatch6[256][2] = {
+	{ 0, 0 },   { 0, 1 },   { 1, 0 },   { 1, 0 },   { 1, 1 },   { 2, 0 },   { 2, 1 },   { 3, 0 },
+	{ 3, 0 },   { 3, 1 },   { 4, 0 },   { 4, 0 },   { 4, 1 },   { 5, 0 },   { 5, 1 },   { 6, 0 },
+	{ 6, 0 },   { 6, 1 },   { 7, 0 },   { 7, 0 },   { 7, 1 },   { 8, 0 },   { 8, 1 },   { 8, 1 },
+	{ 8, 2 },   { 9, 1 },   { 9, 2 },   { 9, 2 },   { 9, 3 },   { 10, 2 },  { 10, 3 },  { 10, 3 },
+	{ 10, 4 },  { 11, 3 },  { 11, 4 },  { 11, 4 },  { 11, 5 },  { 12, 4 },  { 12, 5 },  { 12, 5 },
+	{ 12, 6 },  { 13, 5 },  { 13, 6 },  { 8, 16 },  { 13, 7 },  { 14, 6 },  { 14, 7 },  { 9, 17 },
+	{ 14, 8 },  { 15, 7 },  { 15, 8 },  { 11, 16 }, { 15, 9 },  { 15, 10 }, { 16, 8 },  { 16, 9 },
+	{ 16, 10 }, { 15, 13 }, { 17, 9 },  { 17, 10 }, { 17, 11 }, { 15, 16 }, { 18, 10 }, { 18, 11 },
+	{ 18, 12 }, { 16, 16 }, { 19, 11 }, { 19, 12 }, { 19, 13 }, { 17, 17 }, { 20, 12 }, { 20, 13 },
+	{ 20, 14 }, { 19, 16 }, { 21, 13 }, { 21, 14 }, { 21, 15 }, { 20, 17 }, { 22, 14 }, { 22, 15 },
+	{ 25, 10 }, { 22, 16 }, { 23, 15 }, { 23, 16 }, { 26, 11 }, { 23, 17 }, { 24, 16 }, { 24, 17 },
+	{ 27, 12 }, { 24, 18 }, { 25, 17 }, { 25, 18 }, { 28, 13 }, { 25, 19 }, { 26, 18 }, { 26, 19 },
+	{ 29, 14 }, { 26, 20 }, { 27, 19 }, { 27, 20 }, { 30, 15 }, { 27, 21 }, { 28, 20 }, { 28, 21 },
+	{ 28, 21 }, { 28, 22 }, { 29, 21 }, { 29, 22 }, { 24, 32 }, { 29, 23 }, { 30, 22 }, { 30, 23 },
+	{ 25, 33 }, { 30, 24 }, { 31, 23 }, { 31, 24 }, { 27, 32 }, { 31, 25 }, { 31, 26 }, { 32, 24 },
+	{ 32, 25 }, { 32, 26 }, { 31, 29 }, { 33, 25 }, { 33, 26 }, { 33, 27 }, { 31, 32 }, { 34, 26 },
+	{ 34, 27 }, { 34, 28 }, { 32, 32 }, { 35, 27 }, { 35, 28 }, { 35, 29 }, { 33, 33 }, { 36, 28 },
+	{ 36, 29 }, { 36, 30 }, { 35, 32 }, { 37, 29 }, { 37, 30 }, { 37, 31 }, { 36, 33 }, { 38, 30 },
+	{ 38, 31 }, { 41, 26 }, { 38, 32 }, { 39, 31 }, { 39, 32 }, { 42, 27 }, { 39, 33 }, { 40, 32 },
+	{ 40, 33 }, { 43, 28 }, { 40, 34 }, { 41, 33 }, { 41, 34 }, { 44, 29 }, { 41, 35 }, { 42, 34 },
+	{ 42, 35 }, { 45, 30 }, { 42, 36 }, { 43, 35 }, { 43, 36 }, { 46, 31 }, { 43, 37 }, { 44, 36 },
+	{ 44, 37 }, { 44, 37 }, { 44, 38 }, { 45, 37 }, { 45, 38 }, { 40, 48 }, { 45, 39 }, { 46, 38 },
+	{ 46, 39 }, { 41, 49 }, { 46, 40 }, { 47, 39 }, { 47, 40 }, { 43, 48 }, { 47, 41 }, { 47, 42 },
+	{ 48, 40 }, { 48, 41 }, { 48, 42 }, { 47, 45 }, { 49, 41 }, { 49, 42 }, { 49, 43 }, { 47, 48 },
+	{ 50, 42 }, { 50, 43 }, { 50, 44 }, { 48, 48 }, { 51, 43 }, { 51, 44 }, { 51, 45 }, { 49, 49 },
+	{ 52, 44 }, { 52, 45 }, { 52, 46 }, { 51, 48 }, { 53, 45 }, { 53, 46 }, { 53, 47 }, { 52, 49 },
+	{ 54, 46 }, { 54, 47 }, { 57, 42 }, { 54, 48 }, { 55, 47 }, { 55, 48 }, { 58, 43 }, { 55, 49 },
+	{ 56, 48 }, { 56, 49 }, { 59, 44 }, { 56, 50 }, { 57, 49 }, { 57, 50 }, { 60, 45 }, { 57, 51 },
+	{ 58, 50 }, { 58, 51 }, { 61, 46 }, { 58, 52 }, { 59, 51 }, { 59, 52 }, { 62, 47 }, { 59, 53 },
+	{ 60, 52 }, { 60, 53 }, { 60, 53 }, { 60, 54 }, { 61, 53 }, { 61, 54 }, { 61, 54 }, { 61, 55 },
+	{ 62, 54 }, { 62, 55 }, { 62, 55 }, { 62, 56 }, { 63, 55 }, { 63, 56 }, { 63, 56 }, { 63, 57 },
+	{ 63, 58 }, { 63, 59 }, { 63, 59 }, { 63, 60 }, { 63, 61 }, { 63, 62 }, { 63, 62 }, { 63, 63 }
+};
diff --git a/src/vulkan/runtime/meson.build b/src/vulkan/runtime/meson.build
index 1c02e1a5dd8..54c89d793a0 100644
--- a/src/vulkan/runtime/meson.build
+++ b/src/vulkan/runtime/meson.build
@@ -93,12 +93,31 @@ endif
 
 if prog_glslang.found()
   vulkan_runtime_files += files('vk_texcompress_astc.c', 'vk_texcompress_astc.h')
+  vulkan_runtime_files += files('vk_texcompress_bcn.c', 'vk_texcompress_bcn.h')
   vulkan_runtime_files += custom_target(
     'astc_spv.h',
     input : astc_decoder_glsl_file,
     output : 'astc_spv.h',
     command : [prog_glslang, '-V', '-S', 'comp', '-x', '-o', '@OUTPUT@', '@INPUT@'] + glslang_quiet,
     )
+  vulkan_runtime_files += custom_target(
+    'bc1_spv.h',
+    input : bc1_glsl_file,
+    output : 'bc1_spv.h',
+    command : [prog_glslang, '-V', '-S', 'comp', '-x', '-Ibc3_encode_head_dir', '-o', '@OUTPUT@', '@INPUT@'] + glslang_quiet,
+  )
+  vulkan_runtime_files += custom_target(
+    'bc4_spv.h',
+    input : bc4_glsl_file,
+    output : 'bc4_spv.h',
+    command : [prog_glslang, '-V', '-S', 'comp', '-x', '-Ibc3_encode_head_dir', '-o', '@OUTPUT@', '@INPUT@'] + glslang_quiet,
+  )
+  vulkan_runtime_files += custom_target(
+    'bc3_spv.h',
+    input : etc2_rgba_stitch_glsl_file,
+    output : 'bc3_spv.h',
+    command : [prog_glslang, '-V', '-S', 'comp', '-x', '-Ibc3_encode_head_dir', '-o', '@OUTPUT@', '@INPUT@'] + glslang_quiet,
+  )
 endif
 
 vk_common_entrypoints = custom_target(
diff --git a/src/vulkan/runtime/vk_command_buffer.h b/src/vulkan/runtime/vk_command_buffer.h
index 36099d03a56..1a04b53510f 100644
--- a/src/vulkan/runtime/vk_command_buffer.h
+++ b/src/vulkan/runtime/vk_command_buffer.h
@@ -28,6 +28,7 @@
 #include "vk_object.h"
 #include "util/list.h"
 #include "util/u_dynarray.h"
+#include "vk_texcompress_bcn.h"
 
 #ifdef __cplusplus
 extern "C" {
@@ -130,6 +131,12 @@ struct vk_command_buffer {
    /* This uses the same trick as STACK_ARRAY */
    struct vk_attachment_state *attachments;
    struct vk_attachment_state _attachments[8];
+
+   bool is_need_hard_encode;
+
+   struct vk_texcompress_staging_images* staging_image_list_head;
+
+   struct vk_texcompress_dummy_image* dummy;
 };
 
 VK_DEFINE_HANDLE_CASTS(vk_command_buffer, base, VkCommandBuffer,
diff --git a/src/vulkan/runtime/vk_queue.c b/src/vulkan/runtime/vk_queue.c
index 6216af9d856..15994e3420d 100644
--- a/src/vulkan/runtime/vk_queue.c
+++ b/src/vulkan/runtime/vk_queue.c
@@ -42,6 +42,7 @@
 #include "vk_util.h"
 
 #include "vulkan/wsi/wsi_common.h"
+#include "vk_texcompress_bcn.h"
 
 static VkResult
 vk_queue_start_submit_thread(struct vk_queue *queue);
@@ -1140,7 +1141,22 @@ vk_common_QueueSubmit2KHR(VkQueue _queue,
       }
    }
 
+   int n_command_buffers = 0;
+   for (uint32_t s = 0; s < submitCount; s++) {
+      n_command_buffers += pSubmits[s].commandBufferInfoCount;
+   }
+   const uint32_t const_num_cmd_buf = n_command_buffers;
+   uint32_t num_cmd_buf = 0;
+   struct vk_command_buffer *need_delete_cmd_buf[const_num_cmd_buf];
+
    for (uint32_t i = 0; i < submitCount; i++) {
+      for (uint32_t j = 0; j < pSubmits[i].commandBufferInfoCount; j++) {
+         VK_FROM_HANDLE(vk_command_buffer, commandBuffer, (pSubmits[i].pCommandBufferInfos)[j].commandBuffer);
+         if (commandBuffer->is_need_hard_encode) {
+            need_delete_cmd_buf[num_cmd_buf] = commandBuffer;
+            num_cmd_buf += 1;
+         }
+      }
       struct vulkan_submit_info info = {
          .pNext = pSubmits[i].pNext,
          .command_buffer_count = pSubmits[i].commandBufferInfoCount,
@@ -1156,6 +1172,20 @@ vk_common_QueueSubmit2KHR(VkQueue _queue,
          return result;
    }
 
+   if (num_cmd_buf > 0 && fence == NULL) {
+      vk_texcompress_create_fence(queue->base.device, NULL, &_fence);
+   }
+
+   if (num_cmd_buf > 0) {
+      vk_texcompress_wait_fence(queue->base.device, NULL, &_fence);
+      for (uint32_t i = 0; i < num_cmd_buf; i++) {
+         while (need_delete_cmd_buf[i]->staging_image_list_head != NULL) {
+            vk_texcompress_delete_head(&need_delete_cmd_buf[i]->staging_image_list_head, queue->base.device);
+         }
+         need_delete_cmd_buf[i]->is_need_hard_encode = false;
+      }
+   }
+
    return VK_SUCCESS;
 }
 
diff --git a/src/vulkan/runtime/vk_texcompress_bcn.c b/src/vulkan/runtime/vk_texcompress_bcn.c
new file mode 100644
index 00000000000..2c85d98af5c
--- /dev/null
+++ b/src/vulkan/runtime/vk_texcompress_bcn.c
@@ -0,0 +1,740 @@
+/* Copyright (c) 2017-2023 Hans-Kristian Arntzen
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+#include "vk_texcompress_bcn.h"
+#include "util/bc1_tables.h"
+#include "vk_alloc.h"
+#include "vk_format.h"
+#include "vk_image.h"
+#include "vk_physical_device.h"
+#include "log/log.h"
+
+void
+vk_texcompress_insert_head(staging_images **head, bcn_image *current)
+{
+   staging_images *node = (staging_images *)malloc(sizeof(staging_images));
+   if (node == NULL) {
+      ALOGE("Failed to alloc staging_images node");
+      return;
+   }
+
+   node->current = current;
+   node->next = *head;
+   *head = node;
+}
+
+void
+vk_texcompress_delete_head(staging_images **head, struct vk_device *device)
+{
+   if (*head == NULL) {
+      return;
+   }
+   staging_images *current_head = *head;
+   vk_texcompress_finish_image(device, NULL, current_head->current);
+   *head = current_head->next;
+   free(current_head);
+}
+
+/* type_indexes_mask bits are set/clear for support memory type index as per
+ * struct VkPhysicalDeviceMemoryProperties.memoryTypes[] */
+static uint32_t
+get_mem_type_index(struct vk_device *device, uint32_t type_indexes_mask,
+                   VkMemoryPropertyFlags mem_property)
+{
+   const struct vk_physical_device_dispatch_table *disp = &device->physical->dispatch_table;
+   VkPhysicalDevice _phy_device = vk_physical_device_to_handle(device->physical);
+
+   VkPhysicalDeviceMemoryProperties2 props2 = {
+      .sType = VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_MEMORY_PROPERTIES_2,
+      .pNext = NULL,
+   };
+   disp->GetPhysicalDeviceMemoryProperties2(_phy_device, &props2);
+
+   for (uint32_t i = 0; i < props2.memoryProperties.memoryTypeCount; i++) {
+      if ((type_indexes_mask & (1 << i)) &&
+          ((props2.memoryProperties.memoryTypes[i].propertyFlags & mem_property) == mem_property)) {
+         return i;
+      }
+   }
+
+   return -1;
+}
+
+static VkResult
+create_buffer(struct vk_device *device, VkAllocationCallbacks *allocator, struct vk_texcompress_bcn_state *bc1)
+{
+   VkDeviceSize size = sizeof(float) * (sizeof(stb__OMatch5) + sizeof(stb__OMatch6));
+   VkResult result;
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkBufferCreateInfo buffer_create_info = {
+      .sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO,
+      .size = size,
+      .usage = VK_BUFFER_USAGE_STORAGE_BUFFER_BIT,
+      .sharingMode = VK_SHARING_MODE_EXCLUSIVE,
+   };
+   result =
+      disp->CreateBuffer(_device, &buffer_create_info, allocator, &bc1->bc1_table_buf);
+   if (unlikely(result != VK_SUCCESS))
+      return result;
+
+   VkBufferMemoryRequirementsInfo2 mem_req_info = {
+      .sType = VK_STRUCTURE_TYPE_BUFFER_MEMORY_REQUIREMENTS_INFO_2,
+      .buffer = bc1->bc1_table_buf,
+   };
+   VkMemoryRequirements2 mem_req = {
+      .sType = VK_STRUCTURE_TYPE_MEMORY_REQUIREMENTS_2,
+   };
+   disp->GetBufferMemoryRequirements2(_device, &mem_req_info, &mem_req);
+
+   uint32_t mem_type_index = get_mem_type_index(
+      device, mem_req.memoryRequirements.memoryTypeBits,
+      VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT | VK_MEMORY_PROPERTY_HOST_COHERENT_BIT);
+   if (mem_type_index == -1)
+      return VK_ERROR_OUT_OF_DEVICE_MEMORY;
+
+   VkMemoryAllocateInfo alloc_info = {
+      .sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
+      .allocationSize = mem_req.memoryRequirements.size,
+      .memoryTypeIndex = mem_type_index,
+   };
+   result = disp->AllocateMemory(_device, &alloc_info, allocator, &bc1->bc1_table_mem);
+   if (unlikely(result != VK_SUCCESS))
+      return result;
+
+   result = disp->BindBufferMemory(_device, bc1->bc1_table_buf, bc1->bc1_table_mem, 0);
+   if (unlikely(result != VK_SUCCESS))
+      return result;
+
+   void *data;
+   disp->MapMemory(_device, bc1->bc1_table_mem, 0, size, 0, &data);
+   float (*data_array)[2] = (float(*)[2])data;
+   for (int i = 0; i < 256; i++) {
+      for (int j = 0; j < 2; j++) {
+         data_array[i][j] = (float)stb__OMatch5[i][j];
+         data_array[i + 256][j] = (float)stb__OMatch6[i][j];
+      }
+   }
+   disp->UnmapMemory(_device, bc1->bc1_table_mem);
+
+   return result;
+}
+
+VkResult
+vk_texcompress_create_image(struct vk_device *device, VkAllocationCallbacks *allocator, VkDeviceMemory *pMem,
+                            VkImage *image, VkExtent3D extent, VkFormat format, int size)
+{
+   bool is_dummy = (size != 0);
+   VkResult result;
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkImageCreateInfo image_create_info = {
+      .sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO,
+      .flags = VK_IMAGE_CREATE_MUTABLE_FORMAT_BIT,
+      .imageType = VK_IMAGE_TYPE_2D,
+      .format = format,
+      .extent = extent,
+      .mipLevels = 1,
+      .arrayLayers = 1,
+      .samples = VK_SAMPLE_COUNT_1_BIT,
+      .tiling = VK_IMAGE_TILING_OPTIMAL,
+      .usage = VK_IMAGE_USAGE_STORAGE_BIT | VK_IMAGE_USAGE_SAMPLED_BIT | VK_IMAGE_USAGE_TRANSFER_SRC_BIT,
+      .sharingMode = VK_SHARING_MODE_EXCLUSIVE,
+      .initialLayout = VK_IMAGE_LAYOUT_GENERAL,
+   };
+   result = disp->CreateImage(_device, &image_create_info, allocator, image);
+   if (unlikely(result != VK_SUCCESS)) {
+      return result;
+   }
+
+   if (is_dummy) {
+      VkMemoryRequirements mem_requirements;
+      disp->GetImageMemoryRequirements(_device, *image, &mem_requirements);
+
+      uint32_t mem_type_index = get_mem_type_index(
+         device, mem_requirements.memoryTypeBits, VK_MEMORY_PROPERTY_DEVICE_LOCAL_BIT);
+      if (mem_type_index == -1) {
+         return VK_ERROR_OUT_OF_HOST_MEMORY;
+      }
+
+      VkMemoryAllocateInfo alloc_info = {
+         .sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO,
+         .allocationSize = size,
+         .memoryTypeIndex = mem_type_index,
+      };
+      result = disp->AllocateMemory(_device, &alloc_info, allocator, pMem);
+      if (unlikely(result != VK_SUCCESS)) {
+         return result;
+      }
+   }
+   if (is_dummy) {
+      disp->DestroyImage(_device, *image, allocator);
+   }
+   return result;
+}
+
+void
+vk_texcompress_finish_image(struct vk_device *device, const VkAllocationCallbacks *allocator,
+                            struct vk_texcompress_bcn_image *intermidate_image)
+{
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+   disp->DestroyImage(vk_device_to_handle(device), intermidate_image->image_rgba, NULL);
+   disp->DestroyImage(vk_device_to_handle(device), intermidate_image->image_bc1, NULL);
+   disp->DestroyImage(vk_device_to_handle(device), intermidate_image->image_bc4, NULL);
+   disp->FreeMemory(vk_device_to_handle(device), intermidate_image->image_mem, NULL);
+   vk_free(&device->alloc, intermidate_image);
+}
+
+void bind_memory(struct vk_device *device, VkImage *image, VkDeviceMemory *pMem, int offset)
+{
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+   disp->BindImageMemory(_device, *image, *pMem, offset);
+}
+
+static VkResult
+create_sampler(struct vk_device *device, VkAllocationCallbacks *allocator,
+               VkSampler *sampler, VkFilter filter, VkSamplerMipmapMode mipmapMode)
+{
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkSamplerCreateInfo create_info = {
+      .sType = VK_STRUCTURE_TYPE_SAMPLER_CREATE_INFO,
+      .magFilter = filter,
+      .minFilter = filter,
+      .addressModeU = VK_SAMPLER_ADDRESS_MODE_REPEAT,
+      .addressModeV = VK_SAMPLER_ADDRESS_MODE_REPEAT,
+      .addressModeW = VK_SAMPLER_ADDRESS_MODE_REPEAT,
+      .anisotropyEnable = VK_FALSE,
+      .borderColor = VK_BORDER_COLOR_INT_OPAQUE_BLACK,
+      .unnormalizedCoordinates = VK_FALSE,
+      .compareEnable = VK_FALSE,
+      .mipmapMode = mipmapMode,
+   };
+   VkResult result = disp->CreateSampler(_device, &create_info, allocator, sampler);
+   return result;
+}
+
+static VkResult
+create_bc1_layout(struct vk_device *device, VkAllocationCallbacks *allocator,
+                  struct vk_texcompress_bcn_state *bcn)
+{
+   VkResult result;
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkDescriptorSetLayoutBinding bindings[] = {
+      {
+         .binding = 0,
+         .descriptorType = VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER,
+         .descriptorCount = 1,
+         .stageFlags = VK_SHADER_STAGE_COMPUTE_BIT,
+         .pImmutableSamplers = NULL,
+      },
+      {
+         .binding = 1,
+         .descriptorType = VK_DESCRIPTOR_TYPE_STORAGE_BUFFER,
+         .descriptorCount = 1,
+         .stageFlags = VK_SHADER_STAGE_COMPUTE_BIT,
+         .pImmutableSamplers = NULL,
+      },
+      {
+         .binding = 2,
+         .descriptorType = VK_DESCRIPTOR_TYPE_STORAGE_IMAGE,
+         .descriptorCount = 1,
+         .stageFlags = VK_SHADER_STAGE_COMPUTE_BIT,
+         .pImmutableSamplers = NULL,
+      },
+   };
+
+   VkDescriptorSetLayoutCreateInfo ds_create_info = {
+      .sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO,
+      .flags = VK_DESCRIPTOR_SET_LAYOUT_CREATE_PUSH_DESCRIPTOR_BIT_KHR,
+      .bindingCount = ARRAY_SIZE(bindings),
+      .pBindings = bindings,
+   };
+
+   result = disp->CreateDescriptorSetLayout(_device, &ds_create_info, allocator, &bcn->ds_layout);
+   return result;
+}
+
+static VkResult
+create_bc4_layout(struct vk_device *device, VkAllocationCallbacks *allocator,
+                  struct vk_texcompress_bcn_state *bcn)
+{
+   VkResult result;
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkDescriptorSetLayoutBinding bindings[] = {
+      {
+         .binding = 0,
+         .descriptorType = VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER,
+         .descriptorCount = 1,
+         .stageFlags = VK_SHADER_STAGE_COMPUTE_BIT,
+         .pImmutableSamplers = NULL,
+      },
+      {
+         .binding = 1,
+         .descriptorType = VK_DESCRIPTOR_TYPE_STORAGE_IMAGE,
+         .descriptorCount = 1,
+         .stageFlags = VK_SHADER_STAGE_COMPUTE_BIT,
+         .pImmutableSamplers = NULL,
+      },
+   };
+
+   VkDescriptorSetLayoutCreateInfo ds_create_info = {
+      .sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO,
+      .flags = VK_DESCRIPTOR_SET_LAYOUT_CREATE_PUSH_DESCRIPTOR_BIT_KHR,
+      .bindingCount = ARRAY_SIZE(bindings),
+      .pBindings = bindings,
+   };
+
+   result = disp->CreateDescriptorSetLayout(_device, &ds_create_info, allocator, &bcn->ds_layout);
+   return result;
+}
+
+static VkResult
+create_bc3_layout(struct vk_device *device, VkAllocationCallbacks *allocator,
+                  struct vk_texcompress_bcn_state *bcn)
+{
+   VkResult result;
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkDescriptorSetLayoutBinding bindings[] = {
+      {
+         .binding = 0,
+         .descriptorType = VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER,
+         .descriptorCount = 1,
+         .stageFlags = VK_SHADER_STAGE_COMPUTE_BIT,
+         .pImmutableSamplers = NULL,
+      },
+      {
+         .binding = 1,
+         .descriptorType = VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER,
+         .descriptorCount = 1,
+         .stageFlags = VK_SHADER_STAGE_COMPUTE_BIT,
+         .pImmutableSamplers = NULL,
+      },
+      {
+         .binding = 2,
+         .descriptorType = VK_DESCRIPTOR_TYPE_STORAGE_IMAGE,
+         .descriptorCount = 1,
+         .stageFlags = VK_SHADER_STAGE_COMPUTE_BIT,
+         .pImmutableSamplers = NULL,
+      },
+   };
+
+   VkDescriptorSetLayoutCreateInfo ds_create_info = {
+      .sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO,
+      .flags = VK_DESCRIPTOR_SET_LAYOUT_CREATE_PUSH_DESCRIPTOR_BIT_KHR,
+      .bindingCount = ARRAY_SIZE(bindings),
+      .pBindings = bindings,
+   };
+
+   result = disp->CreateDescriptorSetLayout(_device, &ds_create_info, allocator, &bcn->ds_layout);
+   return result;
+}
+
+static VkResult
+create_pipeline_layout(struct vk_device *device, VkAllocationCallbacks *allocator,
+                       struct vk_texcompress_bcn_state *bcn, enum compute_pipeline_id id)
+{
+   VkResult result = VK_RESULT_MAX_ENUM;
+   switch (id) {
+   case BC1_ENCODE:
+      result = create_bc1_layout(device, allocator, bcn);
+      break;
+   case BC4_ENCODE:
+      result = create_bc4_layout(device, allocator, bcn);
+      break;
+   case BC3_ENCODE:
+      result = create_bc3_layout(device, allocator, bcn);
+      break;
+   default:
+      goto fail;
+   }
+   if (result != VK_SUCCESS) {
+      goto fail;
+   }
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+   VkPipelineLayoutCreateInfo pl_create_info = {
+      .sType = VK_STRUCTURE_TYPE_PIPELINE_LAYOUT_CREATE_INFO,
+      .setLayoutCount = 1,
+      .pSetLayouts = &bcn->ds_layout,
+      .pushConstantRangeCount = 1,
+      .pPushConstantRanges = &(VkPushConstantRange) {VK_SHADER_STAGE_COMPUTE_BIT, 0, 8}
+   };
+   result = disp->CreatePipelineLayout(_device, &pl_create_info, allocator, &bcn->p_layout);
+fail:
+   return result;
+}
+
+static const uint32_t bc1_spv[] = {
+#include "bc1_spv.h"
+};
+
+static const uint32_t bc4_spv[] = {
+#include "bc4_spv.h"
+};
+
+static const uint32_t bc3_spv[] = {
+#include "bc3_spv.h"
+};
+
+static VkResult
+vk_bc1_create_shader_module(struct vk_device *device,
+                             VkAllocationCallbacks *allocator,
+                             struct vk_texcompress_bcn_state *bcn)
+{
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkShaderModuleCreateInfo shader_module_create_info = {
+      .sType = VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO,
+      .pNext = NULL,
+      .flags = 0,
+      .codeSize = sizeof(bc1_spv),
+      .pCode = bc1_spv,
+   };
+
+   return disp->CreateShaderModule(_device, &shader_module_create_info, allocator, &bcn->shader_module);
+}
+
+static VkResult
+vk_bc4_create_shader_module(struct vk_device *device,
+                             VkAllocationCallbacks *allocator,
+                             struct vk_texcompress_bcn_state *bcn)
+{
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkShaderModuleCreateInfo shader_module_create_info = {
+      .sType = VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO,
+      .pNext = NULL,
+      .flags = 0,
+      .codeSize = sizeof(bc4_spv),
+      .pCode = bc4_spv,
+   };
+
+   return disp->CreateShaderModule(_device, &shader_module_create_info, allocator, &bcn->shader_module);
+}
+
+static VkResult
+vk_bc3_create_shader_module(struct vk_device *device,
+                             VkAllocationCallbacks *allocator,
+                             struct vk_texcompress_bcn_state *bcn)
+{
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkShaderModuleCreateInfo shader_module_create_info = {
+      .sType = VK_STRUCTURE_TYPE_SHADER_MODULE_CREATE_INFO,
+      .pNext = NULL,
+      .flags = 0,
+      .codeSize = sizeof(bc3_spv),
+      .pCode = bc3_spv,
+   };
+
+   return disp->CreateShaderModule(_device, &shader_module_create_info, allocator, &bcn->shader_module);
+}
+
+static VkResult
+create_bcn_encode_pipeline(struct vk_device *device,
+                           VkAllocationCallbacks *allocator,
+                           struct vk_texcompress_bcn_state *bcn,
+                           VkPipelineCache pipeline_cache)
+{
+   VkResult result;
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+   VkPipeline pipeline;
+
+   /* compute shader */
+   VkPipelineShaderStageCreateInfo pipeline_shader_stage = {
+      .sType = VK_STRUCTURE_TYPE_PIPELINE_SHADER_STAGE_CREATE_INFO,
+      .stage = VK_SHADER_STAGE_COMPUTE_BIT,
+      .module = bcn->shader_module,
+      .pName = "main",
+   };
+
+   VkComputePipelineCreateInfo vk_pipeline_info = {
+      .sType = VK_STRUCTURE_TYPE_COMPUTE_PIPELINE_CREATE_INFO,
+      .stage = pipeline_shader_stage,
+      .flags = 0,
+      .layout = bcn->p_layout,
+   };
+
+   result = disp->CreateComputePipelines(_device, pipeline_cache, 1, &vk_pipeline_info, allocator, &pipeline);
+   if (result != VK_SUCCESS)
+      return result;
+
+   bcn->pipeline = pipeline;
+
+   return result;
+}
+
+VkPipeline
+vk_texcompress_rgba_get_encode_pipeline(struct vk_device *device, VkAllocationCallbacks *allocator,
+                                        struct vk_texcompress_bcn_state *bcn, VkPipelineCache pipeline_cache,
+                                        enum compute_pipeline_id id)
+{
+   VkResult result;
+
+   simple_mtx_lock(&bcn->mutex);
+
+   if (bcn->pipeline)
+      goto unlock;
+
+   if (id >= MAX_PIPELINE_NUM || id < 0)
+      goto unlock;
+
+   if (id == BC1_ENCODE && !bcn->shader_module) {
+      result = vk_bc1_create_shader_module(device, allocator, bcn);
+      if (result != VK_SUCCESS)
+         goto unlock;
+   }
+
+   if (id == BC4_ENCODE && !bcn->shader_module) {
+      result = vk_bc4_create_shader_module(device, allocator, bcn);
+      if (result != VK_SUCCESS)
+         goto unlock;
+   }
+
+   if (id == BC3_ENCODE && !bcn->shader_module) {
+      result = vk_bc3_create_shader_module(device, allocator, bcn);
+      if (result != VK_SUCCESS)
+         goto unlock;
+   }
+
+   create_bcn_encode_pipeline(device, allocator, bcn, pipeline_cache);
+
+unlock:
+   simple_mtx_unlock(&bcn->mutex);
+   return bcn->pipeline;
+}
+
+static inline void
+fill_desc_image_info_struct(VkDescriptorImageInfo *info, VkSampler sampler, VkImageView img_view,
+                            VkImageLayout img_layout)
+{
+   info->sampler = sampler;
+   info->imageView = img_view;
+   info->imageLayout = img_layout;
+}
+
+static inline void
+fill_write_descriptor_set_image(VkWriteDescriptorSet *set, uint8_t bind_i,
+                                VkDescriptorType desc_type, VkDescriptorImageInfo *image_info)
+{
+   set->sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+   set->pNext = NULL;
+   set->dstSet = VK_NULL_HANDLE;
+   set->dstBinding = bind_i;
+   set->dstArrayElement = 0;
+   set->descriptorCount = 1;
+   set->descriptorType = desc_type;
+   set->pImageInfo = image_info;
+   set->pBufferInfo = NULL;
+   set->pTexelBufferView = NULL;
+}
+
+static inline void
+fill_write_descriptor_set_buffer(VkWriteDescriptorSet *set,
+                                 uint8_t bind_i, VkDescriptorType desc_type,
+                                 VkDescriptorBufferInfo *buf_info)
+{
+   set->sType = VK_STRUCTURE_TYPE_WRITE_DESCRIPTOR_SET;
+   set->pNext = NULL;
+   set->dstSet = VK_NULL_HANDLE;
+   set->dstBinding = bind_i;
+   set->dstArrayElement = 0;
+   set->descriptorCount = 1;
+   set->descriptorType = desc_type;
+   set->pImageInfo = NULL;
+   set->pBufferInfo = buf_info;
+   set->pTexelBufferView = NULL;
+}
+
+void
+vk_texcompress_bc1_fill_write_descriptor_sets(struct vk_texcompress_bcn_state *bcn,
+                                               struct vk_texcompress_bcn_write_descriptor_set *set,
+                                               VkImageView src_img_view, VkImageLayout src_img_layout,
+                                               VkImageView dst_img_view)
+{
+   fill_desc_image_info_struct(&set->src_desc_image_info, bcn->sampler, src_img_view, src_img_layout);
+   fill_write_descriptor_set_image(&set->descriptor_set[0], 0,
+                                   VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, &set->src_desc_image_info);
+
+   set->bc1_buffer_info.buffer = bcn->bc1_table_buf;
+   set->bc1_buffer_info.offset = 0;
+   set->bc1_buffer_info.range = VK_WHOLE_SIZE;
+   fill_write_descriptor_set_buffer(&set->descriptor_set[1], 1,
+                                    VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, &set->bc1_buffer_info);
+
+   fill_desc_image_info_struct(&set->dst_desc_image_info, VK_NULL_HANDLE, dst_img_view, VK_IMAGE_LAYOUT_GENERAL);
+   fill_write_descriptor_set_image(&set->descriptor_set[2], 2,
+                                   VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, &set->dst_desc_image_info);
+}
+
+void
+vk_texcompress_bc4_fill_write_descriptor_sets(struct vk_texcompress_bcn_state *bcn,
+                                              struct vk_texcompress_bc4_write_descriptor_set *set,
+                                              VkImageView src_img_view, VkImageLayout src_img_layout,
+                                              VkImageView dst_img_view)
+{
+   fill_desc_image_info_struct(&set->src_desc_image_info, bcn->sampler, src_img_view, src_img_layout);
+   fill_write_descriptor_set_image(&set->descriptor_set[0], 0,
+                                   VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, &set->src_desc_image_info);
+
+   fill_desc_image_info_struct(&set->dst_desc_image_info, VK_NULL_HANDLE, dst_img_view, VK_IMAGE_LAYOUT_GENERAL);
+   fill_write_descriptor_set_image(&set->descriptor_set[1], 1,
+                                   VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, &set->dst_desc_image_info);
+}
+
+void
+vk_texcompress_bc3_fill_write_descriptor_sets(struct vk_texcompress_bcn_state *bcn,
+                                               struct vk_texcompress_bcn_write_descriptor_set *set,
+                                               VkImageView bc1_src_img_view, VkImageView bc4_src_img_view,
+                                               VkImageView dst_img_view)
+{
+   fill_desc_image_info_struct(&set->src_desc_image_info, bcn->sampler, bc1_src_img_view, VK_IMAGE_LAYOUT_GENERAL);
+   fill_write_descriptor_set_image(&set->descriptor_set[0], 0,
+                                   VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, &set->src_desc_image_info);
+
+   fill_desc_image_info_struct(&set->src_desc_image_info_2, bcn->sampler, bc4_src_img_view, VK_IMAGE_LAYOUT_GENERAL);
+   fill_write_descriptor_set_image(&set->descriptor_set[1], 1,
+                                   VK_DESCRIPTOR_TYPE_COMBINED_IMAGE_SAMPLER, &set->src_desc_image_info_2);
+
+   fill_desc_image_info_struct(&set->dst_desc_image_info, VK_NULL_HANDLE, dst_img_view, VK_IMAGE_LAYOUT_GENERAL);
+   fill_write_descriptor_set_image(&set->descriptor_set[2], 2,
+                                   VK_DESCRIPTOR_TYPE_STORAGE_IMAGE, &set->dst_desc_image_info);
+}
+
+VkResult
+vk_texcompress_bcn_init(struct vk_device *device, VkAllocationCallbacks *allocator,
+                        VkPipelineCache pipeline_cache,
+                        struct vk_texcompress_bcn_state *bcn[])
+{
+   VkResult result;
+
+   for (int pi = VK_TEXCOMPRESS_BCN_NUM_COMPUTE_PIPELINES - 1; pi >= 0; --pi) {
+      /* memory to be freed as part of vk_bcn_decode_finish() */
+      bcn[pi] = vk_zalloc(allocator, sizeof(struct vk_texcompress_bcn_state), 8, VK_SYSTEM_ALLOCATION_SCOPE_OBJECT);
+      if (bcn[pi] == NULL) {
+         return VK_ERROR_OUT_OF_HOST_MEMORY;
+      }
+      simple_mtx_init(&(bcn[pi])->mutex, mtx_plain);
+   }
+
+   result = create_buffer(device, allocator, bcn[BC1_ENCODE]);
+   if (result != VK_SUCCESS)
+      goto fail;
+
+   result = create_sampler(device, allocator, &bcn[BC1_ENCODE]->sampler,
+                           VK_FILTER_LINEAR, VK_SAMPLER_MIPMAP_MODE_LINEAR);
+   if (result != VK_SUCCESS)
+      goto fail;
+
+   result = create_pipeline_layout(device, allocator, bcn[BC1_ENCODE], BC1_ENCODE);
+   if (result != VK_SUCCESS)
+      goto fail;
+
+   result = create_sampler(device, allocator, &bcn[BC4_ENCODE]->sampler,
+                           VK_FILTER_LINEAR, VK_SAMPLER_MIPMAP_MODE_LINEAR);
+   if (result != VK_SUCCESS)
+      goto fail;
+
+   result = create_pipeline_layout(device, allocator, bcn[BC4_ENCODE], BC4_ENCODE);
+   if (result != VK_SUCCESS)
+      goto fail;
+
+   // 由于VK_FORMAT_R32G32_UINT格式的imageView格式不包含VK_FORMAT_FEATURE_SAMPLED_IMAGE_FILTER_LINEAR_BIT, 采用nearest
+   result = create_sampler(device, allocator, &bcn[BC3_ENCODE]->sampler,
+                           VK_FILTER_NEAREST, VK_SAMPLER_MIPMAP_MODE_NEAREST);
+   if (result != VK_SUCCESS)
+      goto fail;
+
+   result = create_pipeline_layout(device, allocator, bcn[BC3_ENCODE], BC3_ENCODE);
+   if (result != VK_SUCCESS)
+      goto fail;
+
+fail:
+   return result;
+}
+
+void
+vk_texcompress_bcn_finish(struct vk_device *device,
+                          VkAllocationCallbacks *allocator,
+                          struct vk_texcompress_bcn_state *bcn[])
+{
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   for (int pi = VK_TEXCOMPRESS_BCN_NUM_COMPUTE_PIPELINES - 1; pi >= 0; --pi) {
+      if (bcn[pi]->pipeline)
+         disp->DestroyPipeline(_device, bcn[pi]->pipeline, allocator);
+
+      disp->DestroyPipelineLayout(_device, bcn[pi]->p_layout, allocator);
+      disp->DestroyShaderModule(_device, bcn[pi]->shader_module, allocator);
+      disp->DestroyDescriptorSetLayout(_device, bcn[pi]->ds_layout, allocator);
+
+      if (pi == BC1_ENCODE) {
+         disp->DestroyBuffer(_device, bcn[pi]->bc1_table_buf, allocator);
+         disp->FreeMemory(_device, bcn[pi]->bc1_table_mem, allocator);
+      }
+      vk_free(allocator, bcn[pi]);
+   }
+}
+
+void
+vk_texcompress_create_fence(struct vk_device *device, VkAllocationCallbacks *allocator, VkFence *fence)
+{
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkFenceCreateInfo create_info = {
+      .sType = VK_STRUCTURE_TYPE_FENCE_CREATE_INFO,
+      .flags = VK_FENCE_CREATE_SIGNALED_BIT,
+   };
+   VkResult result = disp->CreateFence(_device, &create_info, allocator, fence);
+   if (result != VK_SUCCESS) {
+      ALOGE("Failed to create fence");
+   }
+}
+
+void
+vk_texcompress_wait_fence(struct vk_device *device, VkAllocationCallbacks *allocator, VkFence *fence)
+{
+   VkDevice _device = vk_device_to_handle(device);
+   const struct vk_device_dispatch_table *disp = &device->dispatch_table;
+
+   VkResult result = disp->WaitForFences(_device, 1, fence, VK_TRUE, UINT64_MAX);
+   if (result != VK_SUCCESS) {
+      ALOGE("Failed to wait for fence");
+   }
+}
\ No newline at end of file
diff --git a/src/vulkan/runtime/vk_texcompress_bcn.h b/src/vulkan/runtime/vk_texcompress_bcn.h
new file mode 100644
index 00000000000..ca8189b657e
--- /dev/null
+++ b/src/vulkan/runtime/vk_texcompress_bcn.h
@@ -0,0 +1,142 @@
+/* Copyright (c) 2017-2023 Hans-Kristian Arntzen
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining
+ * a copy of this software and associated documentation files (the
+ * "Software"), to deal in the Software without restriction, including
+ * without limitation the rights to use, copy, modify, merge, publish,
+ * distribute, sublicense, and/or sell copies of the Software, and to
+ * permit persons to whom the Software is furnished to do so, subject to
+ * the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+ * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+ * CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+ * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+ * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+#ifndef VK_TEXCOMPRESS_BCN_H
+#define VK_TEXCOMPRESS_BCN_H
+
+#include "util/simple_mtx.h"
+#include "vk_fence.h"
+#include "vk_device.h"
+
+#define VK_TEXCOMPRESS_BCN_NUM_COMPUTE_PIPELINES 3
+#define VK_TEXCOMPRESS_BCN_WRITE_DESC_SET_COUNT 3  // 每个着色器最多有3个描述符集
+
+struct vk_texcompress_bcn_write_descriptor_set {
+   VkWriteDescriptorSet descriptor_set[VK_TEXCOMPRESS_BCN_WRITE_DESC_SET_COUNT];
+   VkDescriptorImageInfo dst_desc_image_info;
+   VkDescriptorImageInfo src_desc_image_info;
+   VkDescriptorImageInfo src_desc_image_info_2;
+   VkDescriptorBufferInfo bc1_buffer_info;
+};
+
+struct vk_texcompress_bc4_write_descriptor_set {
+   VkWriteDescriptorSet descriptor_set[2];
+   VkDescriptorImageInfo dst_desc_image_info;
+   VkDescriptorImageInfo src_desc_image_info;
+};
+
+typedef struct vk_texcompress_bcn_image {
+   VkImage image_dummy;
+   VkImage image_rgba;
+   VkImage image_bc1;
+   VkImage image_bc4;
+   VkDeviceMemory image_mem;
+} bcn_image;
+
+struct vk_texcompress_dummy_image {
+   VkImage image_dummy;
+   VkDeviceMemory image_mem;
+};
+
+typedef struct vk_texcompress_staging_images {
+   struct vk_texcompress_bcn_image *current;
+   struct vk_texcompress_staging_images* next;
+} staging_images;
+
+enum compute_pipeline_id {
+   BC1_ENCODE = 0,
+   BC4_ENCODE,
+   BC3_ENCODE,
+   MAX_PIPELINE_NUM
+};
+
+struct vk_texcompress_bcn_state {
+   simple_mtx_t mutex;
+   VkDescriptorSetLayout ds_layout;
+   VkPipelineLayout p_layout;
+   VkPipeline pipeline;
+   VkShaderModule shader_module;
+   VkSampler sampler;
+
+   VkDeviceMemory bc1_table_mem;
+   VkBuffer bc1_table_buf;
+};
+
+void
+vk_texcompress_insert_head(staging_images **head, bcn_image *current);
+
+void
+vk_texcompress_delete_head(staging_images **head, struct vk_device *device);
+   
+VkResult
+vk_texcompress_create_image(struct vk_device *device, VkAllocationCallbacks *allocator, VkDeviceMemory *pMem,
+                            VkImage *image, VkExtent3D extent, VkFormat format, int size);
+
+void bind_memory(struct vk_device *device, VkImage *image, VkDeviceMemory *pMem, int offset);
+
+void
+vk_texcompress_finish_image(struct vk_device *device, const VkAllocationCallbacks *allocator,
+                            struct vk_texcompress_bcn_image *intermidate_image);
+
+void
+vk_texcompress_bc1_fill_write_descriptor_sets(struct vk_texcompress_bcn_state *bcn,
+                                              struct vk_texcompress_bcn_write_descriptor_set *set,
+                                              VkImageView src_img_view, VkImageLayout src_img_layout,
+                                              VkImageView dst_img_view);
+
+void
+vk_texcompress_bc4_fill_write_descriptor_sets(struct vk_texcompress_bcn_state *bcn,
+                                              struct vk_texcompress_bc4_write_descriptor_set *set,
+                                              VkImageView src_img_view, VkImageLayout src_img_layout,
+                                              VkImageView dst_img_view);
+
+void
+vk_texcompress_bc3_fill_write_descriptor_sets(struct vk_texcompress_bcn_state *bcn,
+                                              struct vk_texcompress_bcn_write_descriptor_set *set,
+                                              VkImageView bc1_src_img_view, VkImageView bc4_src_img_view,
+                                              VkImageView dst_img_view);
+
+VkPipeline
+vk_texcompress_rgba_get_encode_pipeline(struct vk_device *device,
+                                        VkAllocationCallbacks *allocator,
+                                        struct vk_texcompress_bcn_state *bcn,
+                                        VkPipelineCache pipeline_cache,
+                                        enum compute_pipeline_id id);
+
+VkResult
+vk_texcompress_bcn_init(struct vk_device *device,
+                        VkAllocationCallbacks *allocator,
+                        VkPipelineCache pipeline_cache,
+                        struct vk_texcompress_bcn_state *bcn[]);
+
+void
+vk_texcompress_bcn_finish(struct vk_device *device,
+                          VkAllocationCallbacks *allocator,
+                          struct vk_texcompress_bcn_state *bcn[]);
+
+
+void
+vk_texcompress_create_fence(struct vk_device *device, VkAllocationCallbacks *allocator, VkFence *fence);
+
+void
+vk_texcompress_wait_fence(struct vk_device *device, VkAllocationCallbacks *allocator, VkFence *fence);
+
+#endif /* VK_TEXCOMPRESS_BCN_H */
