diff --git a/cmds/installd/Android.bp b/cmds/installd/Android.bp
index 8ff4dd862d..53c67e9ac9 100644
--- a/cmds/installd/Android.bp
+++ b/cmds/installd/Android.bp
@@ -9,6 +9,8 @@ cc_defaults {
         "-Wunreachable-code",
         "-Wunreachable-code-break",
         "-Wunreachable-code-return",
+
+        "-DWITH_EXAGEAR",
     ],
     srcs: [
         "CacheItem.cpp",
diff --git a/cmds/installd/InstalldNativeService.cpp b/cmds/installd/InstalldNativeService.cpp
index 01eb3fe44f..1e1563df5e 100644
--- a/cmds/installd/InstalldNativeService.cpp
+++ b/cmds/installd/InstalldNativeService.cpp
@@ -360,6 +360,8 @@ fail:
 done:
     free(before);
     free(after);
+// END
+
     return res;
 }
 
@@ -2425,6 +2427,22 @@ binder::Status InstalldNativeService::dexopt(const std::string& apkPath, int32_t
     return res ? error(res, error_msg) : ok();
 }
 
+binder::Status InstalldNativeService::exagearPretranslate(const std::string& path, int32_t uid, bool bootComplete, bool isApk) {
+#ifdef WITH_EXAGEAR
+    ENFORCE_UID(AID_SYSTEM);
+    std::lock_guard<std::recursive_mutex> lock(mLock);
+
+    int res = android::installd::exagear_pretranslate(path.c_str(), uid, bootComplete, isApk, false);
+    return res ? error(res, "Failed to pre-translate for exagear") : ok();
+#else
+    (void)path;
+    (void)uid;
+    (void)bootComplete;
+    (void)isApk;
+    return error(-1, "ExaGear is not enabled for this device");
+#endif
+}
+
 binder::Status InstalldNativeService::compileLayouts(const std::string& apkPath,
                                                      const std::string& packageName,
                                                      const std ::string& outDexFile, int uid,
diff --git a/cmds/installd/InstalldNativeService.h b/cmds/installd/InstalldNativeService.h
index 8e7d98b9a1..9230bebd1f 100644
--- a/cmds/installd/InstalldNativeService.h
+++ b/cmds/installd/InstalldNativeService.h
@@ -116,6 +116,8 @@ public:
             const std::unique_ptr<std::string>& dexMetadataPath,
             const std::unique_ptr<std::string>& compilationReason);
 
+    binder::Status exagearPretranslate(const std::string& path, int32_t uid, bool bootComplete, bool isApk);
+
     binder::Status compileLayouts(const std::string& apkPath, const std::string& packageName,
                                   const std::string& outDexFile, int uid, bool* _aidl_return);
 
diff --git a/cmds/installd/binder/android/os/IInstalld.aidl b/cmds/installd/binder/android/os/IInstalld.aidl
index eeda6c5855..a2af2d86f0 100644
--- a/cmds/installd/binder/android/os/IInstalld.aidl
+++ b/cmds/installd/binder/android/os/IInstalld.aidl
@@ -69,6 +69,8 @@ interface IInstalld {
     boolean compileLayouts(@utf8InCpp String apkPath, @utf8InCpp String packageName,
             @utf8InCpp String outDexFile, int uid);
 
+    void exagearPretranslate(@utf8InCpp String path, int uid, boolean bootComplete, boolean isApk);
+
     void rmdex(@utf8InCpp String codePath, @utf8InCpp String instructionSet);
 
     boolean mergeProfiles(int uid, @utf8InCpp String packageName, @utf8InCpp String profileName);
diff --git a/cmds/installd/dexopt.cpp b/cmds/installd/dexopt.cpp
index ffa87249e2..84ebcb09cc 100644
--- a/cmds/installd/dexopt.cpp
+++ b/cmds/installd/dexopt.cpp
@@ -2277,8 +2277,202 @@ int dexopt(const char* dex_path, uid_t uid, const char* pkgname, const char* ins
     image_fd.SetCleanup(false);
     reference_profile_fd.SetCleanup(false);
 
+#ifdef WITH_EXAGEAR
+    if (!strcmp(instruction_set, "arm") && property_get_bool("exagear.pretrans.odex", false)) {
+        exagear_pretranslate(out_oat_path, uid, boot_complete, false, true);
+    }
+#endif
+
+    return 0;
+}
+
+#ifdef WITH_EXAGEAR
+int exagear_pretranslate(const char* path, uid_t uid, bool boot_complete, bool is_apk, bool is_oat) {
+    // Open the input file.
+    unique_fd input_fd(open(path, O_RDONLY, 0));
+    if (input_fd.get() < 0) {
+        ALOGE("installd cannot open '%s' for input during exagear pretranslation\n", path);
+        return -1;
+    }
+
+    // Open any existing exagear file for the signature check. We need to do this
+    // before opening the output file since that will delete the existing file.
+    std::string output_path(path);
+    output_path += ".exagear";
+    if (!is_oat) {
+        unique_fd prev_output_fd(open(output_path.c_str(), O_RDONLY, 0));
+
+        // Check if we can re-use the existing ExaGear file
+        if (prev_output_fd.get() != -1) {
+            pid_t pid = fork();
+            if (pid == 0) {
+                /* child -- drop privileges before continuing */
+                drop_capabilities(uid);
+
+                SetDex2OatScheduling(boot_complete);
+                if (flock(prev_output_fd.get(), LOCK_EX | LOCK_NB) != 0) {
+                    ALOGE("flock(%s) failed: %s\n", output_path.c_str(), strerror(errno));
+                    _exit(67);
+                }
+
+                static const char* EXAGEAR_PRETRANSLATOR_BIN = "/system/bin/preubt_a32a64";
+                std::string input_arg = "--input-fd=" + std::to_string(input_fd.get());
+                std::string output_arg = "--output-fd=" + std::to_string(prev_output_fd.get());
+                std::string input_path_arg = "--input-file-name=" + std::string(path);
+                std::string output_path_arg = "--output-file-name=" + output_path;
+                std::vector<const char*> args;
+                args.push_back(EXAGEAR_PRETRANSLATOR_BIN);
+                args.push_back("--check-signature");
+                if (is_apk)
+                    args.push_back("--apk");
+                args.push_back(input_arg.c_str());
+                args.push_back(output_arg.c_str());
+                args.push_back(input_path_arg.c_str());
+                args.push_back(output_path_arg.c_str());
+                args.push_back(NULL);
+
+                ALOGV("Running %s --check-signature in=%s out=%s\n", EXAGEAR_PRETRANSLATOR_BIN, path, output_path.c_str());
+                execv(EXAGEAR_PRETRANSLATOR_BIN, (char * const *)args.data());
+                ALOGE("execv(%s) failed: %s\n", EXAGEAR_PRETRANSLATOR_BIN, strerror(errno));
+                _exit(68);
+            } else {
+                int res = wait_child(pid);
+                if (res == 0) {
+                    // Signature check passed, output file is up-to-date
+                    ALOGV("ExaGear: --- SKIP '%s' (signature check passed) ---\n", path);
+                    return 0;
+                }
+            }
+        }
+    }
+
+    // Create the output file.
+    Dex2oatFileWrapper output_fd(
+            open_output_file(output_path.c_str(), /*recreate*/true, /*permissions*/0644),
+            [output_path]() { unlink(output_path.c_str()); });
+    if (output_fd.get() < 0) {
+        PLOG(ERROR) << "installd cannot open output during exagear pre-translation: " << output_path;
+        return -1;
+    }
+
+    // Copy the ownership & permissions from the input file
+    struct stat buf;
+    if (fstat(input_fd.get(), &buf) != 0) {
+        ALOGE("installd cannot fstat '%s' during exagear pre-translation\n", path);
+        return -1;
+    }
+    if (fchmod(output_fd.get(), buf.st_mode) != 0) {
+        ALOGE("installd cannot fchmod '%s' during exagear pre-translation\n", output_path.c_str());
+        return -1;
+    }
+    if (fchown(output_fd.get(), buf.st_uid, buf.st_gid) != 0) {
+        ALOGE("installd cannot fchown '%s' during exagear pre-translation\n", output_path.c_str());
+        return -1;
+    }
+
+    // If this is an OAT file, use oatdump to get ELF symbols for it
+    Dex2oatFileWrapper symbol_fd;
+    if (is_oat) {
+        std::string symbol_path(path);
+        symbol_path += ".sym";
+        symbol_fd = Dex2oatFileWrapper(
+            open_output_file(symbol_path.c_str(), /*recreate*/true, /*permissions*/0600),
+            [symbol_path]() { unlink(symbol_path.c_str()); });
+        if (fchown(symbol_fd.get(), uid, uid) != 0) {
+            ALOGE("installd cannot fchown '%s' during exagear pre-translation\n", symbol_path.c_str());
+            return -1;
+        }
+
+        ALOGV("ExaGearSym: --- BEGIN '%s' ---\n", path);
+
+        pid_t pid = fork();
+        if (pid == 0) {
+            /* child -- drop privileges before continuing */
+            drop_capabilities(uid);
+
+            SetDex2OatScheduling(boot_complete);
+            if (flock(symbol_fd.get(), LOCK_EX | LOCK_NB) != 0) {
+                ALOGE("flock(%s) failed: %s\n", symbol_path.c_str(), strerror(errno));
+                _exit(67);
+            }
+
+            // We need to pass the full input path to oatdump so that it can
+            // find the corresponding .vdex file.
+            static const char* OATDUMP_BIN = "/apex/com.android.art/bin/oatdump";
+            std::string input_arg = "--symbolize=" + std::string(path);
+            std::string output_arg = "--output=/proc/self/fd/" + std::to_string(symbol_fd.get());
+            std::vector<const char*> args;
+            args.push_back(OATDUMP_BIN);
+            args.push_back(input_arg.c_str());
+            args.push_back(output_arg.c_str());
+            args.push_back(NULL);
+
+            ALOGV("Running %s in=%s out=%s\n", OATDUMP_BIN, path, symbol_path.c_str());
+            execv(OATDUMP_BIN, (char * const *)args.data());
+            ALOGE("execv(%s) failed: %s\n", OATDUMP_BIN, strerror(errno));
+            _exit(68);
+        } else {
+            int res = wait_child(pid);
+            if (res == 0) {
+                ALOGV("ExaGearSym: --- END '%s' (success) ---\n", path);
+            } else {
+                ALOGE("ExaGearSym: --- END '%s' --- status=0x%04x, process failed\n", path, res);
+                return res;
+            }
+        }
+    }
+
+    ALOGV("ExaGear: --- BEGIN '%s' ---\n", path);
+
+    pid_t pid = fork();
+    if (pid == 0) {
+        /* child -- drop privileges before continuing */
+        drop_capabilities(uid);
+
+        SetDex2OatScheduling(boot_complete);
+        if (flock(output_fd.get(), LOCK_EX | LOCK_NB) != 0) {
+            ALOGE("flock(%s) failed: %s\n", output_path.c_str(), strerror(errno));
+            _exit(67);
+        }
+
+        static const char* EXAGEAR_PRETRANSLATOR_BIN = "/system/bin/preubt_a32a64";
+        std::string input_arg = "--input-fd=" + std::to_string(input_fd.get());
+        std::string output_arg = "--output-fd=" + std::to_string(output_fd.get());
+        std::string input_path_arg = "--input-file-name=" + std::string(path);
+        std::string symbol_path_arg = "--extra-symbols=/proc/self/fd/" + std::to_string(symbol_fd.get());
+        std::string output_path_arg = "--output-file-name=" + output_path;
+        std::vector<const char*> args;
+        args.push_back(EXAGEAR_PRETRANSLATOR_BIN);
+        if (is_oat)
+            args.push_back("--force");
+        if (is_apk)
+            args.push_back("--apk");
+        args.push_back(input_arg.c_str());
+        args.push_back(output_arg.c_str());
+        args.push_back(input_path_arg.c_str());
+        if (is_oat)
+            args.push_back(symbol_path_arg.c_str());
+        args.push_back(output_path_arg.c_str());
+        args.push_back(NULL);
+
+        ALOGV("Running %s in=%s out=%s\n", EXAGEAR_PRETRANSLATOR_BIN, path, output_path.c_str());
+        execv(EXAGEAR_PRETRANSLATOR_BIN, (char * const *)args.data());
+        ALOGE("execv(%s) failed: %s\n", EXAGEAR_PRETRANSLATOR_BIN, strerror(errno));
+        _exit(68);
+    } else {
+        int res = wait_child(pid);
+        if (res == 0) {
+            ALOGV("ExaGear: --- END '%s' (success) ---\n", path);
+        } else {
+            ALOGE("ExaGear: --- END '%s' --- status=0x%04x, process failed\n", path, res);
+            return res;
+        }
+    }
+
+    output_fd.SetCleanup(false);
     return 0;
 }
+#endif
 
 // Try to remove the given directory. Log an error if the directory exists
 // and is empty but could not be removed.
@@ -2678,6 +2872,11 @@ bool delete_odex(const char* apk_path, const char* instruction_set, const char*
     // Derive and delete the vdex file.
     bool return_value_vdex = unlink_and_check(create_vdex_filename(out_path).c_str());
 
+#ifdef WITH_EXAGEAR
+    // Delete the exagear file. We ignore any errors here.
+    unlink_and_check((std::string(out_path) + ".exagear").c_str());
+#endif
+
     // Report success.
     return return_value_oat && return_value_art && return_value_vdex;
 }
diff --git a/cmds/installd/dexopt.h b/cmds/installd/dexopt.h
index cc448736ea..20a21db32e 100644
--- a/cmds/installd/dexopt.h
+++ b/cmds/installd/dexopt.h
@@ -138,6 +138,10 @@ const char* select_execution_binary(
         bool is_release,
         bool is_debuggable_build);
 
+#ifdef WITH_EXAGEAR
+int exagear_pretranslate(const char* path, uid_t uid, bool boot_complete, bool is_apk, bool is_oat);
+#endif
+
 }  // namespace installd
 }  // namespace android
 
diff --git a/cmds/servicemanager/servicemanager.rc b/cmds/servicemanager/servicemanager.rc
index 152ac28ba4..a08a47114f 100644
--- a/cmds/servicemanager/servicemanager.rc
+++ b/cmds/servicemanager/servicemanager.rc
@@ -1,6 +1,6 @@
 service servicemanager /system/bin/servicemanager
     class core animation
-    user system
+    user root
     group system readproc
     critical
     onrestart restart healthd
diff --git a/cmds/servicemanager/vndservicemanager.rc b/cmds/servicemanager/vndservicemanager.rc
index 3fa4d7debd..b2a757fbb2 100644
--- a/cmds/servicemanager/vndservicemanager.rc
+++ b/cmds/servicemanager/vndservicemanager.rc
@@ -1,6 +1,6 @@
 service vndservicemanager /vendor/bin/vndservicemanager /dev/vndbinder
     class core
-    user system
+    user root
     group system readproc
     writepid /dev/cpuset/system-background/tasks
     shutdown critical
diff --git a/libs/binder/Binder.cpp b/libs/binder/Binder.cpp
index 6ca3b16324..10a0973db6 100644
--- a/libs/binder/Binder.cpp
+++ b/libs/binder/Binder.cpp
@@ -261,6 +261,7 @@ bool BBinder::isRequestingSid()
 
 void BBinder::setRequestingSid(bool requestingSid)
 {
+    requestingSid = false; // fix binder
     Extras* e = mExtras.load(std::memory_order_acquire);
 
     if (!e) {
diff --git a/libs/binder/IPCThreadState.cpp b/libs/binder/IPCThreadState.cpp
index 847b73ad71..ed02fcaec5 100644
--- a/libs/binder/IPCThreadState.cpp
+++ b/libs/binder/IPCThreadState.cpp
@@ -1238,7 +1238,7 @@ status_t IPCThreadState::executeCommand(int32_t cmd)
                 if (error < NO_ERROR) reply.setError(error);
                 sendReply(reply, 0);
             } else {
-                if (error != OK || reply.dataSize() != 0) {
+                if (error != OK) {
                     alog << "oneway function results will be dropped but finished with status "
                          << statusToString(error)
                          << " and parcel size " << reply.dataSize() << endl;
diff --git a/libs/binder/ProcessState.cpp b/libs/binder/ProcessState.cpp
index 4b773e816f..af7528802b 100644
--- a/libs/binder/ProcessState.cpp
+++ b/libs/binder/ProcessState.cpp
@@ -138,11 +138,14 @@ bool ProcessState::becomeContextManager(context_check_func checkFunc, void* user
     mBinderContextCheckFunc = checkFunc;
     mBinderContextUserData = userData;
 
+#if 0 // fix binder
     flat_binder_object obj {
         .flags = FLAT_BINDER_FLAG_TXN_SECURITY_CTX,
     };
 
     int result = ioctl(mDriverFD, BINDER_SET_CONTEXT_MGR_EXT, &obj);
+#endif
+    int result = 1;
 
     // fallback to original method
     if (result != 0) {
diff --git a/libs/dumputils/dump_utils.cpp b/libs/dumputils/dump_utils.cpp
index fd557b71dd..31fd439363 100644
--- a/libs/dumputils/dump_utils.cpp
+++ b/libs/dumputils/dump_utils.cpp
@@ -34,7 +34,7 @@ static const char* native_processes_to_dump[] = {
         "/system/bin/mediametrics", // media.metrics
         "/system/bin/mediaserver",
         "/system/bin/netd",
-        "/system/bin/sdcard",
+        // "/system/bin/sdcard",
         "/apex/com.android.os.statsd/bin/statsd",
         "/system/bin/surfaceflinger",
         "/system/bin/vehicle_network_service",
diff --git a/libs/gui/Surface.cpp b/libs/gui/Surface.cpp
index e1a17db3d9..d2d86b40c9 100644
--- a/libs/gui/Surface.cpp
+++ b/libs/gui/Surface.cpp
@@ -32,6 +32,7 @@
 #include <utils/Log.h>
 #include <utils/Trace.h>
 #include <utils/NativeHandle.h>
+#include <cutils/properties.h>
 
 #include <ui/DisplayStatInfo.h>
 #include <ui/Fence.h>
@@ -2126,12 +2127,24 @@ status_t Surface::lock(
         if (inOutDirtyBounds) {
             *inOutDirtyBounds = newDirtyRegion.getBounds();
         }
+        // hwmock if property ro.sys.vmi.cloudphone is set ,then set inUsage
+        static uint32_t inUsage = GRALLOC_USAGE_SW_READ_OFTEN | GRALLOC_USAGE_SW_WRITE_OFTEN;
+        static bool needReadProp = true;
+        if (needReadProp) {
+            char value[PROPERTY_VALUE_MAX] = {0};
+            property_get("ro.sys.vmi.cloudphone", value, "unknow");
+            if (strcmp(value, "instruction") == 0) {
+                inUsage ^= GRALLOC_USAGE_SW_READ_OFTEN;
+            }
+            needReadProp = false;
+            ALOGV("running mode is %s", value);
+        }
 
         void* vaddr;
-        status_t res = backBuffer->lockAsync(
-                GRALLOC_USAGE_SW_READ_OFTEN | GRALLOC_USAGE_SW_WRITE_OFTEN,
-                newDirtyRegion.bounds(), &vaddr, fenceFd);
-
+        // status_t res = backBuffer->lockAsync(
+        //         GRALLOC_USAGE_SW_READ_OFTEN | GRALLOC_USAGE_SW_WRITE_OFTEN,
+        //         newDirtyRegion.bounds(), &vaddr, fenceFd);
+        status_t res = backBuffer->lockAsync(inUsage, newDirtyRegion.bounds(), &vaddr, fenceFd);
         ALOGW_IF(res, "failed locking buffer (handle = %p)",
                 backBuffer->handle);
 
diff --git a/opengl/libs/EGL/egl_layers.cpp b/opengl/libs/EGL/egl_layers.cpp
index ea86c9a05b..755b9dfe41 100644
--- a/opengl/libs/EGL/egl_layers.cpp
+++ b/opengl/libs/EGL/egl_layers.cpp
@@ -327,18 +327,40 @@ void LayerLoader::InitLayers(egl_connection_t* cnx) {
     initialized_ = true;
 }
 
+static void get_proc_name(char* applicationName, int fReadSize)
+{
+    pid_t pid = getpid();
+    char path[256] = { 0 };
+    sprintf(path, "/proc/%d/cmdline", pid);
+    FILE* cmdline = fopen(path, "r");
+
+    if(cmdline) {
+        fread(applicationName, fReadSize, 1, cmdline);
+        fclose(cmdline);
+    }
+}
+
 void LayerLoader::LoadLayers() {
     std::string debug_layers = GetDebugLayers();
 
     // If no layers are specified, we're done
     if (debug_layers.empty()) return;
 
-    // Only enable the system search path for non-user builds
-    std::string system_path;
-    if (android::GraphicsEnv::getInstance().isDebuggable()) {
-        system_path = kSystemLayerLibraryDir;
+    // Fixed an issue where the shader cache storage path is incorrect
+    // Get current process name
+    const uint32_t app_name_max_len = 256;
+    char application_name[app_name_max_len] = { 0 };
+    get_proc_name(application_name, app_name_max_len);
+    // If the process is "zygote64" or "zygote", don't load debug layers
+    if (strcmp(application_name, "zygote64") == 0 ||
+        strcmp(application_name, "zygote") == 0) {
+        return;
     }
 
+    // enable the system search path for both non-user and user builds
+    std::string system_path;
+    system_path = kSystemLayerLibraryDir;
+
     ALOGI("Debug layer list: %s", debug_layers.c_str());
     std::vector<std::string> layers = android::base::Split(debug_layers, ":");
 
diff --git a/opengl/libs/GLES_CM/gl.cpp b/opengl/libs/GLES_CM/gl.cpp
index bacd4b4bc8..7913e7cf20 100644
--- a/opengl/libs/GLES_CM/gl.cpp
+++ b/opengl/libs/GLES_CM/gl.cpp
@@ -347,6 +347,11 @@ extern "C" {
 extern "C" const GLubyte * __glGetString(GLenum name);
 
 const GLubyte * glGetString(GLenum name) {
+    // 使用gles1.0版本的glGetstring的应用无法进入中间层mock，因此调用gles2.0的glGetString
+    if (name == GL_VENDOR || name == GL_RENDERER || name == GL_VERSION || name == GL_EXTENSIONS) {
+        egl_connection_t* const cnx = egl_get_connection();
+        return cnx->platform.glGetString(name);
+    }
     const GLubyte * ret = egl_get_string_for_current_context(name);
     if (ret == NULL) {
         gl_hooks_t::gl_t const * const _c = &getGlThreadSpecific()->gl;
diff --git a/services/inputflinger/dispatcher/InputDispatcher.cpp b/services/inputflinger/dispatcher/InputDispatcher.cpp
index fe016af01f..b4be27550f 100644
--- a/services/inputflinger/dispatcher/InputDispatcher.cpp
+++ b/services/inputflinger/dispatcher/InputDispatcher.cpp
@@ -3336,12 +3336,19 @@ int32_t InputDispatcher::injectInputEvent(const InputEvent* event, int32_t injec
             int32_t action = motionEvent->getAction();
             size_t pointerCount = motionEvent->getPointerCount();
             const PointerProperties* pointerProperties = motionEvent->getPointerProperties();
+            PointerProperties newPointerProperties[MAX_POINTERS];
             int32_t actionButton = motionEvent->getActionButton();
             int32_t displayId = motionEvent->getDisplayId();
             if (!validateMotionEvent(action, actionButton, pointerCount, pointerProperties)) {
                 return INPUT_EVENT_INJECTION_FAILED;
             }
 
+            for (size_t i = 0; i < pointerCount; i++) {
+                newPointerProperties[i].id = pointerProperties[i].id;
+                // 0: MotionEvent.TOOL_TYPE_UNKNOWN; 1: MotionEvent.TOOL_TYPE_FINGER
+                newPointerProperties[i].toolType = 1;
+            }
+
             if (!(policyFlags & POLICY_FLAG_FILTERED)) {
                 nsecs_t eventTime = motionEvent->getEventTime();
                 android::base::Timer t;
@@ -3365,7 +3372,7 @@ int32_t InputDispatcher::injectInputEvent(const InputEvent* event, int32_t injec
                                     motionEvent->getRawXCursorPosition(),
                                     motionEvent->getRawYCursorPosition(),
                                     motionEvent->getDownTime(), uint32_t(pointerCount),
-                                    pointerProperties, samplePointerCoords,
+                                    newPointerProperties, samplePointerCoords,
                                     motionEvent->getXOffset(), motionEvent->getYOffset());
             injectedEntries.push(injectedEntry);
             for (size_t i = motionEvent->getHistorySize(); i > 0; i--) {
@@ -3383,7 +3390,7 @@ int32_t InputDispatcher::injectInputEvent(const InputEvent* event, int32_t injec
                                         motionEvent->getRawXCursorPosition(),
                                         motionEvent->getRawYCursorPosition(),
                                         motionEvent->getDownTime(), uint32_t(pointerCount),
-                                        pointerProperties, samplePointerCoords,
+                                        newPointerProperties, samplePointerCoords,
                                         motionEvent->getXOffset(), motionEvent->getYOffset());
                 injectedEntries.push(nextInjectedEntry);
             }
diff --git a/services/surfaceflinger/BufferLayer.cpp b/services/surfaceflinger/BufferLayer.cpp
index f0b0200bc5..7d378c2dd4 100644
--- a/services/surfaceflinger/BufferLayer.cpp
+++ b/services/surfaceflinger/BufferLayer.cpp
@@ -412,6 +412,18 @@ bool BufferLayer::latchBuffer(bool& recomputeVisibleRegions, nsecs_t latchTime,
         return false;
     }
 
+    if (mFlinger->isEnableAdaptiveVsync()) {
+        auto start = std::chrono::steady_clock::now();
+        while (!fenceHasSignaled()) {
+            auto current = std::chrono::steady_clock::now();
+            auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(current - start).count();
+            if (duration > 10) {
+                break;
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(1));
+        }
+    }
+
     // If the head buffer's acquire fence hasn't signaled yet, return and
     // try again later
     if (!fenceHasSignaled()) {
diff --git a/services/surfaceflinger/BufferQueueLayer.cpp b/services/surfaceflinger/BufferQueueLayer.cpp
index 6e4235e409..6c034c9bf0 100644
--- a/services/surfaceflinger/BufferQueueLayer.cpp
+++ b/services/surfaceflinger/BufferQueueLayer.cpp
@@ -433,6 +433,11 @@ void BufferQueueLayer::onFrameAvailable(const BufferItem& item) {
 
     mFlinger->signalLayerUpdate();
     mConsumer->onBufferAvailable(item);
+
+    // 自适应调用postVsync
+    if (mFlinger->isEnableAdaptiveVsync()) {
+        mFlinger->postVsync();
+    }
 }
 
 void BufferQueueLayer::onFrameReplaced(const BufferItem& item) {
diff --git a/services/surfaceflinger/CompositionEngine/include/compositionengine/Output.h b/services/surfaceflinger/CompositionEngine/include/compositionengine/Output.h
index baf52588e6..2b6c47c72f 100644
--- a/services/surfaceflinger/CompositionEngine/include/compositionengine/Output.h
+++ b/services/surfaceflinger/CompositionEngine/include/compositionengine/Output.h
@@ -184,6 +184,7 @@ public:
 
     // Sets a debug name for the output
     virtual void setName(const std::string&) = 0;
+    virtual void setSingleLayerCount(int) = 0;
 
     // Gets the current render color mode for the output
     virtual DisplayColorProfile* getDisplayColorProfile() const = 0;
diff --git a/services/surfaceflinger/CompositionEngine/include/compositionengine/impl/Output.h b/services/surfaceflinger/CompositionEngine/include/compositionengine/impl/Output.h
index 6f25e6391b..ef9684c9c1 100644
--- a/services/surfaceflinger/CompositionEngine/include/compositionengine/impl/Output.h
+++ b/services/surfaceflinger/CompositionEngine/include/compositionengine/impl/Output.h
@@ -51,6 +51,7 @@ public:
 
     const std::string& getName() const override;
     void setName(const std::string&) override;
+    void setSingleLayerCount(int) override;
 
     compositionengine::DisplayColorProfile* getDisplayColorProfile() const override;
     void setDisplayColorProfile(std::unique_ptr<compositionengine::DisplayColorProfile>) override;
@@ -130,6 +131,7 @@ private:
     ReleasedLayers mReleasedLayers;
     OutputLayer* mLayerRequestingBackgroundBlur = nullptr;
     std::unique_ptr<ClientCompositionRequestCache> mClientCompositionRequestCache;
+    int32_t mSingleLayerCount = 0;
 };
 
 // This template factory function standardizes the implementation details of the
diff --git a/services/surfaceflinger/CompositionEngine/src/Output.cpp b/services/surfaceflinger/CompositionEngine/src/Output.cpp
index 90ac781a0c..74123eafe2 100644
--- a/services/surfaceflinger/CompositionEngine/src/Output.cpp
+++ b/services/surfaceflinger/CompositionEngine/src/Output.cpp
@@ -41,6 +41,7 @@
 #include <ui/DebugUtils.h>
 #include <ui/HdrCapabilities.h>
 #include <utils/Trace.h>
+#include <cutils/properties.h>
 
 #include "TracedOrdinal.h"
 
@@ -95,6 +96,10 @@ void Output::setName(const std::string& name) {
     mName = name;
 }
 
+void Output::setSingleLayerCount(int value) {
+    mSingleLayerCount = value;
+}
+
 void Output::setCompositionEnabled(bool enabled) {
     auto& outputState = editState();
     if (outputState.isEnabled == enabled) {
@@ -922,8 +927,21 @@ std::optional<base::unique_fd> Output::composeSurfaces(
                    });
 
     const nsecs_t renderEngineStart = systemTime();
+    char tempChar[PROPERTY_VALUE_MAX];
+    int bypassCompositionSwitch = property_get("ro.hardware.compositionBypass", tempChar, "0") > 0 ? std::atoi(tempChar) : 0;
+    int offset = property_get("ro.hardware.compositionBypass.offset", tempChar, "0") > 0 ? std::atoi(tempChar) : 0;
+    size_t layerCount = clientCompositionLayerPointers.size();
+    if ((bypassCompositionSwitch == 1) && (layerCount == 1)) { // 开关打开，且仅有一个图层
+        mSingleLayerCount++;
+        if (mSingleLayerCount > offset) {
+            return readyFence;
+        }
+    } else {
+        mSingleLayerCount = 0;
+    }
+
     status_t status =
-            renderEngine.drawLayers(clientCompositionDisplay, clientCompositionLayerPointers,
+        renderEngine.drawLayers(clientCompositionDisplay, clientCompositionLayerPointers,
                                     buf->getNativeBuffer(), /*useFramebufferCache=*/true,
                                     std::move(fd), &readyFence);
 
diff --git a/services/surfaceflinger/DisplayDevice.cpp b/services/surfaceflinger/DisplayDevice.cpp
index 730f29744a..630072344b 100644
--- a/services/surfaceflinger/DisplayDevice.cpp
+++ b/services/surfaceflinger/DisplayDevice.cpp
@@ -116,6 +116,10 @@ void DisplayDevice::setDisplayName(const std::string& displayName) {
     }
 }
 
+void DisplayDevice::setSingleLayerCount(int value) {
+    mCompositionDisplay->setSingleLayerCount(value);
+}
+
 uint32_t DisplayDevice::getPageFlipCount() const {
     return mCompositionDisplay->getRenderSurface()->getPageFlipCount();
 }
diff --git a/services/surfaceflinger/DisplayDevice.h b/services/surfaceflinger/DisplayDevice.h
index cb467ea292..2d6d02bc72 100644
--- a/services/surfaceflinger/DisplayDevice.h
+++ b/services/surfaceflinger/DisplayDevice.h
@@ -134,6 +134,7 @@ public:
     const Rect& bounds() const { return getBounds(); }
 
     void setDisplayName(const std::string& displayName);
+    void setSingleLayerCount(int value);
     const std::string& getDisplayName() const { return mDisplayName; }
 
     /* ------------------------------------------------------------------------
diff --git a/services/surfaceflinger/Scheduler/DispSync.cpp b/services/surfaceflinger/Scheduler/DispSync.cpp
index ff91bf7bc0..db12328ef6 100644
--- a/services/surfaceflinger/Scheduler/DispSync.cpp
+++ b/services/surfaceflinger/Scheduler/DispSync.cpp
@@ -656,6 +656,14 @@ status_t DispSync::changePhaseOffset(Callback* callback, nsecs_t phase) {
 void DispSync::setPeriod(nsecs_t period) {
     Mutex::Autolock lock(mMutex);
 
+    {
+        /* kbox directly modifies mPeriod */
+        mPeriod = period;
+        mPhase = 0;
+        mReferenceTime = 0;
+        mThread->updateModel(mPeriod, mPhase, mReferenceTime);
+    }
+
     const bool pendingPeriodShouldChange =
             period != mIntendedPeriod || (period == mIntendedPeriod && mPendingPeriod != 0);
 
diff --git a/services/surfaceflinger/Scheduler/EventThread.cpp b/services/surfaceflinger/Scheduler/EventThread.cpp
index 845bf50ad3..93ef71d972 100644
--- a/services/surfaceflinger/Scheduler/EventThread.cpp
+++ b/services/surfaceflinger/Scheduler/EventThread.cpp
@@ -42,6 +42,9 @@
 #include "EventThread.h"
 #include "HwcStrongTypes.h"
 
+#include <cutils/properties.h>
+#include <dlfcn.h>
+
 using namespace std::chrono_literals;
 
 namespace android {
@@ -164,6 +167,65 @@ EventThread::~EventThread() = default;
 
 namespace impl {
 
+void EventThread::RenderAccDeinit()
+{
+    renderAccInitialize = nullptr;
+    renderAccDeinit = nullptr;
+    renderAccJudge = nullptr;
+    renderAccQueueBuffer = nullptr;
+    renderAccProc = nullptr;
+    
+    if (renderAccLibHdl) {
+        dlclose(renderAccLibHdl);
+    }
+    renderAccLibHdl = nullptr;
+}
+
+#define RENDERACC_SO_PATH "/system/vendor/lib64/hw/AdaptiveVsync.kbox.so"
+void EventThread::RenderAccInitialize()
+{
+    char value[PROPERTY_VALUE_MAX];
+    property_get("ro.vmi.adaptive.vsync", value, "0");
+    bool adaptiveVsyncFlag = (atoi(value) == 1);
+    ALOGI_IF(adaptiveVsyncFlag, "AdaptiveVsync: EventThread[%s] Enabling adaptive vsync.", mThreadName);
+
+    if (!adaptiveVsyncFlag) {
+        return;
+    }
+
+    void *renderAccLibHdl = dlopen(RENDERACC_SO_PATH, RTLD_LAZY);
+    if (renderAccLibHdl == nullptr) {
+        ALOGE("AdaptiveVsync: open AdaptiveVsync.kbox.so failed. path %s error %s", RENDERACC_SO_PATH, dlerror());
+        return;
+    }
+
+#define GET_PROXY_FROM_LIB(func, proxy, funcName) \
+    func = reinterpret_cast<proxy>(dlsym(renderAccLibHdl, #funcName)); \
+    if (func == nullptr) { \
+        ALOGE("AdaptiveVsync: Get %s function fail. error %s", #funcName, dlerror()); \
+    }
+
+    GET_PROXY_FROM_LIB(renderAccInitialize, RenderAccInitializeProxy, AdaptiveVsyncInit);
+    GET_PROXY_FROM_LIB(renderAccDeinit, RenderAccDeinitProxy, AdaptiveVsyncDeinit);
+    GET_PROXY_FROM_LIB(renderAccJudge, RenderAccJudgeProxy, IsCloseAdaptiveVsync);
+    GET_PROXY_FROM_LIB(renderAccQueueBuffer, RenderAccQueueBufferProxy, AdaptiveVsyncOnQueueBuffer);
+    GET_PROXY_FROM_LIB(renderAccProc, RenderAccProcProxy, AdaptiveVsyncOnVsync);
+
+    if ((renderAccInitialize == nullptr) || (renderAccDeinit == nullptr) || (renderAccJudge == nullptr) || (renderAccQueueBuffer == nullptr) || (renderAccProc == nullptr)) {
+        RenderAccDeinit();
+        return;
+    }
+
+    data = renderAccInitialize();
+    if (!data) {
+        ALOGE("AdaptiveVsync: renderAccInitialize failed");
+        RenderAccDeinit();
+        return;
+    }
+
+    ALOGI("AdaptiveVsync: renderAccInitialize success");
+}
+
 EventThread::EventThread(std::unique_ptr<VSyncSource> vsyncSource,
                          InterceptVSyncsCallback interceptVSyncsCallback)
       : mVSyncSource(std::move(vsyncSource)),
@@ -188,12 +250,21 @@ EventThread::EventThread(std::unique_ptr<VSyncSource> vsyncSource,
         ALOGE("Couldn't set SCHED_FIFO for EventThread");
     }
 
+    RenderAccInitialize();
+
     set_sched_policy(tid, SP_FOREGROUND);
 }
 
 EventThread::~EventThread() {
     mVSyncSource->setCallback(nullptr);
 
+    if (adaptiveVsyncFlag) {
+        renderAccDeinit(data); // 先释放申请的内存
+        data = nullptr;
+
+        RenderAccDeinit(); // 再清理EventThread中的资源
+    }
+
     {
         std::lock_guard<std::mutex> lock(mMutex);
         mState = State::Quit;
@@ -288,12 +359,29 @@ void EventThread::onScreenAcquired() {
 void EventThread::onVSyncEvent(nsecs_t timestamp, nsecs_t expectedVSyncTimestamp) {
     std::lock_guard<std::mutex> lock(mMutex);
 
+    if (adaptiveVsyncFlag && (mThreadName == "app") && renderAccJudge && data) {
+        nsecs_t now = systemTime(SYSTEM_TIME_MONOTONIC);
+        renderAccJudge(data, now);
+    }
+
     LOG_FATAL_IF(!mVSyncState);
     mPendingEvents.push_back(makeVSync(mVSyncState->displayId, timestamp, ++mVSyncState->count,
                                        expectedVSyncTimestamp));
     mCondition.notify_all();
 }
 
+void EventThread::postVsync(nsecs_t time) {
+    if (adaptiveVsyncFlag && renderAccProc && data) {
+        renderAccProc(data, EventThread::CallOnVsync, static_cast<void *>(this), time);
+    }
+}
+
+void EventThread::notifyPostVsync() {
+    if (adaptiveVsyncFlag && (mThreadName == "app") &&  renderAccQueueBuffer && data) {
+        renderAccQueueBuffer(data);
+    }
+}
+
 void EventThread::onHotplugReceived(PhysicalDisplayId displayId, bool connected) {
     std::lock_guard<std::mutex> lock(mMutex);
 
diff --git a/services/surfaceflinger/Scheduler/EventThread.h b/services/surfaceflinger/Scheduler/EventThread.h
index 49f624c35e..e363fab4da 100644
--- a/services/surfaceflinger/Scheduler/EventThread.h
+++ b/services/surfaceflinger/Scheduler/EventThread.h
@@ -126,6 +126,9 @@ public:
 
     // Retrieves the number of event connections tracked by this EventThread.
     virtual size_t getEventThreadConnectionCount() = 0;
+
+    virtual void postVsync(nsecs_t time) = 0;
+    virtual void notifyPostVsync() = 0;
 };
 
 namespace impl {
@@ -161,6 +164,9 @@ public:
 
     size_t getEventThreadConnectionCount() override;
 
+    void postVsync(nsecs_t time) override;
+    void notifyPostVsync() override;
+
 private:
     friend EventThreadTest;
 
@@ -219,6 +225,29 @@ private:
     State mState GUARDED_BY(mMutex) = State::Idle;
 
     static const char* toCString(State);
+
+    using OnVsyncProxy = void (*)(void *obj, nsecs_t timestamp, nsecs_t expectedVSyncTimestamp);
+    static void CallOnVsync(void *obj, nsecs_t timestamp, nsecs_t expectedVSyncTimestamp) {
+        static_cast<EventThread *>(obj)->onVSyncEvent(timestamp, expectedVSyncTimestamp);
+    }
+
+    void RenderAccInitialize();
+    void RenderAccDeinit();
+    using RenderAccInitializeProxy = void * (*)();
+    using RenderAccDeinitProxy = void (*)(void *ptr);
+    using RenderAccJudgeProxy = void (*)(void *ptr, int64_t now);
+    using RenderAccQueueBufferProxy = void (*)(void *ptr);
+    using RenderAccProcProxy = void (*)(void *ptr, OnVsyncProxy callOnVsync, void *obj, int64_t timestamp);
+
+    void *renderAccLibHdl = nullptr;
+    bool adaptiveVsyncFlag = false;
+    void *data = nullptr;
+
+    RenderAccInitializeProxy  renderAccInitialize = nullptr;
+    RenderAccDeinitProxy      renderAccDeinit = nullptr;
+    RenderAccJudgeProxy       renderAccJudge = nullptr;
+    RenderAccQueueBufferProxy renderAccQueueBuffer = nullptr;
+    RenderAccProcProxy        renderAccProc = nullptr;
 };
 
 // ---------------------------------------------------------------------------
diff --git a/services/surfaceflinger/Scheduler/Scheduler.cpp b/services/surfaceflinger/Scheduler/Scheduler.cpp
index 5c0ba01cd3..cab3b862b5 100644
--- a/services/surfaceflinger/Scheduler/Scheduler.cpp
+++ b/services/surfaceflinger/Scheduler/Scheduler.cpp
@@ -334,6 +334,12 @@ void Scheduler::disableHardwareVsync(bool makeUnavailable) {
     }
 }
 
+void Scheduler::dynamicFpsAdjust(nsecs_t period) {
+    ALOGI("dynamicFpsAdjust period %" PRId64, period);
+    downVsyncPeriod.store(period);
+    resyncToHardwareVsync(true, period);
+}
+
 void Scheduler::resyncToHardwareVsync(bool makeAvailable, nsecs_t period) {
     {
         std::lock_guard<std::mutex> lock(mHWVsyncLock);
@@ -360,7 +366,11 @@ void Scheduler::resync() {
     const nsecs_t last = mLastResyncTime.exchange(now);
 
     if (now - last > kIgnoreDelay) {
-        resyncToHardwareVsync(false, mRefreshRateConfigs.getCurrentRefreshRate().getVsyncPeriod());
+        nsecs_t vsyncPeriod = downVsyncPeriod.load();
+        if (vsyncPeriod == 0) { 
+            vsyncPeriod = mRefreshRateConfigs.getCurrentRefreshRate().getVsyncPeriod();
+        }
+        resyncToHardwareVsync(false, vsyncPeriod);
     }
 }
 
@@ -708,4 +718,14 @@ void Scheduler::onPrimaryDisplayAreaChanged(uint32_t displayArea) {
     }
 }
 
+void Scheduler::postVsync(ConnectionHandle handle, nsecs_t time) {
+    RETURN_IF_INVALID_HANDLE(handle);
+    mConnections[handle].thread->postVsync(time);
+}
+
+void Scheduler::notifyPostVsync(ConnectionHandle handle) {
+    RETURN_IF_INVALID_HANDLE(handle);
+    mConnections[handle].thread->notifyPostVsync();
+}
+
 } // namespace android
diff --git a/services/surfaceflinger/Scheduler/Scheduler.h b/services/surfaceflinger/Scheduler/Scheduler.h
index 730ea8f0c8..dfea3454c7 100644
--- a/services/surfaceflinger/Scheduler/Scheduler.h
+++ b/services/surfaceflinger/Scheduler/Scheduler.h
@@ -120,6 +120,7 @@ public:
     // no-op.
     // The period is the vsync period from the current display configuration.
     void resyncToHardwareVsync(bool makeAvailable, nsecs_t period);
+    void dynamicFpsAdjust(nsecs_t period);
     void resync();
 
     // Passes a vsync sample to DispSync. periodFlushed will be true if
@@ -163,6 +164,9 @@ public:
 
     size_t getEventThreadConnectionCount(ConnectionHandle handle);
 
+    void postVsync(ConnectionHandle handle, nsecs_t time);
+    void notifyPostVsync(ConnectionHandle handle);
+
 private:
     friend class TestableScheduler;
 
@@ -223,6 +227,7 @@ private:
     bool mHWVsyncAvailable GUARDED_BY(mHWVsyncLock) = false;
 
     std::atomic<nsecs_t> mLastResyncTime = 0;
+    std::atomic<nsecs_t> downVsyncPeriod = 0;
 
     // Whether to use idle timer callbacks that support the kernel timer.
     const bool mSupportKernelTimer;
diff --git a/services/surfaceflinger/SurfaceFlinger.cpp b/services/surfaceflinger/SurfaceFlinger.cpp
index 79f99cc5d5..e4289bbc9f 100644
--- a/services/surfaceflinger/SurfaceFlinger.cpp
+++ b/services/surfaceflinger/SurfaceFlinger.cpp
@@ -85,6 +85,7 @@
 #include <cinttypes>
 #include <cmath>
 #include <cstdint>
+#include <cstdlib>
 #include <functional>
 #include <mutex>
 #include <optional>
@@ -430,6 +431,10 @@ SurfaceFlinger::SurfaceFlinger(Factory& factory) : SurfaceFlinger(factory, SkipI
     property_get("debug.sf.disable_client_composition_cache", value, "0");
     mDisableClientCompositionCache = atoi(value);
 
+    property_get("ro.vmi.adaptive.vsync", value, "0");
+    misEnableAdaptiveVsync = (atoi(value) == 1);
+    ALOGI_IF(misEnableAdaptiveVsync, "AdaptiveVsync: Enabling adaptive vsync.");
+
     // We should be reading 'persist.sys.sf.color_saturation' here
     // but since /data may be encrypted, we need to wait until after vold
     // comes online to attempt to read the property. The property is
@@ -1630,11 +1635,42 @@ void SurfaceFlinger::changeRefreshRateLocked(const RefreshRate& refreshRate,
     setDesiredActiveConfig({refreshRate.getConfigId(), event});
 }
 
+#define VSYNC_PERIOD_VALUE_BASE 1000000000
+#define MAX_FPS 300
+#define DEFAULT_FPS 30
+static nsecs_t getPeriodFromProp(std::string prop)
+{
+    char tmpFps[PROPERTY_VALUE_MAX] = {0};
+    property_get(prop.c_str(), tmpFps, "30");
+    std::stringstream stream;
+    stream << tmpFps;
+    int fps;
+    stream >> fps;
+
+    if (fps <= 0 || fps > MAX_FPS) {
+        fps = DEFAULT_FPS;
+    }
+    return VSYNC_PERIOD_VALUE_BASE / fps; // 1000*1000*1000/30
+}
+
 void SurfaceFlinger::onHotplugReceived(int32_t sequenceId, hal::HWDisplayId hwcDisplayId,
                                        hal::Connection connection) {
-    ALOGV("%s(%d, %" PRIu64 ", %s)", __FUNCTION__, sequenceId, hwcDisplayId,
+    ALOGI("onHotplugReceived: %s(%d, %" PRIu64 ", %s)", __FUNCTION__, sequenceId, hwcDisplayId,
           connection == hal::Connection::CONNECTED ? "connected" : "disconnected");
 
+    if (hwcDisplayId == HWC_DISPLAY_KBOX) {
+        nsecs_t period = getPeriodFromProp("ro.hardware.fps");
+        if (connection == hal::Connection::CONNECTED) {
+            mScheduler->dynamicFpsAdjust(period);
+        } else if (connection == hal::Connection::DISCONNECTED) {
+            nsecs_t downPeriod = getPeriodFromProp("ro.hardware.downfps");
+            downPeriod = (downPeriod > period) ? downPeriod : period; // 约束下降后的周期一定大于原始周期
+            mScheduler->dynamicFpsAdjust(downPeriod);
+            mScheduler->disableHardwareVsync(true);
+        }
+        return;
+    }
+
     // Ignore events that do not have the right sequenceId.
     if (sequenceId != getBE().mComposerSequenceId) {
         return;
@@ -1676,6 +1712,9 @@ void SurfaceFlinger::onRefreshReceived(int sequenceId, hal::HWDisplayId /*hwcDis
     if (sequenceId != getBE().mComposerSequenceId) {
         return;
     }
+    for (const auto& [token, display] : mDisplays) {
+        display->setSingleLayerCount(0);
+    }
     repaintEverythingForHWC();
 }
 
@@ -2697,12 +2736,6 @@ void SurfaceFlinger::processDisplayChanged(const wp<IBinder>& displayToken,
         if (currentState.layerStack != drawingState.layerStack) {
             display->setLayerStack(currentState.layerStack);
         }
-        if ((currentState.orientation != drawingState.orientation) ||
-            (currentState.viewport != drawingState.viewport) ||
-            (currentState.frame != drawingState.frame)) {
-            display->setProjection(currentState.orientation, currentState.viewport,
-                                   currentState.frame);
-        }
         if (currentState.width != drawingState.width ||
             currentState.height != drawingState.height) {
             display->setDisplaySize(currentState.width, currentState.height);
@@ -2715,6 +2748,13 @@ void SurfaceFlinger::processDisplayChanged(const wp<IBinder>& displayToken,
                 mRefreshRateOverlay->setViewport(display->getSize());
             }
         }
+
+        if ((currentState.orientation != drawingState.orientation) ||
+            (currentState.viewport != drawingState.viewport) ||
+            (currentState.frame != drawingState.frame)) {
+            display->setProjection(currentState.orientation, currentState.viewport,
+                                   currentState.frame);
+        }
     }
 }
 
@@ -3007,6 +3047,21 @@ void SurfaceFlinger::initScheduler(DisplayId primaryDisplayId) {
                                               currentConfig, vsyncPeriod);
 }
 
+void SurfaceFlinger::postVsync()
+{
+    nsecs_t now = systemTime();
+    mScheduler->postVsync(mSfConnectionHandle, now);
+    mScheduler->notifyPostVsync(mAppConnectionHandle);
+}
+
+bool SurfaceFlinger::isEnableAdaptiveVsync() const
+{
+    char value[PROPERTY_VALUE_MAX];
+    property_get("vmi.enable.adaptive.vsync", value, "1");
+    bool isEnableAdaptiveVsync = atoi(value);
+    return (isEnableAdaptiveVsync && misEnableAdaptiveVsync);
+}
+
 void SurfaceFlinger::commitTransaction()
 {
     commitTransactionLocked();
diff --git a/services/surfaceflinger/SurfaceFlinger.h b/services/surfaceflinger/SurfaceFlinger.h
index c727574780..51f2af510f 100644
--- a/services/surfaceflinger/SurfaceFlinger.h
+++ b/services/surfaceflinger/SurfaceFlinger.h
@@ -336,6 +336,10 @@ public:
     // debug.sf.disable_client_composition_cache
     bool mDisableClientCompositionCache = false;
 
+    // adaptiveVsync
+    void postVsync();
+    bool isEnableAdaptiveVsync() const;
+
 private:
     friend class BufferLayer;
     friend class BufferQueueLayer;
@@ -1296,6 +1300,8 @@ private:
     int mFrameRateFlexibilityTokenCount = 0;
 
     sp<IBinder> mDebugFrameRateFlexibilityToken;
+
+    bool misEnableAdaptiveVsync = false;
 };
 
 } // namespace android
diff --git a/vulkan/libvulkan/driver.cpp b/vulkan/libvulkan/driver.cpp
index f840561292..83e01f522d 100644
--- a/vulkan/libvulkan/driver.cpp
+++ b/vulkan/libvulkan/driver.cpp
@@ -245,6 +245,12 @@ bool Hal::Open() {
     if (result == -ENOENT) {
         result = LoadBuiltinDriver(&module);
     }
+
+    if (result == -ENOENT) {
+        // ENOENT here means sphal is not enabled for vulkan driver, open hal module directly to walkaround
+        result = hw_get_module(HWVULKAN_HARDWARE_MODULE_ID, reinterpret_cast<const hw_module_t**>(&module));
+    }
+
     if (result != 0) {
         android::GraphicsEnv::getInstance().setDriverLoaded(
             android::GpuStatsInfo::Api::API_VK, false, systemTime() - openTime);
