diff --git a/lmkd.cpp b/lmkd.cpp
index 3515fdb..c240c58 100644
--- a/lmkd.cpp
+++ b/lmkd.cpp
@@ -37,6 +37,7 @@
 #include <sys/time.h>
 #include <sys/types.h>
 #include <time.h>
+#include <ctime>
 #include <unistd.h>
 
 #include <cutils/properties.h>
@@ -81,6 +82,12 @@
 #define MEMCG_MEMORYSW_USAGE "/dev/memcg/memory.memsw.usage_in_bytes"
 #define ZONEINFO_PATH "/proc/zoneinfo"
 #define MEMINFO_PATH "/proc/meminfo"
+
+#define CONTAINERID_PATH "/data/containerid"
+#define CONTAINERRENDER_PATH "/data/renderpath"
+#define VMEMINFO_PATH "/sys/kernel/debug/gpu_statistics"
+#define MEM_LIMT 40
+#define PARSE_LINE_BREAK "\n"
 #define VMSTAT_PATH "/proc/vmstat"
 #define PROC_STATUS_TGID_FIELD "Tgid:"
 #define LINE_MAX 128
@@ -334,21 +341,18 @@ struct zoneinfo_zone {
 enum zoneinfo_node_field {
     ZI_NODE_NR_INACTIVE_FILE = 0,
     ZI_NODE_NR_ACTIVE_FILE,
-    ZI_NODE_WORKINGSET_REFAULT,
     ZI_NODE_FIELD_COUNT
 };
 
 static const char* const zoneinfo_node_field_names[ZI_NODE_FIELD_COUNT] = {
     "nr_inactive_file",
     "nr_active_file",
-    "workingset_refault",
 };
 
 union zoneinfo_node_fields {
     struct {
         int64_t nr_inactive_file;
         int64_t nr_active_file;
-        int64_t workingset_refault;
     } field;
     int64_t arr[ZI_NODE_FIELD_COUNT];
 };
@@ -361,7 +365,7 @@ struct zoneinfo_node {
 };
 
 /* for now two memory nodes is more than enough */
-#define MAX_NR_NODES 2
+#define MAX_NR_NODES 4
 
 struct zoneinfo {
     int node_count;
@@ -369,7 +373,6 @@ struct zoneinfo {
     int64_t totalreserve_pages;
     int64_t total_inactive_file;
     int64_t total_active_file;
-    int64_t total_workingset_refault;
 };
 
 /* Fields to parse in /proc/meminfo */
@@ -451,6 +454,7 @@ enum vmstat_field {
     VS_INACTIVE_FILE,
     VS_ACTIVE_FILE,
     VS_WORKINGSET_REFAULT,
+    VS_WORKINGSET_REFAULT_FILE,
     VS_PGSCAN_KSWAPD,
     VS_PGSCAN_DIRECT,
     VS_PGSCAN_DIRECT_THROTTLE,
@@ -462,6 +466,7 @@ static const char* const vmstat_field_names[MI_FIELD_COUNT] = {
     "nr_inactive_file",
     "nr_active_file",
     "workingset_refault",
+    "workingset_refault_file",
     "pgscan_kswapd",
     "pgscan_direct",
     "pgscan_direct_throttle",
@@ -473,6 +478,7 @@ union vmstat {
         int64_t nr_inactive_file;
         int64_t nr_active_file;
         int64_t workingset_refault;
+        int64_t workingset_refault_file;
         int64_t pgscan_kswapd;
         int64_t pgscan_direct;
         int64_t pgscan_direct_throttle;
@@ -516,6 +522,11 @@ static struct adjslot_list procadjslot_list[ADJTOSLOT_COUNT];
 
 #define MAX_DISTINCT_OOM_ADJ 32
 #define KILLCNT_INVALID_IDX 0xFF
+#define BUFFER_SIZE_128 128
+#define BUFFER_SIZE_256 256
+#define BUFFER_SIZE_13 13
+#define FCAST_SPACE 512
+
 /*
  * Because killcnt array is sparse a two-level indirection is used
  * to keep the size small. killcnt_idx stores index of the element in
@@ -537,6 +548,14 @@ static int clamp(int low, int high, int value) {
     return max(min(value, high), low);
 }
 
+//用于关闭文件描述符,并将文件描述符置为-1
+static void close_fd(int& fd) {
+    if (fd > 0) {
+        close(fd);
+        fd = -1;
+    }
+}
+
 static bool parse_int64(const char* str, int64_t* ret) {
     char* endptr;
     long long val = strtoll(str, &endptr, 10);
@@ -1563,8 +1582,8 @@ static void zoneinfo_parse_protection(char *buf, struct zoneinfo_zone *zone) {
 }
 
 static int zoneinfo_parse_zone(char **buf, struct zoneinfo_zone *zone) {
-    for (char *line = strtok_r(NULL, "\n", buf); line;
-         line = strtok_r(NULL, "\n", buf)) {
+    for (char *line = strtok_r(NULL, PARSE_LINE_BREAK, buf); line;
+         line = strtok_r(NULL, PARSE_LINE_BREAK, buf)) {
         char *cp;
         char *ap;
         char *save_ptr;
@@ -1617,8 +1636,8 @@ static int zoneinfo_parse_zone(char **buf, struct zoneinfo_zone *zone) {
 static int zoneinfo_parse_node(char **buf, struct zoneinfo_node *node) {
     int fields_to_match = ZI_NODE_FIELD_COUNT;
 
-    for (char *line = strtok_r(NULL, "\n", buf); line;
-         line = strtok_r(NULL, "\n", buf)) {
+    for (char *line = strtok_r(NULL, PARSE_LINE_BREAK, buf); line;
+         line = strtok_r(NULL, PARSE_LINE_BREAK, buf)) {
         char *cp;
         char *ap;
         char *save_ptr;
@@ -1671,8 +1690,8 @@ static int zoneinfo_parse(struct zoneinfo *zi) {
         return -1;
     }
 
-    for (line = strtok_r(buf, "\n", &save_ptr); line;
-         line = strtok_r(NULL, "\n", &save_ptr)) {
+    for (line = strtok_r(buf, PARSE_LINE_BREAK, &save_ptr); line;
+         line = strtok_r(NULL, PARSE_LINE_BREAK, &save_ptr)) {
         int node_id;
         if (sscanf(line, "Node %d, zone %" STRINGIFY(LINE_MAX) "s", &node_id, zone_name) == 2) {
             if (!node || node->id != node_id) {
@@ -1719,7 +1738,6 @@ static int zoneinfo_parse(struct zoneinfo *zi) {
         }
         zi->total_inactive_file += node->fields.field.nr_inactive_file;
         zi->total_active_file += node->fields.field.nr_active_file;
-        zi->total_workingset_refault += node->fields.field.workingset_refault;
     }
     return 0;
 }
@@ -1766,8 +1784,8 @@ static int meminfo_parse(union meminfo *mi) {
         return -1;
     }
 
-    for (line = strtok_r(buf, "\n", &save_ptr); line;
-         line = strtok_r(NULL, "\n", &save_ptr)) {
+    for (line = strtok_r(buf, PARSE_LINE_BREAK, &save_ptr); line;
+         line = strtok_r(NULL, PARSE_LINE_BREAK, &save_ptr)) {
         if (!meminfo_parse_line(line, mi)) {
             ALOGE("%s parse error", file_data.filename);
             return -1;
@@ -1821,8 +1839,8 @@ static int vmstat_parse(union vmstat *vs) {
         return -1;
     }
 
-    for (line = strtok_r(buf, "\n", &save_ptr); line;
-         line = strtok_r(NULL, "\n", &save_ptr)) {
+    for (line = strtok_r(buf, PARSE_LINE_BREAK, &save_ptr); line;
+         line = strtok_r(NULL, PARSE_LINE_BREAK, &save_ptr)) {
         if (!vmstat_parse_line(line, vs)) {
             ALOGE("%s parse error", file_data.filename);
             return -1;
@@ -2341,6 +2359,7 @@ static void mp_event_psi(int data, uint32_t events, struct polling_params *poll_
     bool cut_thrashing_limit = false;
     int min_score_adj = 0;
     long since_thrashing_reset_ms;
+    int64_t workingset_refault_file;
 
     if (clock_gettime(CLOCK_MONOTONIC_COARSE, &curr_tm) != 0) {
         ALOGE("Failed to get current time");
@@ -2366,6 +2385,8 @@ static void mp_event_psi(int data, uint32_t events, struct polling_params *poll_
         ALOGE("Failed to parse vmstat!");
         return;
     }
+    /* Starting 5.9 kernel workingset_refault vmstat field was renamed workingset_refault_file */
+    workingset_refault_file = vs.field.workingset_refault ? : vs.field.workingset_refault_file;
 
     if (meminfo_parse(&mi) < 0) {
         ALOGE("Failed to parse meminfo!");
@@ -2378,7 +2399,7 @@ static void mp_event_psi(int data, uint32_t events, struct polling_params *poll_
         cycle_after_kill = true;
         /* Reset file-backed pagecache size and refault amounts after a kill */
         base_file_lru = vs.field.nr_inactive_file + vs.field.nr_active_file;
-        init_ws_refault = vs.field.workingset_refault;
+        init_ws_refault = workingset_refault_file;
         thrashing_reset_tm = curr_tm;
         prev_thrash_growth = 0;
     }
@@ -2388,7 +2409,7 @@ static void mp_event_psi(int data, uint32_t events, struct polling_params *poll_
         if (!swap_low_threshold) {
             swap_low_threshold = mi.field.total_swap * swap_free_low_percentage / 100;
         }
-        swap_is_low = mi.field.free_swap < swap_low_threshold;
+        swap_is_low = mi.field.free_swap <= swap_low_threshold;
     }
 
     /* Identify reclaim state */
@@ -2399,12 +2420,12 @@ static void mp_event_psi(int data, uint32_t events, struct polling_params *poll_
     } else if (vs.field.pgscan_kswapd > init_pgscan_kswapd) {
         init_pgscan_kswapd = vs.field.pgscan_kswapd;
         reclaim = KSWAPD_RECLAIM;
-    } else if (vs.field.workingset_refault == prev_workingset_refault) {
+    } else if (workingset_refault_file == prev_workingset_refault) {
         /* Device is not thrashing and not reclaiming, bail out early until we see these stats changing*/
         goto no_kill;
     }
 
-    prev_workingset_refault = vs.field.workingset_refault;
+    prev_workingset_refault = workingset_refault_file;
 
      /*
      * It's possible we fail to find an eligible process to kill (ex. no process is
@@ -2421,7 +2442,7 @@ static void mp_event_psi(int data, uint32_t events, struct polling_params *poll_
     if (since_thrashing_reset_ms > THRASHING_RESET_INTERVAL_MS) {
         long windows_passed;
         /* Calculate prev_thrash_growth if we crossed THRASHING_RESET_INTERVAL_MS */
-        prev_thrash_growth = (vs.field.workingset_refault - init_ws_refault) * 100
+        prev_thrash_growth = (workingset_refault_file - init_ws_refault) * 100
                             / (base_file_lru + 1);
         windows_passed = (since_thrashing_reset_ms / THRASHING_RESET_INTERVAL_MS);
         /*
@@ -2435,12 +2456,12 @@ static void mp_event_psi(int data, uint32_t events, struct polling_params *poll_
 
         /* Record file-backed pagecache size when crossing THRASHING_RESET_INTERVAL_MS */
         base_file_lru = vs.field.nr_inactive_file + vs.field.nr_active_file;
-        init_ws_refault = vs.field.workingset_refault;
+        init_ws_refault = workingset_refault_file;
         thrashing_reset_tm = curr_tm;
         thrashing_limit = thrashing_limit_pct;
     } else {
         /* Calculate what % of the file-backed pagecache refaulted so far */
-        thrashing = (vs.field.workingset_refault - init_ws_refault) * 100 / (base_file_lru + 1);
+        thrashing = (workingset_refault_file - init_ws_refault) * 100 / (base_file_lru + 1);
     }
     /* Add previous cycle's decayed thrashing amount */
     thrashing += prev_thrash_growth;
@@ -2659,7 +2680,7 @@ static void mp_event_common(int data, uint32_t events, struct polling_params *po
         kill_skip_count = 0;
     }
 
-    if (meminfo_parse(&mi) < 0 || zoneinfo_parse(&zi) < 0) {
+    if (meminfo_parse(&mi) < 0) {
         ALOGE("Failed to get free memory!");
         return;
     }
@@ -3101,6 +3122,8 @@ static int init(void) {
         pidfd_supported = true;
         close(pidfd);
     }
+    property_set("sys.lmk.time_difference","30");
+    property_set("sys.lmk.memlog_activationflag","1");
     ALOGI("Process polling is %s", pidfd_supported ? "supported" : "not supported" );
 
     return 0;
@@ -3154,12 +3177,136 @@ static void call_handler(struct event_handler_info* handler_info,
     }
 }
 
+//该接口非线程安全，不支持并发调用 forecast_space提供给共用此函数读取文件的调用，自定义预估读取字节。
+static char* read_file(int& file_fd, int forecast_space = 0) {
+    static ssize_t buf_size = PAGE_SIZE; 
+    ssize_t size;
+    static char *new_buf, *buf = NULL;
+    
+    if (buf == NULL && (buf = static_cast<char*>(malloc(buf_size))) == NULL) {
+        ALOGE("open vmem temp sapce fail");
+        return NULL;
+    }   
+    while (true) {
+        size = read_all(file_fd, buf, forecast_space <= 0 ? buf_size - 1 : forecast_space - 1);
+        if (size < 0) {
+            ALOGE("Read vmemflie fail");
+            close_fd(file_fd);
+            free(buf);
+            buf = NULL;
+            return NULL;
+        }
+        // 过量使用会进行减少内存
+        if (buf_size != PAGE_SIZE && (buf_size >= MEM_LIMT * size)) {
+            free(buf);
+            new_buf = NULL;
+            buf = NULL;
+            buf_size = PAGE_SIZE;
+            if ((buf = static_cast<char*>(malloc(buf_size))) == NULL) {
+                ALOGE("open vmem temp sapce fail");
+                return NULL;
+            }
+            continue;
+        }
+        if (size < buf_size - 1) {
+            break;
+        }
+        if ((new_buf = static_cast<char*>(realloc(buf, buf_size * 2))) == NULL) {
+            ALOGE("realloc fail!");
+            return NULL;
+        }
+        buf = new_buf;
+        buf_size *= 2;
+    }
+    return buf;
+}
+
+static void get_mem_info(char* m_buf, char* m_line) {
+    // 不采用strstr,字符串匹配会带来性能损失，确定了需要的内容在 1 2 3 15 16行。直接按需读取即可
+    char *line, *rest = NULL;
+    int lineCount = 1;
+    line = strtok_r(m_buf, PARSE_LINE_BREAK, &rest);
+    while (line != NULL) {
+        if (lineCount <= 3 || lineCount == 15) {
+            snprintf(m_line + strlen(m_line), BUFFER_SIZE_256 - strlen(m_line) - 1, "%s|", line);
+        } else if (lineCount == 16) {
+            snprintf(m_line + strlen(m_line), BUFFER_SIZE_256 - strlen(m_line) - 1, "%s", line);
+            break;
+        }
+        ++lineCount;
+        line = strtok_r(NULL, PARSE_LINE_BREAK, &rest);
+    }
+}
+
+static void get_vmem_info(char* m_buf, char* m_line, const char* m_containerid) {
+    char *line, *rest = NULL;
+    line = strtok_r(m_buf, PARSE_LINE_BREAK, &rest);
+    while (line != NULL) {
+        if (strstr(line, m_containerid)) {
+            // 由于文件位数较为固定，直接读26位后的数据，提高性能
+            snprintf(m_line + strlen(m_line), BUFFER_SIZE_256 - strlen(m_line) - 1, "%s", line + 26);
+            break;
+        }
+        line = strtok_r(NULL, PARSE_LINE_BREAK, &rest);
+    }
+}
+
+static void print_meminfo(const char* m_file, const char* m_containerid = NULL) {
+    int file_fd = -1;
+    char* buf = NULL;
+    char mem_info_line[BUFFER_SIZE_256] = {0};
+
+    file_fd = open(m_file, O_RDONLY | O_CLOEXEC);
+    if (file_fd < 0) {
+        ALOGE("%s file open fail", m_file);
+    } else  {
+        buf = m_containerid == NULL ? read_file(file_fd, PAGE_SIZE) : read_file(file_fd);
+        if (buf == NULL) {
+            ALOGE("read %s file fail!", m_file);
+        } else {
+            if (m_containerid == NULL) {
+                get_mem_info(buf, mem_info_line);
+                ALOGI("MemInfo:[%s]", mem_info_line);
+            } else {
+                get_vmem_info(buf, mem_info_line, m_containerid); 
+                ALOGI("VMemInfo:[%s]", mem_info_line);
+            }
+        }
+        close_fd(file_fd);
+    }
+}
+
+static void get_containerdata(char* buf, const char* path, ssize_t bytes) {
+    //确保container文件在读取前一定生成
+    int redo_time = 5;
+    int pathFd = -1;
+    while ((pathFd = open(path, O_RDONLY | O_CLOEXEC)) < 0 && --redo_time >= 0) {
+        sleep(1);
+    }
+    if (pathFd < 0) {
+        ALOGE("%s file open fail", path);
+        return;
+    }
+    ssize_t bytesRead = read(pathFd, buf, bytes - 1);
+    if (bytesRead < 0) {
+        ALOGE("Read %s flie fail", path);
+    }
+    close_fd(pathFd);
+}
+
+
 static void mainloop(void) {
     struct event_handler_info* handler_info;
     struct polling_params poll_params;
     struct timespec curr_tm;
     struct epoll_event *evt;
     long delay = -1;
+    long delayEpollTime = 20000;
+    int time_difference = property_get_int32("sys.lmk.time_difference", 30);
+    char container_id[BUFFER_SIZE_13] = {0};
+
+    get_containerdata(container_id, CONTAINERID_PATH, BUFFER_SIZE_13);
+    std::time_t timeTemp = std::time(nullptr);
 
     poll_params.poll_handler = NULL;
     poll_params.paused_handler = NULL;
@@ -3185,7 +3332,6 @@ static void mainloop(void) {
 
                 /* Wait for events until the next polling timeout */
                 nevents = epoll_wait(epollfd, events, maxevents, delay);
-
                 /* Update current time after wait */
                 clock_gettime(CLOCK_MONOTONIC_COARSE, &curr_tm);
                 poll_now = (get_time_diff_ms(&poll_params.last_poll_tm, &curr_tm) >=
@@ -3210,8 +3356,8 @@ static void mainloop(void) {
                     }
                 }
             } else {
-                /* Wait for events with no timeout */
-                nevents = epoll_wait(epollfd, events, maxevents, -1);
+                /* epoll事件阻塞事件修改为delayEpollTime毫秒*/
+                nevents = epoll_wait(epollfd, events, maxevents, delayEpollTime);
             }
         }
 
@@ -3251,6 +3397,17 @@ static void mainloop(void) {
                 call_handler(handler_info, &poll_params, evt->events);
             }
         }
+
+        /* 获取time_t类型开销小于property_get_int32 */
+        std::time_t timeStamp = std::time(nullptr);
+        if (timeStamp - timeTemp >= time_difference) {
+            if (property_get_int32("sys.lmk.memlog_activationflag", 1)) {
+                timeTemp = timeStamp;
+                print_meminfo(MEMINFO_PATH);
+                print_meminfo(VMEMINFO_PATH, container_id);
+                time_difference = property_get_int32("sys.lmk.time_difference", 30);
+            }
+        }
     }
 }
 
